
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../prepare_genome/">
      
      
        <link rel="next" href="../update_genome_metadata/">
      
      
      <link rel="icon" href="../../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>prepare_seq_region - Ensembl-Genomio</title>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/stylesheets/main.4b4a2bd9.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../../.." title="Ensembl-Genomio" class="md-header__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-Genomio
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              prepare_seq_region
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../../.." title="Ensembl-Genomio" class="md-nav__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../../images/logo.png" alt="logo">

    </a>
    Ensembl-Genomio
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    src
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            src
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [brc4]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [brc4]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1_1" id="__nav_4_1_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [runnable]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [runnable]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2" id="__nav_4_1_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [io]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2">
            <span class="md-nav__icon md-icon"></span>
            [io]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2_1" id="__nav_4_1_1_2_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [genomio]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            [genomio]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../assembly/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    assembly
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_1" id="__nav_4_1_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/get_assembly_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_assembly_data
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../db_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    db_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../events/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    events
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_3" id="__nav_4_1_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events/format_events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_events
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_loader
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../fastaprep/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    fastaprep
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_6" id="__nav_4_1_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            fastaprep
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../fastaprep/process_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_fasta
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genbank/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genbank
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_7" id="__nav_4_1_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/extract_from_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/get_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_genbank
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_8" id="__nav_4_1_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/compare_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/dump_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/dump_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../gff3/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    gff3
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_9" id="__nav_4_1_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/functional_annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    functional_annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/process_gff3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_gff3
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest_maker/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_maker
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_13" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_13" id="__nav_4_1_1_2_1_13_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_13_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_13">
            <span class="md-nav__icon md-icon"></span>
            metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    prepare_seq_region
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    prepare_seq_region
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region" class="md-nav__link">
    prepare_seq_region
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
    <nav class="md-nav" aria-label="InputSchema">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.InputSchema.reformat_args" class="md-nav__link">
    reformat_args()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.UnknownMetadata" class="md-nav__link">
    UnknownMetadata
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.add_insdc_seq_region_name" class="md-nav__link">
    add_insdc_seq_region_name()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.add_mitochondrial_codon_table" class="md-nav__link">
    add_mitochondrial_codon_table()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.add_translation_table" class="md-nav__link">
    add_translation_table()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.exclude_seq_regions" class="md-nav__link">
    exclude_seq_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_codon_table" class="md-nav__link">
    get_codon_table()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_gbff_seq_regions" class="md-nav__link">
    get_gbff_seq_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_genbank_id" class="md-nav__link">
    get_genbank_id()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_organelle" class="md-nav__link">
    get_organelle()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_report_regions" class="md-nav__link">
    get_report_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.make_seq_region" class="md-nav__link">
    make_seq_region()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.merge_seq_regions" class="md-nav__link">
    merge_seq_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.prepare_seq_region_metadata" class="md-nav__link">
    prepare_seq_region_metadata()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.report_to_csv" class="md-nav__link">
    report_to_csv()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../update_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    update_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_14" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_14" id="__nav_4_1_1_2_1_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_14_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_14">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json_schema_validator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_validator
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seq_region_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_16" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../utils/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    utils
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_16" id="__nav_4_1_1_2_1_16_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_16_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_16">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/archive_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    archive_utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/json_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [example]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            [example]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../example/calculate/calculations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    calculations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../example/google_docs/google_docs_eg/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    google_docs
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../license/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    License
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region" class="md-nav__link">
    prepare_seq_region
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
    <nav class="md-nav" aria-label="InputSchema">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.InputSchema.reformat_args" class="md-nav__link">
    reformat_args()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.UnknownMetadata" class="md-nav__link">
    UnknownMetadata
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.add_insdc_seq_region_name" class="md-nav__link">
    add_insdc_seq_region_name()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.add_mitochondrial_codon_table" class="md-nav__link">
    add_mitochondrial_codon_table()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.add_translation_table" class="md-nav__link">
    add_translation_table()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.exclude_seq_regions" class="md-nav__link">
    exclude_seq_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_codon_table" class="md-nav__link">
    get_codon_table()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_gbff_seq_regions" class="md-nav__link">
    get_gbff_seq_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_genbank_id" class="md-nav__link">
    get_genbank_id()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_organelle" class="md-nav__link">
    get_organelle()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.get_report_regions" class="md-nav__link">
    get_report_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.make_seq_region" class="md-nav__link">
    make_seq_region()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.merge_seq_regions" class="md-nav__link">
    merge_seq_regions()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.prepare_seq_region_metadata" class="md-nav__link">
    prepare_seq_region_metadata()
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.metadata.prepare_seq_region.report_to_csv" class="md-nav__link">
    report_to_csv()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>prepare_seq_region</h1>

<div class="doc doc-object doc-module">



<a id="src.ensembl.io.genomio.metadata.prepare_seq_region"></a>
  <div class="doc doc-contents first">
  
      <p>Construct a seq_region metadata file from INSDC files.</p>

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.InputSchema" class="doc doc-heading">
          <code>InputSchema</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="argschema.ArgSchema">ArgSchema</span></code></p>

  
      <p>Input arguments expected by the entry point of this module.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InputSchema</span><span class="p">(</span><span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchema</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Input arguments expected by the entry point of this module.&quot;&quot;&quot;</span>

    <span class="n">genome_file</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">InputFile</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Genome metadata JSON file path&quot;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">report_file</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">InputFile</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC/RefSeq sequences report file path to parse&quot;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">gbff_file</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">InputFile</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">allow_none</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">dump_default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC/RefSeq GBFF file path to parse&quot;</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">dst_dir</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">OutputDir</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">dump_default</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span>
        <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Output folder for the processed sequence regions JSON file&quot;</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">brc4_mode</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">Boolean</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">dump_default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Include INSDC sequence region names&quot;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">to_exclude</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">String</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">dump_default</span><span class="o">=</span><span class="s2">&quot;None&quot;</span><span class="p">,</span>
        <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence region names to exclude, in a list-like string&quot;</span><span class="p">},</span>
    <span class="p">)</span>

    <span class="nd">@post_load</span>
    <span class="k">def</span> <span class="nf">reformat_args</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span>  <span class="c1"># pylint: disable=unused-argument</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Processes arguments to may need additional parsing after being correctly loaded.</span>

<span class="sd">        Args:</span>
<span class="sd">            data: Loaded arguments.</span>
<span class="sd">            **kwargs: Arbitrary keyword arguments.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># &quot;to_exclude&quot; argument will be a list-like string that needs to be parsed into a list</span>
        <span class="n">data</span><span class="p">[</span><span class="s2">&quot;to_exclude&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ast</span><span class="o">.</span><span class="n">literal_eval</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;to_exclude&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.metadata.prepare_seq_region.InputSchema.reformat_args" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">reformat_args</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Processes arguments to may need additional parsing after being correctly loaded.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>data</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Loaded arguments.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>**kwargs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Arbitrary keyword arguments.</p>
            </div>
          </td>
          <td>
                <code>{}</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@post_load</span>
<span class="k">def</span> <span class="nf">reformat_args</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="o">**</span><span class="n">kwargs</span>  <span class="c1"># pylint: disable=unused-argument</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Processes arguments to may need additional parsing after being correctly loaded.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: Loaded arguments.</span>
<span class="sd">        **kwargs: Arbitrary keyword arguments.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># &quot;to_exclude&quot; argument will be a list-like string that needs to be parsed into a list</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;to_exclude&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ast</span><span class="o">.</span><span class="n">literal_eval</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;to_exclude&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.UnknownMetadata" class="doc doc-heading">
          <code>UnknownMetadata</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>If a metadata if not supported or recognized.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">80</span>
<span class="normal">81</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">UnknownMetadata</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;If a metadata if not supported or recognized.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>

</div>



<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.add_insdc_seq_region_name" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">add_insdc_seq_region_name</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">synonym_sources</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the list of sequence regions with their corresponding INSDC sequence region names.</p>
<p>"BRC4_seq_region_name" and "EBI_seq_region_name" fields are added to each sequence region: the
former will contain the corresponding INSDC name whilst the latter will contain the current name.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegion">SeqRegion</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>synonym_sources</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Synonym sources to use for the BRC4 name, in order of preference.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.metadata.prepare_seq_region.UnknownMetadata" href="#src.ensembl.io.genomio.metadata.prepare_seq_region.UnknownMetadata">UnknownMetadata</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If no synonym name is found for a sequence region.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_insdc_seq_region_name</span><span class="p">(</span>
    <span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">synonym_sources</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the list of sequence regions with their corresponding INSDC sequence region names.</span>

<span class="sd">    &quot;BRC4_seq_region_name&quot; and &quot;EBI_seq_region_name&quot; fields are added to each sequence region: the</span>
<span class="sd">    former will contain the corresponding INSDC name whilst the latter will contain the current name.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        synonym_sources: Synonym sources to use for the BRC4 name, in order of preference.</span>

<span class="sd">    Raises:</span>
<span class="sd">        UnknownMetadata: If no synonym name is found for a sequence region.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">synonym_sources</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">synonym_sources</span> <span class="o">=</span> <span class="n">SYNONYM_RESOURCES</span>
    <span class="n">new_seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="n">names</span> <span class="o">=</span> <span class="p">{</span><span class="n">synonym</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]:</span> <span class="n">synonym</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">synonym</span> <span class="ow">in</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;synonyms&quot;</span><span class="p">,</span> <span class="p">[])}</span>
        <span class="c1"># Choose the synonym to use as the BRC name, using the first valid name from the source list</span>
        <span class="k">for</span> <span class="n">source_name</span> <span class="ow">in</span> <span class="n">synonym_sources</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">source_name</span> <span class="ow">in</span> <span class="n">names</span><span class="p">:</span>
                <span class="n">brc_name</span> <span class="o">=</span> <span class="n">names</span><span class="p">[</span><span class="n">source_name</span><span class="p">]</span>
                <span class="k">break</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Cannot set BRC4 sequence region name for </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">brc_name</span> <span class="o">=</span> <span class="n">brc_name</span><span class="o">.</span><span class="n">partition</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;BRC4_seq_region_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">brc_name</span>
        <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;EBI_seq_region_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="n">new_seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">new_seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.add_mitochondrial_codon_table" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">taxon_id</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</p>
<p>If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegion">SeqRegion</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>taxon_id</code></td>
          <td>
                <code>int</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The species taxon ID.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">taxon_id</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the mitochondrial codon table to each sequence region (when missing) based on the taxon ID.</span>

<span class="sd">    If no mitochondrial genetic code can be found for the given taxon ID nothing will be changed.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        taxon_id: The species taxon ID.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">url</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;https://www.ebi.ac.uk/ena/data/taxonomy/v1/taxon/tax-id/</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">taxon_id</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Content-Type&quot;</span><span class="p">:</span> <span class="s2">&quot;application/json&quot;</span><span class="p">},</span> <span class="n">timeout</span><span class="o">=</span><span class="mi">60</span><span class="p">)</span>
    <span class="n">decoded</span> <span class="o">=</span> <span class="n">response</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
    <span class="k">if</span> <span class="s2">&quot;mitochondrialGeneticCode&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">decoded</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No mitochondria genetic code found for taxon </span><span class="si">{</span><span class="n">taxon_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">genetic_code</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">decoded</span><span class="p">[</span><span class="s2">&quot;mitochondrialGeneticCode&quot;</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">seqr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="c1"># Do not overwrite any existing mitochondrial codon table</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">location</span> <span class="o">==</span> <span class="s2">&quot;mitochondrial_chromosome&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">):</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genetic_code</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.add_translation_table" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">add_translation_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">location_codon</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Adds the translation codon table to each sequence region (when missing) based on its location.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegion">SeqRegion</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>location_codon</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, int]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of known codon tables for known locations.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_translation_table</span><span class="p">(</span>
    <span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">location_codon</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds the translation codon table to each sequence region (when missing) based on its location.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        location_codon: Map of known codon tables for known locations.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">location_codon</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">location_codon</span> <span class="o">=</span> <span class="n">LOCATION_CODON</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="c1"># Do not overwrite any existing codon table</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;codon_table&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">location_codon</span><span class="p">):</span>
            <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location_codon</span><span class="p">[</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]]</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.exclude_seq_regions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the list of sequence regions with the ones from the exclusion list removed.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_regions</code></td>
          <td>
                <code><span title="typing.List">List</span>[<span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegion">SeqRegion</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence regions.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>to_exclude</code></td>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence region names to exclude.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span>
<span class="normal">98</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">],</span> <span class="n">to_exclude</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the list of sequence regions with the ones from the exclusion list removed.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_regions: Sequence regions.</span>
<span class="sd">        to_exclude: Sequence region names to exclude.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filtered_seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;name&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Remove seq_region </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">filtered_seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">filtered_seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.get_codon_table" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the codon table number from a given a GenBank sequence record (if present).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>record</code></td>
          <td>
                <code><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>GenBank sequence record.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the codon table number from a given a GenBank sequence record (if present).</span>

<span class="sd">    Args:</span>
<span class="sd">        record: GenBank sequence record.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">table_number</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;transl_table&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">table_number</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;transl_table&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">table_number</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.get_gbff_seq_regions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the sequence regions found in the GBFF file (if any).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gbff_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to GBFF file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegionDict">SeqRegionDict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A dict of SeqRegions, with their name as the key.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequence regions found in the GBFF file (if any).</span>

<span class="sd">    Args:</span>
<span class="sd">        gbff_path: Path to GBFF file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dict of SeqRegions, with their name as the key.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gbff_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
            <span class="n">seqr</span><span class="p">:</span> <span class="n">SeqRegion</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)}</span>
            <span class="c1"># Is the seq_region circular?</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span>
            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;topology&quot;</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span><span class="p">):</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;circular&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Is there a genetic code defined?</span>
            <span class="n">codon_table</span> <span class="o">=</span> <span class="n">get_codon_table</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">codon_table</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;codon_table&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">codon_table</span>
            <span class="c1"># Is it an organelle?</span>
            <span class="n">location</span> <span class="o">=</span> <span class="n">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">location</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">location</span>
            <span class="c1"># Is there a comment stating the Genbank record this is based on?</span>
            <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">genbank_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="s2">&quot;INSDC&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">genbank_id</span><span class="p">}]</span>
            <span class="c1"># Store the seq_region</span>
            <span class="n">seq_regions</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqr</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.get_genbank_id" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the GenBank accession from a given sequence record (if present).</p>
<p>Only useful for RefSeq sequence records, where the GenBank accession is stored in a comment.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>record</code></td>
          <td>
                <code><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence record.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_genbank_id</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the GenBank accession from a given sequence record (if present).</span>

<span class="sd">    Only useful for RefSeq sequence records, where the GenBank accession is stored in a comment.</span>

<span class="sd">    Args:</span>
<span class="sd">        record: Sequence record.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genbank_id</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="s2">&quot;comment&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
        <span class="n">comment</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;comment&quot;</span><span class="p">]</span>
        <span class="n">comment</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ \n\r]+&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">comment</span><span class="p">)</span>
        <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;The reference sequence was derived from ([^\.]+)\.&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
            <span class="n">genbank_id</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genbank_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.get_organelle" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">molecule_location</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the organelle location from the given GenBank record (if present).</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>record</code></td>
          <td>
                <code><span title="Bio.SeqRecord.SeqRecord">SeqRecord</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>GenBank sequence record.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>molecule_location</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of sequence type to SO location.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>KeyError</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If the location is not part of the controlled vocabulary.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_organelle</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the organelle location from the given GenBank record (if present).</span>

<span class="sd">    Args:</span>
<span class="sd">        record: GenBank sequence record.</span>
<span class="sd">        molecule_location: Map of sequence type to SO location.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the location is not part of the controlled vocabulary.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">molecule_location</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">molecule_location</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span>
    <span class="n">location</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;organelle&quot;</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
            <span class="n">organelle</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;organelle&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">organelle</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="c1"># Remove plastid prefix</span>
            <span class="n">with_prefix</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^(plastid|mitochondrion):(.+)$&quot;</span><span class="p">,</span> <span class="n">organelle</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">with_prefix</span><span class="p">:</span>
                <span class="n">organelle</span> <span class="o">=</span> <span class="n">with_prefix</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="c1"># Get controlled name</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">location</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">organelle</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">organelle</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">location</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.get_report_regions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_report_regions</span><span class="p">(</span><span class="n">report_path</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the sequence regions found in the report file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>report_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to the sequence regions report file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>is_refseq</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>True if the source of the report is RefSeq, false if INSDC.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegionDict">SeqRegionDict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A dict of SeqRegions, with their name as the key.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_report_regions</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegionDict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the sequence regions found in the report file.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: Path to the sequence regions report file.</span>
<span class="sd">        is_refseq: True if the source of the report is RefSeq, false if INSDC.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A dict of SeqRegions, with their name as the key.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the report in a CSV format</span>
    <span class="n">report_csv</span> <span class="o">=</span> <span class="n">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Feed the csv string to the CSV reader</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">DictReader</span><span class="p">(</span><span class="n">report_csv</span><span class="o">.</span><span class="n">splitlines</span><span class="p">(),</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">quoting</span><span class="o">=</span><span class="n">csv</span><span class="o">.</span><span class="n">QUOTE_NONE</span><span class="p">)</span>
    <span class="c1"># Create the seq_regions</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
        <span class="n">seq_region</span> <span class="o">=</span> <span class="n">make_seq_region</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">seq_region</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="n">seq_regions</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_region</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.main" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Module's entry-point.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Module&#39;s entry-point.&quot;&quot;&quot;</span>
    <span class="n">mod</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchemaParser</span><span class="p">(</span><span class="n">schema_type</span><span class="o">=</span><span class="n">InputSchema</span><span class="p">)</span>
    <span class="n">prepare_seq_region_metadata</span><span class="p">(</span>
        <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;genome_file&quot;</span><span class="p">],</span>
        <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;report_file&quot;</span><span class="p">],</span>
        <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;gbff_file&quot;</span><span class="p">],</span>
        <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;dst_dir&quot;</span><span class="p">],</span>
        <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;brc4_mode&quot;</span><span class="p">],</span>
        <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;to_exclude&quot;</span><span class="p">],</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.make_seq_region" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">make_seq_region</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">,</span> <span class="n">synonym_map</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">molecule_location</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns a sequence region from the information provided.</p>
<p>An empty sequence region will be returned if not accession information is found.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>data</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>a dict from the report representing one line, where the key is the column name.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>is_refseq</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>True if the source is RefSeq, false if INSDC.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>synonym_map</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of INSDC report column names to sequence region field names.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>molecule_location</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Map of sequence type to SO location.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>KeyError</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If the sequence location is not recognised.</p>
            </div>
          </td>
        </tr>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.metadata.prepare_seq_region.UnknownMetadata" href="#src.ensembl.io.genomio.metadata.prepare_seq_region.UnknownMetadata">UnknownMetadata</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>if the sequence role is not recognised.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">make_seq_region</span><span class="p">(</span>
    <span class="n">data</span><span class="p">:</span> <span class="n">Dict</span><span class="p">,</span>
    <span class="n">is_refseq</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">synonym_map</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">molecule_location</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRegion</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a sequence region from the information provided.</span>

<span class="sd">    An empty sequence region will be returned if not accession information is found.</span>

<span class="sd">    Args:</span>
<span class="sd">        data: a dict from the report representing one line, where the key is the column name.</span>
<span class="sd">        is_refseq: True if the source is RefSeq, false if INSDC.</span>
<span class="sd">        synonym_map: Map of INSDC report column names to sequence region field names.</span>
<span class="sd">        molecule_location: Map of sequence type to SO location.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If the sequence location is not recognised.</span>
<span class="sd">        UnknownMetadata: if the sequence role is not recognised.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">synonym_map</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">synonym_map</span> <span class="o">=</span> <span class="n">SYNONYM_MAP</span>
    <span class="k">if</span> <span class="n">molecule_location</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">molecule_location</span> <span class="o">=</span> <span class="n">MOLECULE_LOCATION</span>
    <span class="n">seq_region</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Set accession as the sequence region name</span>
    <span class="n">src</span> <span class="o">=</span> <span class="s2">&quot;RefSeq&quot;</span> <span class="k">if</span> <span class="n">is_refseq</span> <span class="k">else</span> <span class="s2">&quot;GenBank&quot;</span>
    <span class="n">accession_id</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s2">-Accn&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">accession_id</span> <span class="ow">and</span> <span class="p">(</span><span class="n">accession_id</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession_id</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;No </span><span class="si">{</span><span class="n">src</span><span class="si">}</span><span class="s1"> accession ID found for </span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;Sequence-Name&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>
    <span class="c1"># Add synonyms</span>
    <span class="n">synonyms</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">field</span><span class="p">,</span> <span class="n">source</span> <span class="ow">in</span> <span class="n">synonym_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
            <span class="n">synonym</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">source</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]}</span>
            <span class="n">synonyms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">synonym</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">synonyms</span><span class="p">:</span>
        <span class="n">synonyms</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">])</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonyms</span>
    <span class="c1"># Add sequence length</span>
    <span class="n">field</span> <span class="o">=</span> <span class="s2">&quot;Sequence-Length&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">field</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">casefold</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;na&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">field</span><span class="p">])</span>
    <span class="c1"># Add coordinate system and location</span>
    <span class="n">seq_role</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Sequence-Role&quot;</span><span class="p">]</span>
    <span class="c1"># Scaffold?</span>
    <span class="k">if</span> <span class="n">seq_role</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;unplaced-scaffold&quot;</span><span class="p">,</span> <span class="s2">&quot;unlocalized-scaffold&quot;</span><span class="p">):</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;scaffold&quot;</span>
    <span class="c1"># Chromosome? Check location</span>
    <span class="k">elif</span> <span class="n">seq_role</span> <span class="o">==</span> <span class="s2">&quot;assembled-molecule&quot;</span><span class="p">:</span>
        <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
        <span class="n">location</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;Assigned-Molecule-Location/Type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="c1"># Get location metadata</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">seq_region</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">molecule_location</span><span class="p">[</span><span class="n">location</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence location: </span><span class="si">{</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">UnknownMetadata</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized sequence role: </span><span class="si">{</span><span class="n">seq_role</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq_region</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.merge_seq_regions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">merge_seq_regions</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Merges sequence regions from different sources.</p>
<p>When combining two regions matching the same key, the "right" seq region data will take precedence
over the "left".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>left</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegionDict">SeqRegionDict</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Dictionary of sequence regions with names as keys.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>right</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegionDict">SeqRegionDict</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Dictionary of sequence regions with names as keys.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[<span title="src.ensembl.io.genomio.metadata.prepare_seq_region.SeqRegion">SeqRegion</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list of merged sequence regions, sorted by their name.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">merge_seq_regions</span><span class="p">(</span>
    <span class="n">left</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">right</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">SeqRegionDict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRegion</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Merges sequence regions from different sources.</span>

<span class="sd">    When combining two regions matching the same key, the &quot;right&quot; seq region data will take precedence</span>
<span class="sd">    over the &quot;left&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        left: Dictionary of sequence regions with names as keys.</span>
<span class="sd">        right: Dictionary of sequence regions with names as keys.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list of merged sequence regions, sorted by their name.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">left</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">left</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">right</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">right</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Get all the names</span>
    <span class="n">all_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">left</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">right</span><span class="p">)</span>
    <span class="c1"># Create the seq_regions, merge if needed</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">all_names</span><span class="p">:</span>
        <span class="n">left_seqr</span> <span class="o">=</span> <span class="n">left</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{})</span>
        <span class="n">right_seqr</span> <span class="o">=</span> <span class="n">right</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="p">{})</span>
        <span class="k">if</span> <span class="n">left_seqr</span> <span class="ow">and</span> <span class="n">right_seqr</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">left_seqr</span><span class="p">,</span> <span class="o">**</span><span class="n">right_seqr</span><span class="p">}</span>
        <span class="k">elif</span> <span class="n">left_seqr</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="n">left_seqr</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">final_seqr</span> <span class="o">=</span> <span class="n">right_seqr</span>
        <span class="n">seq_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">final_seqr</span><span class="p">)</span>
    <span class="n">seq_regions</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">seq_regions</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.prepare_seq_region_metadata" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">prepare_seq_region_metadata</span><span class="p">(</span><span class="n">genome_file</span><span class="p">,</span> <span class="n">report_file</span><span class="p">,</span> <span class="n">gbff_file</span><span class="p">,</span> <span class="n">dst_dir</span><span class="p">,</span> <span class="n">brc4_mode</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">to_exclude</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Prepares the sequence region metadata found in the INSDC/RefSeq report and GBFF files.</p>
<p>The sequence region information is loaded from both sources and combined. Elements are added/excluded
as requested, and the final sequence region metadata is dumped in a JSON file that follows the schema
defined in "schemas/seq_region_schema.json".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>genome_file</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Genome metadata JSON file path.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>report_file</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>INSDC/RefSeq sequences report file path to parse.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gbff_file</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="os.PathLike">PathLike</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>INSDC/RefSeq GBFF file path to parse.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>dst_dir</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Output folder for the processed sequence regions JSON file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>brc4_mode</code></td>
          <td>
                <code>bool</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Include INSDC sequence region names?</p>
            </div>
          </td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>to_exclude</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence region names to exclude.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">prepare_seq_region_metadata</span><span class="p">(</span>
    <span class="n">genome_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">report_file</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">gbff_file</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">],</span>
    <span class="n">dst_dir</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span>
    <span class="n">brc4_mode</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">to_exclude</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepares the sequence region metadata found in the INSDC/RefSeq report and GBFF files.</span>

<span class="sd">    The sequence region information is loaded from both sources and combined. Elements are added/excluded</span>
<span class="sd">    as requested, and the final sequence region metadata is dumped in a JSON file that follows the schema</span>
<span class="sd">    defined in &quot;schemas/seq_region_schema.json&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        genome_file: Genome metadata JSON file path.</span>
<span class="sd">        report_file: INSDC/RefSeq sequences report file path to parse.</span>
<span class="sd">        gbff_file: INSDC/RefSeq GBFF file path to parse.</span>
<span class="sd">        dst_dir: Output folder for the processed sequence regions JSON file.</span>
<span class="sd">        brc4_mode: Include INSDC sequence region names?</span>
<span class="sd">        to_exclude: Sequence region names to exclude.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_data</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">genome_file</span><span class="p">)</span>
    <span class="n">dst_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">dst_dir</span><span class="p">)</span>
    <span class="n">dst_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Final file name</span>
    <span class="n">metadata_type</span> <span class="o">=</span> <span class="s2">&quot;seq_region&quot;</span>
    <span class="n">metadata_file</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metadata_type</span><span class="si">}</span><span class="s2">.json&quot;</span>
    <span class="n">final_path</span> <span class="o">=</span> <span class="n">dst_dir</span> <span class="o">/</span> <span class="n">metadata_file</span>
    <span class="n">is_refseq</span> <span class="o">=</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">][</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;GCF_&quot;</span><span class="p">)</span>
    <span class="c1"># Get the sequence regions from the report and gbff files, and merge them</span>
    <span class="n">report_regions</span> <span class="o">=</span> <span class="n">get_report_regions</span><span class="p">(</span><span class="n">report_file</span><span class="p">,</span> <span class="n">is_refseq</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">gbff_file</span><span class="p">:</span>
        <span class="n">gbff_regions</span> <span class="o">=</span> <span class="n">get_gbff_seq_regions</span><span class="p">(</span><span class="n">gbff_file</span><span class="p">)</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">merge_seq_regions</span><span class="p">(</span><span class="n">report_regions</span><span class="p">,</span> <span class="n">gbff_regions</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">report_regions</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="c1"># Exclude seq_regions from a list</span>
    <span class="k">if</span> <span class="n">to_exclude</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">exclude_seq_regions</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">to_exclude</span><span class="p">)</span>
    <span class="c1"># Setup the BRC4_seq_region_name</span>
    <span class="k">if</span> <span class="n">brc4_mode</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">add_insdc_seq_region_name</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span>
    <span class="c1"># Add translation and mitochondrial codon tables</span>
    <span class="n">add_translation_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">)</span>
    <span class="n">add_mitochondrial_codon_table</span><span class="p">(</span><span class="n">seq_regions</span><span class="p">,</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;species&quot;</span><span class="p">][</span><span class="s2">&quot;taxonomy_id&quot;</span><span class="p">])</span>
    <span class="c1"># Print out the file</span>
    <span class="n">print_json</span><span class="p">(</span><span class="n">final_path</span><span class="p">,</span> <span class="n">seq_regions</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.metadata.prepare_seq_region.report_to_csv" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns an assembly report as a CSV string.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>report_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>path to a seq_region file from INSDC/RefSeq</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Tuple">Tuple</span>[str, <span title="typing.Dict">Dict</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The data as a string in CSV format, and the head metadata as a dictionary.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/metadata/prepare_seq_region.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">report_to_csv</span><span class="p">(</span><span class="n">report_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns an assembly report as a CSV string.</span>

<span class="sd">    Args:</span>
<span class="sd">        report_path: path to a seq_region file from INSDC/RefSeq</span>

<span class="sd">    Returns:</span>
<span class="sd">        The data as a string in CSV format, and the head metadata as a dictionary.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">report_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">report</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">metadata</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">last_head</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">report</span><span class="p">:</span>
            <span class="c1"># Ignore header</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="c1"># Get metadata values if possible</span>
                <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;# (.+?): (.+?)$&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                    <span class="n">metadata</span><span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)]</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
                <span class="n">last_head</span> <span class="o">=</span> <span class="n">line</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">last_head</span><span class="p">:</span>
                    <span class="n">data</span> <span class="o">+=</span> <span class="n">last_head</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="n">last_head</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">data</span> <span class="o">+=</span> <span class="n">line</span>
        <span class="k">return</span> <span class="n">data</span><span class="p">,</span> <span class="n">metadata</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../../../../..", "features": [], "search": "../../../../../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../../../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
    
  </body>
</html>