
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../extract_annotation/">
      
      
        <link rel="next" href="../../manifest/">
      
      
      <link rel="icon" href="../../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>process - Ensembl-Genomio (v0.1)</title>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/stylesheets/main.4b4a2bd9.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#src.ensembl.io.genomio.gff3.process" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../../.." title="Ensembl-Genomio (v0.1)" class="md-header__button md-logo" aria-label="Ensembl-Genomio (v0.1)" data-md-component="logo">
      
  <img src="../../../../../../img/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-Genomio (v0.1)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              process
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../../.." title="Ensembl-Genomio (v0.1)" class="md-nav__button md-logo" aria-label="Ensembl-Genomio (v0.1)" data-md-component="logo">
      
  <img src="../../../../../../img/logo.png" alt="logo">

    </a>
    Ensembl-Genomio (v0.1)
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    src
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            src
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../brc4/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    brc4
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            brc4
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../brc4/runnable/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    runnable
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_1_1" id="__nav_4_1_1_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            runnable
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2" id="__nav_4_1_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    io
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2">
            <span class="md-nav__icon md-icon"></span>
            io
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genomio
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1" id="__nav_4_1_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            genomio
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../assembly/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    assembly
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_1" id="__nav_4_1_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/download/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../database/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    database
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_2" id="__nav_4_1_1_2_1_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_2">
            <span class="md-nav__icon md-icon"></span>
            database
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../database/factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    factory
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../events/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    events
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_3" id="__nav_4_1_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events/dump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events/format/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events/load/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../fasta/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    fasta
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_4" id="__nav_4_1_1_2_1_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_4">
            <span class="md-nav__icon md-icon"></span>
            fasta
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../fasta/process/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genbank/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genbank
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_5" id="__nav_4_1_1_2_1_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_5">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/download/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genbank/extract_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_data
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_6" id="__nav_4_1_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/dump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/extend/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extend
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/prepare/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_stats/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genome_stats
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_7" id="__nav_4_1_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            genome_stats
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_stats/compare/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_stats/dump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_8" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    gff3
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_8" id="__nav_4_1_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_8_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../extract_annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    process
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    process
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process" class="md-nav__link">
    process
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFGeneMerger" class="md-nav__link">
    GFFGeneMerger
  </a>
  
    <nav class="md-nav" aria-label="GFFGeneMerger">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFGeneMerger.merge" class="md-nav__link">
    merge()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFParserCommon" class="md-nav__link">
    GFFParserCommon
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFParserError" class="md-nav__link">
    GFFParserError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier" class="md-nav__link">
    GFFSimplifier
  </a>
  
    <nav class="md-nav" aria-label="GFFSimplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.build_transcript" class="md-nav__link">
    build_transcript()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.cds_gene" class="md-nav__link">
    cds_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.clean_extra_exons" class="md-nav__link">
    clean_extra_exons()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.format_gene_segments" class="md-nav__link">
    format_gene_segments()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.format_mobile_element" class="md-nav__link">
    format_mobile_element()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_cds" class="md-nav__link">
    gene_to_cds()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_exon" class="md-nav__link">
    gene_to_exon()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.generate_stable_id" class="md-nav__link">
    generate_stable_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.move_cds_to_mrna" class="md-nav__link">
    move_cds_to_mrna()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_cds_id" class="md-nav__link">
    normalize_cds_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene" class="md-nav__link">
    normalize_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene_id" class="md-nav__link">
    normalize_gene_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_pseudogene_cds" class="md-nav__link">
    normalize_pseudogene_cds()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_transcript_id" class="md-nav__link">
    normalize_transcript_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.remove_cds_from_pseudogene" class="md-nav__link">
    remove_cds_from_pseudogene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.remove_prefixes" class="md-nav__link">
    remove_prefixes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.simpler_gff3" class="md-nav__link">
    simpler_gff3()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_for_gene" class="md-nav__link">
    transcript_for_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_gene" class="md-nav__link">
    transcript_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.valid_id" class="md-nav__link">
    valid_id()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.Records" class="md-nav__link">
    Records
  </a>
  
    <nav class="md-nav" aria-label="Records">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.Records.to_gff" class="md-nav__link">
    to_gff()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../manifest/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    manifest
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_9" id="__nav_4_1_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            manifest
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/check_integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    check_integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/compute_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compute_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest/generate/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    generate
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_10" id="__nav_4_1_1_2_1_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_10">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_10_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/json/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    json
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_10_1" id="__nav_4_1_1_2_1_10_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="7" aria-labelledby="__nav_4_1_1_2_1_10_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_10_1">
            <span class="md-nav__icon md-icon"></span>
            json
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json/factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json/validate/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    validate
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_11" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../seq_region/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    seq_region
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_11" id="__nav_4_1_1_2_1_11_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_11_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_11">
            <span class="md-nav__icon md-icon"></span>
            seq_region
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/dump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region/prepare/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_12" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../utils/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    utils
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_12" id="__nav_4_1_1_2_1_12_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_12_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_12">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/archive_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    archive_utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/json_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    example
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            example
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../../example/calculate/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    calculate
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_1" id="__nav_4_1_2_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            calculate
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../example/calculate/calculations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    calculations
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../../example/google_docs/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    google_docs
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_2_2" id="__nav_4_1_2_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2_2">
            <span class="md-nav__icon md-icon"></span>
            google_docs
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../example/google_docs/google_docs_eg/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    google_docs_eg
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../license/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    License
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process" class="md-nav__link">
    process
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFGeneMerger" class="md-nav__link">
    GFFGeneMerger
  </a>
  
    <nav class="md-nav" aria-label="GFFGeneMerger">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFGeneMerger.merge" class="md-nav__link">
    merge()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFParserCommon" class="md-nav__link">
    GFFParserCommon
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFParserError" class="md-nav__link">
    GFFParserError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier" class="md-nav__link">
    GFFSimplifier
  </a>
  
    <nav class="md-nav" aria-label="GFFSimplifier">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.build_transcript" class="md-nav__link">
    build_transcript()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.cds_gene" class="md-nav__link">
    cds_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.clean_extra_exons" class="md-nav__link">
    clean_extra_exons()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.format_gene_segments" class="md-nav__link">
    format_gene_segments()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.format_mobile_element" class="md-nav__link">
    format_mobile_element()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_cds" class="md-nav__link">
    gene_to_cds()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_exon" class="md-nav__link">
    gene_to_exon()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.generate_stable_id" class="md-nav__link">
    generate_stable_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.move_cds_to_mrna" class="md-nav__link">
    move_cds_to_mrna()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_cds_id" class="md-nav__link">
    normalize_cds_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene" class="md-nav__link">
    normalize_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene_id" class="md-nav__link">
    normalize_gene_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_pseudogene_cds" class="md-nav__link">
    normalize_pseudogene_cds()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_transcript_id" class="md-nav__link">
    normalize_transcript_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.remove_cds_from_pseudogene" class="md-nav__link">
    remove_cds_from_pseudogene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.remove_prefixes" class="md-nav__link">
    remove_prefixes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.simpler_gff3" class="md-nav__link">
    simpler_gff3()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_for_gene" class="md-nav__link">
    transcript_for_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_gene" class="md-nav__link">
    transcript_gene()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.GFFSimplifier.valid_id" class="md-nav__link">
    valid_id()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.Records" class="md-nav__link">
    Records
  </a>
  
    <nav class="md-nav" aria-label="Records">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.Records.to_gff" class="md-nav__link">
    to_gff()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.gff3.process.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>process</h1>

<div class="doc doc-object doc-module">



<a id="src.ensembl.io.genomio.gff3.process"></a>
  <div class="doc doc-contents first">
  
      <p>Standardize the gene model representation of a GFF3 file, and extract the functional annotation
in a separate file.</p>

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.gff3.process.GFFGeneMerger" class="doc doc-heading">
          <code>GFFGeneMerger</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.gff3.process.GFFParserCommon" href="#src.ensembl.io.genomio.gff3.process.GFFParserCommon">GFFParserCommon</a></code></p>

  
      <p>Specialized class to merge split genes in a GFF3 file, prior to further parsing.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GFFGeneMerger</span><span class="p">(</span><span class="n">GFFParserCommon</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Specialized class to merge split genes in a GFF3 file, prior to further parsing.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merge genes in a gff that are split in multiple lines</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">merged</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">in_gff_fh</span><span class="p">:</span>
                <span class="c1"># Skip comments</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                    <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Parse one line</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>
                    <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">attr_fields</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
                    <span class="n">attrs</span> <span class="o">=</span> <span class="p">{}</span>
                    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">attr_fields</span><span class="p">:</span>
                        <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)</span>
                        <span class="n">attrs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

                    <span class="c1"># Check this is a gene to merge; cache it then</span>
                    <span class="k">if</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_types</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;part&quot;</span> <span class="ow">in</span> <span class="n">attrs</span> <span class="ow">or</span> <span class="s2">&quot;is_ordered&quot;</span> <span class="ow">in</span> <span class="n">attrs</span><span class="p">):</span>
                        <span class="n">to_merge</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span>

                    <span class="c1"># If not, merge previous gene if needed, and print the line</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
                            <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
                            <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                                <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
                            <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                            <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
                            <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>
                            <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
                        <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Print last merged gene if there is one</span>
            <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
                <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                    <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
                <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
                <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">merged</span>

    <span class="k">def</span> <span class="nf">_merge_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">to_merge</span><span class="p">:</span> <span class="n">List</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a single gene gff3 line merged from separate parts.</span>

<span class="sd">        Args:</span>
<span class="sd">            to_merge: List of gff3 lines with gene parts.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merge gene in </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span><span class="si">}</span><span class="s2"> parts&quot;</span><span class="p">)</span>
        <span class="n">min_start</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="n">max_end</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merge part: </span><span class="si">{</span><span class="n">gene</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>

            <span class="k">if</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">min_start</span> <span class="ow">or</span> <span class="n">min_start</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">min_start</span> <span class="o">=</span> <span class="n">start</span>
            <span class="k">if</span> <span class="n">end</span> <span class="o">&gt;</span> <span class="n">max_end</span> <span class="ow">or</span> <span class="n">max_end</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">max_end</span> <span class="o">=</span> <span class="n">end</span>

        <span class="c1"># Take the first line as template and replace things</span>
        <span class="n">new_gene</span> <span class="o">=</span> <span class="n">to_merge</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">new_gene</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">min_start</span><span class="p">)</span>
        <span class="n">new_gene</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">max_end</span><span class="p">)</span>

        <span class="n">attrs</span> <span class="o">=</span> <span class="n">new_gene</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="n">attrs</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;;is_ordered=true&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;;part=\d+/\d+&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">attrs</span><span class="p">)</span>
        <span class="n">new_gene</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">=</span> <span class="n">attrs</span>

        <span class="k">return</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">new_gene</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFGeneMerger.merge" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">merge</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Merge genes in a gff that are split in multiple lines</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">merge</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Merge genes in a gff that are split in multiple lines</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">merged</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">in_gff_fh</span><span class="p">:</span>
            <span class="c1"># Skip comments</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Parse one line</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>
                <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">attr_fields</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
                <span class="n">attrs</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">attr_fields</span><span class="p">:</span>
                    <span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)</span>
                    <span class="n">attrs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

                <span class="c1"># Check this is a gene to merge; cache it then</span>
                <span class="k">if</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_types</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;part&quot;</span> <span class="ow">in</span> <span class="n">attrs</span> <span class="ow">or</span> <span class="s2">&quot;is_ordered&quot;</span> <span class="ow">in</span> <span class="n">attrs</span><span class="p">):</span>
                    <span class="n">to_merge</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span>

                <span class="c1"># If not, merge previous gene if needed, and print the line</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
                        <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
                        <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                            <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
                        <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                        <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
                        <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>
                        <span class="n">to_merge</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Print last merged gene if there is one</span>
        <span class="k">if</span> <span class="n">to_merge</span><span class="p">:</span>
            <span class="n">merged_str</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">line_to_merge</span> <span class="ow">in</span> <span class="n">to_merge</span><span class="p">:</span>
                <span class="n">merged_str</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">line_to_merge</span><span class="p">))</span>
            <span class="n">merged</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">new_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_merge_genes</span><span class="p">(</span><span class="n">to_merge</span><span class="p">)</span>
            <span class="n">out_gff_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">merged</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.gff3.process.GFFParserCommon" class="doc doc-heading">
          <code>GFFParserCommon</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Heritable class to share the list of feature types supported or ignored by the parser</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GFFParserCommon</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Heritable class to share the list of feature types supported or ignored by the parser&quot;&quot;&quot;</span>

    <span class="c1"># Supported gene level biotypes</span>
    <span class="n">gene_types</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ncRNA_gene&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c1"># Exception: non_gene_types that are nonetheless kept in the GFF3 file</span>
    <span class="n">non_gene_types</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;transposable_element&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c1"># Supported transcript level biotypes</span>
    <span class="n">transcript_types</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;C_gene_segment&quot;</span><span class="p">,</span>
        <span class="s2">&quot;guide_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;lnc_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;miRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;misc_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ncRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;piRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pseudogenic_rRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pseudogenic_transcript&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pseudogenic_tRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ribozyme&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RNase_MRP_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RNase_P_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;rRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;scRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;snoRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;snRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SRP_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;telomerase_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;transcript&quot;</span><span class="p">,</span>
        <span class="s2">&quot;tRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;V_gene_segment&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c1"># Biotypes that are ignored, and removed from the final GFF3 file</span>
    <span class="n">ignored_gene_types</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;cDNA_match&quot;</span><span class="p">,</span>
        <span class="s2">&quot;centromere&quot;</span><span class="p">,</span>
        <span class="s2">&quot;D_loop&quot;</span><span class="p">,</span>
        <span class="s2">&quot;direct_repeat&quot;</span><span class="p">,</span>
        <span class="s2">&quot;dispersed_repeat&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gap&quot;</span><span class="p">,</span>
        <span class="s2">&quot;intron&quot;</span><span class="p">,</span>
        <span class="s2">&quot;inverted_repeat&quot;</span><span class="p">,</span>
        <span class="s2">&quot;long_terminal_repeat&quot;</span><span class="p">,</span>
        <span class="s2">&quot;microsatellite&quot;</span><span class="p">,</span>
        <span class="s2">&quot;origin_of_replication&quot;</span><span class="p">,</span>
        <span class="s2">&quot;region&quot;</span><span class="p">,</span>
        <span class="s2">&quot;repeat_region&quot;</span><span class="p">,</span>
        <span class="s2">&quot;satellite_DNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sequence_feature&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sequence_secondary_structure&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sequence_uncertainty&quot;</span><span class="p">,</span>
        <span class="s2">&quot;STS&quot;</span><span class="p">,</span>
        <span class="s2">&quot;tandem_repeat&quot;</span><span class="p">,</span>
        <span class="s2">&quot;telomere&quot;</span><span class="p">,</span>
        <span class="s2">&quot;terminal%2Cinverted&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c1"># Ignored biotypes that are under a gene parent feature</span>
    <span class="n">ignored_transcript_types</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;3&#39;UTR&quot;</span><span class="p">,</span>
        <span class="s2">&quot;5&#39;UTR&quot;</span><span class="p">,</span>
        <span class="s2">&quot;antisense_RNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;intron&quot;</span><span class="p">,</span>
        <span class="s2">&quot;non_canonical_five_prime_splice_site&quot;</span><span class="p">,</span>
        <span class="s2">&quot;non_canonical_three_prime_splice_site&quot;</span><span class="p">,</span>
    <span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.gff3.process.GFFParserError" class="doc doc-heading">
          <code>GFFParserError</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Error when parsing a GFF3 file.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">53</span>
<span class="normal">54</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GFFParserError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Error when parsing a GFF3 file.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier" class="doc doc-heading">
          <code>GFFSimplifier</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.gff3.process.GFFParserCommon" href="#src.ensembl.io.genomio.gff3.process.GFFParserCommon">GFFParserCommon</a></code></p>

  
      <p>Parse a GGF3 file and output a cleaned up GFF3 + annotation json file.</p>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.gff3.process.GFFParserError" href="#src.ensembl.io.genomio.gff3.process.GFFParserError">GFFParserError</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If an error cannot be automatically fixed.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GFFSimplifier</span><span class="p">(</span><span class="n">GFFParserCommon</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse a GGF3 file and output a cleaned up GFF3 + annotation json file.</span>

<span class="sd">    Raises:</span>
<span class="sd">        GFFParserError: If an error cannot be automatically fixed.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Multiple parameters to automate various fixes</span>
    <span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">gene_cds_skip_others</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">allow_pseudogene_with_CDS</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">exclude_seq_regions</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">validate_gene_id</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">min_id_length</span> <span class="o">=</span> <span class="mi">8</span>
    <span class="n">stable_id_prefix</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">current_stable_id_number</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">make_missing_stable_ids</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">records</span> <span class="o">=</span> <span class="n">Records</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">FunctionalAnnotations</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="n">genome_path</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">genome_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">genome_fh</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">genome_fh</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="n">make_missing_stable_ids</span>

    <span class="k">def</span> <span class="nf">simpler_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load a GFF3 from INSDC and rewrite it in a simpler version,</span>
<span class="sd">        and also write a functional_annotation file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">allowed_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_types</span>
        <span class="n">ignored_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignored_gene_types</span>
        <span class="n">transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_types</span>
        <span class="n">allowed_non_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_gene_types</span>
        <span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span>
        <span class="n">to_exclude</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">exclude_seq_regions</span>

        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">:</span>
            <span class="n">fail_types</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_gff_fh</span><span class="p">):</span>
                <span class="n">new_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip seq_region </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">continue</span>

                <span class="c1"># Root features (usually genes)</span>
                <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="c1"># Skip or format depending on the feature type</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">ignored_gene_types</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">transcript_types</span><span class="p">:</span>
                        <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                        <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cds_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">):</span>
                        <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_mobile_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>

                    <span class="c1"># Normalize the gene structure</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">allowed_gene_types</span><span class="p">:</span>
                        <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">allowed_non_gene_types</span><span class="p">:</span>
                        <span class="k">pass</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">fail_types</span><span class="p">[</span><span class="s2">&quot;gene=&quot;</span> <span class="o">+</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="n">message</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Unsupported feature type: </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> (for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
                        <span class="nb">print</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">skip_unrecognized</span><span class="p">:</span>
                            <span class="k">del</span> <span class="n">feat</span>
                            <span class="k">continue</span>

                    <span class="n">new_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_record</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">fail_types</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">skip_unrecognized</span><span class="p">:</span>
                <span class="n">fail_errors</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fail_types</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized types found:</span><span class="se">\n</span><span class="s2">   </span><span class="si">{</span><span class="n">fail_errors</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">format_mobile_element</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Given a mobile_genetic_element feature, transform it into a transposable_element&quot;&quot;&quot;</span>

        <span class="c1"># Change mobile_genetic_element into a transposable_element feature</span>
        <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">:</span>
            <span class="n">mobile_element_type</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mobile_element_type&quot;</span><span class="p">,</span> <span class="p">[])</span>
            <span class="k">if</span> <span class="n">mobile_element_type</span><span class="p">:</span>
                <span class="c1"># Get the type (and name) from the attrib</span>
                <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                    <span class="n">element_type</span><span class="p">,</span> <span class="n">element_name</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                    <span class="n">description</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">element_name</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">element_type</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">description</span> <span class="o">=</span> <span class="n">element_type</span>

                <span class="c1"># Keep the metadata in the description if the type is known</span>
                <span class="k">if</span> <span class="n">element_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;transposon&quot;</span><span class="p">,</span> <span class="s2">&quot;retrotransposon&quot;</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;transposable_element&quot;</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;product&quot;</span><span class="p">):</span>
                        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;product&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">description</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mobile genetic element &#39;mobile_element_type&#39; is not transposon: </span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">return</span> <span class="n">feat</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mobile genetic element does not have a &#39;mobile_element_type&#39; tag&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">feat</span>
        <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> is not a supported TE feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">feat</span>

        <span class="c1"># Generate ID if needed and add it to the functional annotation</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">)</span>
        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>

        <span class="k">return</span> <span class="n">feat</span>

    <span class="k">def</span> <span class="nf">format_gene_segments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the equivalent Ensembl biotype feature for gene segment transcript features.</span>

<span class="sd">        Supported features: &quot;C_gene_segment&quot; and &quot;V_gene_segment&quot;.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript: Gene segment transcript feature.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Change V/C_gene_segment into a its corresponding transcript names</span>
        <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;C_gene_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;V_gene_segment&quot;</span><span class="p">):</span>
            <span class="n">standard_name</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;standard_name&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">biotype</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\b(immunoglobulin|ig)\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
                <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;IG_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">elif</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\bt[- _]cell\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
                <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;TR_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unexpected &#39;standard_name&#39; content for feature </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">standard_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">transcript</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="n">biotype</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">normalize_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a normalized gene structure, separate from the functional elements.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene: Gene object to normalize.</span>
<span class="sd">            functional_annotation: List of feature annotations (appended by this method).</span>
<span class="sd">            fail_types: List of feature types that are not supported (appended by this method).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># New gene ID</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># Gene with no subfeatures: need to create a transcript at least</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for lone gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>

        <span class="c1"># Count features</span>
        <span class="n">fcounter</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">])</span>

        <span class="c1"># Transform gene - CDS to gene-transcript-exon-CDS</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fcounter</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
                <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript-exon feats for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> CDSs)&quot;</span><span class="p">)</span>
                <span class="n">transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">transcripts</span>

            <span class="c1"># Transform gene - exon to gene-transcript-exon</span>
            <span class="k">elif</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
                <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> exons)&quot;</span><span class="p">)</span>
                <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_exon</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Check that we don&#39;t mix</span>
            <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
                <span class="c1"># Move CDS(s) from parent gene to parent mRNA if needed</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">move_cds_to_mrna</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
                <span class="c1"># Special case with extra exons</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_extra_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># Remove CDS from pseudogenes</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># TRANSCRIPTS</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcripts</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">)</span>

        <span class="c1"># PSEUDOGENE CDS IDs</span>
        <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">normalize_pseudogene_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1"># Finally, store gene functional annotation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)</span>

        <span class="c1"># replace qualifiers</span>
        <span class="n">old_gene_qualifiers</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_gene_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]}</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">_normalize_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a normalized transcript.&quot;&quot;&quot;</span>
        <span class="n">allowed_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_types</span>
        <span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span>

        <span class="n">transcripts_to_delete</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_transcript_types</span>
                <span class="ow">and</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignored_transcript_types</span>
            <span class="p">):</span>
                <span class="n">fail_types</span><span class="p">[</span><span class="s2">&quot;transcript=&quot;</span> <span class="o">+</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">message</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unrecognized transcript type: </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="sa">f</span><span class="s2">&quot; for </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">skip_unrecognized</span><span class="p">:</span>
                    <span class="n">transcripts_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
                    <span class="k">continue</span>

            <span class="c1"># New transcript ID</span>
            <span class="n">transcript_number</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_transcript_id</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">transcript_number</span><span class="p">)</span>

            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_gene_segments</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>

            <span class="c1"># Store transcript functional annotation</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">transcript</span><span class="p">,</span> <span class="s2">&quot;transcript&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

            <span class="c1"># Replace qualifiers</span>
            <span class="n">old_transcript_qualifiers</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="s2">&quot;source&quot;</span> <span class="ow">in</span> <span class="n">old_transcript_qualifiers</span><span class="p">:</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_transcript_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>

            <span class="c1"># EXONS AND CDS</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcript_subfeatures</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">transcript</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">transcripts_to_delete</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">transcripts_to_delete</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">_normalize_transcript_subfeatures</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">fail_types</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a transcript with normalized sub-features.&quot;&quot;&quot;</span>
        <span class="n">ignored_transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignored_transcript_types</span>
        <span class="n">cds_found</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">exons_to_delete</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">tcount</span><span class="p">,</span> <span class="n">feat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="c1"># Replace qualifiers</span>
                <span class="n">old_exon_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>
                <span class="k">if</span> <span class="s2">&quot;source&quot;</span> <span class="ow">in</span> <span class="n">old_exon_qualifiers</span><span class="p">:</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">old_exon_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="c1"># New CDS ID</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_cds&quot;</span>

                <span class="c1"># Store CDS functional annotation (only once)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">cds_found</span><span class="p">:</span>
                    <span class="n">cds_found</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;translation&quot;</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

                <span class="c1"># Replace qualifiers</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                    <span class="s2">&quot;phase&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;phase&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">],</span>
                <span class="p">}</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">ignored_transcript_types</span><span class="p">:</span>
                    <span class="n">exons_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tcount</span><span class="p">)</span>
                    <span class="k">continue</span>

                <span class="n">fail_types</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;sub_transcript=</span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">message</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unrecognized exon type for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; (for transcript </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> of type </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span><span class="p">:</span>
                    <span class="n">exons_to_delete</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tcount</span><span class="p">)</span>
                    <span class="k">continue</span>

        <span class="k">if</span> <span class="n">exons_to_delete</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">exons_to_delete</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">transcript_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ncrna</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a gene for lone transcripts: &#39;gene&#39; for tRNA/rRNA, and &#39;ncRNA&#39; for all others</span>

<span class="sd">        Args:</span>
<span class="sd">            ncrna: the transcript for which we want to create a gene.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The gene that contains the transcript.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
        <span class="k">if</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">):</span>
            <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the transcript </span><span class="si">{</span><span class="n">ncrna</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> in a </span><span class="si">{</span><span class="n">new_type</span><span class="si">}</span><span class="s2"> parent feature&quot;</span><span class="p">)</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">ncrna</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">new_type</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">ncrna</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">id</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">cds_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a gene created for a lone CDS.&quot;&quot;&quot;</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Put the lone CDS in gene-mRNA parent features&quot;</span><span class="p">)</span>

        <span class="c1"># Create a transcript, add the CDS</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">cds</span><span class="p">]</span>

        <span class="c1"># Add an exon too</span>
        <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
        <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

        <span class="c1"># Create a gene, add the transcript</span>
        <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">):</span>
            <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">gene_type</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">transcript_for_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a transcript, from a gene without one.&quot;&quot;&quot;</span>

        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;transcript&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">gene_to_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a list of transcripts (with exons), from a gene with only CDS children.&quot;&quot;&quot;</span>

        <span class="n">gene_cds_skip_others</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_cds_skip_others</span>
        <span class="n">transcripts_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">del_transcript</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">cds</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gene_cds_skip_others</span><span class="p">:</span>
                    <span class="n">del_transcript</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                    <span class="s2">&quot;Can not create a chain &#39;transcript - exon - CDS&#39;&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; when the gene children are not all CDSs&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> of type </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> is child of gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>

            <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>

            <span class="c1"># Add to transcript or create a new one</span>
            <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transcripts_dict</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Create new mRNA for </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span>
            <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
            <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>
            <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">del_transcript</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>

        <span class="n">transcripts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transcripts_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="k">return</span> <span class="n">transcripts</span>

    <span class="k">def</span> <span class="nf">build_transcript</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a transcript with same metadata as the gene provided.&quot;&quot;&quot;</span>

        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">move_cds_to_mrna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Move CDS child features of a gene, to the mRNA.</span>

<span class="sd">        This is to fix the case where we have the following structure:</span>
<span class="sd">        gene -&gt; [ mRNA, CDSs ]</span>
<span class="sd">        and change it to</span>
<span class="sd">        gene -&gt; [ mRNA -&gt; [ CDSs ] ]</span>
<span class="sd">        The mRNA might have exons, in which case check that they match the CDS coordinates.</span>

<span class="sd">        Raises an exception if the feature structure is not recognized.</span>

<span class="sd">        Args:</span>
<span class="sd">            A gene with only one transcript, to check and fix.</span>

<span class="sd">        Returns:</span>
<span class="sd">            The gene where the CDSs have been moved, if needed.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># First, count the types</span>
        <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">cdss</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">gene_subf_clean</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">gene_subf_clean</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Nothing to fix here, no CDSs to move</span>
            <span class="k">return</span> <span class="n">gene</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mrnas</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Can&#39;t fix gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: contains several mRNAs and CDSs, all children of the gene&quot;</span>
            <span class="p">)</span>

        <span class="n">mrna</span> <span class="o">=</span> <span class="n">mrnas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Check if there are exons (or CDSs) under the mRNA</span>
        <span class="n">sub_exons</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">sub_cdss</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">sub_cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="n">sub_exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_cdss</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">sub_cdss</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">cdss</span><span class="p">,</span> <span class="n">sub_exons</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: move </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span><span class="si">}</span><span class="s2"> CDSs to the mRNA&quot;</span><span class="p">)</span>
        <span class="c1"># No more issues? move the CDSs</span>
        <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">cdss</span>
        <span class="c1"># And remove them from the gene</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subf_clean</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mrna</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_check_sub_cdss</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">sub_cdss</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_cdss</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> has CDSs as children of the gene and mRNA&quot;</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_check_sub_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">cdss</span><span class="p">,</span> <span class="n">sub_exons</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check that the exons of the mRNA and the CDSs match&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_exons</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># Check that they match the CDS outside</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sub_exons</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">):</span>
                <span class="c1"># Now that all coordinates are the same</span>
                <span class="n">coord_exons</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">exon</span><span class="o">.</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">sub_exons</span><span class="p">]</span>
                <span class="n">coord_cdss</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cds</span> <span class="ow">in</span> <span class="n">cdss</span><span class="p">]</span>

                <span class="k">if</span> <span class="n">coord_exons</span> <span class="o">!=</span> <span class="n">coord_cdss</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> CDSs and exons under the mRNA do not match&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> CDSs and exons under the mRNA do not match (different count)&quot;</span>
                <span class="p">)</span>

    <span class="k">def</span> <span class="nf">clean_extra_exons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove extra exons, already existing in the mRNA.</span>

<span class="sd">        This is a special case where a gene contains proper mRNAs, etc. but also</span>
<span class="sd">        extra exons for the same features. Those exons usually have an ID starting with</span>
<span class="sd">        &quot;id-&quot;, so that&#39;s what we use to detect them.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">exons</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">others</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                <span class="n">exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">others</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">exons</span> <span class="ow">and</span> <span class="n">mrnas</span><span class="p">:</span>
            <span class="n">exon_has_id</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="c1"># Check if the exon ids start with &quot;id-&quot;, which is an indication that they do not belong here</span>
            <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;id-&quot;</span><span class="p">):</span>
                    <span class="n">exon_has_id</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">exon_has_id</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">exon_has_id</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons</span><span class="p">):</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove </span><span class="si">{</span><span class="n">exon_has_id</span><span class="si">}</span><span class="s2"> extra exons from </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">mrnas</span>
                    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">others</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t remove extra exons for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">, not all start with &#39;id-&#39;&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">gene</span>

    <span class="k">def</span> <span class="nf">gene_to_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns an intermediary transcript for a gene with direct exon children.&quot;&quot;&quot;</span>

        <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">transcript</span>

    <span class="k">def</span> <span class="nf">normalize_gene_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove any unnecessary prefixes around the gene ID.</span>

<span class="sd">        Generate a new stable id if it is not recognized as valid.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene: Gene feature to normalize.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A normalized gene id.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene-&quot;</span><span class="p">,</span> <span class="s2">&quot;gene:&quot;</span><span class="p">]</span>
        <span class="n">new_gene_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

        <span class="c1"># In case the gene id is not valid, use the GeneID</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">new_gene_id</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene id is not valid: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">qual</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="k">if</span> <span class="s2">&quot;Dbxref&quot;</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">xref</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">[</span><span class="s2">&quot;Dbxref&quot;</span><span class="p">]:</span>
                    <span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">xref</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">:</span>
                        <span class="n">new_gene_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using GeneID </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> for stable_id instead of </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">return</span> <span class="n">new_gene_id</span>

            <span class="c1"># Make a new stable_id</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">:</span>
                <span class="n">new_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;New id: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">new_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">new_id</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t use invalid gene id for </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new_gene_id</span>

    <span class="k">def</span> <span class="nf">generate_stable_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a new unique gene stable_id with a prefix.</span>

<span class="sd">        The id is made up of a prefix and a number, which is auto incremented.</span>
<span class="sd">        Define the prefix with the param &quot;stable_id_prefix&quot;,</span>
<span class="sd">        or use the genome organism_abbrev and prepend &quot;TMP_&quot; to it.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span>
                <span class="n">org</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;BRC4&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism_abbrev&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">org</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_PREFIX_&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_&quot;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span> <span class="o">=</span> <span class="n">prefix</span>

        <span class="n">number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">new_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">=</span> <span class="n">number</span>

        <span class="k">return</span> <span class="n">new_id</span>

    <span class="k">def</span> <span class="nf">valid_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check that the format of a stable id is valid.&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">validate_gene_id</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>

        <span class="n">min_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_id_length</span>

        <span class="c1"># Trna (from tRNAscan)</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^Trna&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is a Trna from tRNA-scan: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># Coordinates</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^.+:\d+..\d+&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is a coordinate: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># Special characters</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ |]&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id contains special characters: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="c1"># Min length</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_length</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is too short (&lt;</span><span class="si">{</span><span class="n">min_length</span><span class="si">}</span><span class="s2">) </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">False</span>

        <span class="k">return</span> <span class="kc">True</span>

    <span class="k">def</span> <span class="nf">normalize_transcript_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Use a gene ID and a number to make a formatted transcript ID.&quot;&quot;&quot;</span>

        <span class="n">transcript_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="s2">_t</span><span class="si">{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">transcript_id</span>

    <span class="k">def</span> <span class="nf">normalize_cds_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check the CDS ID is proper:</span>
<span class="sd">        - Remove any unnecessary prefixes around the CDS ID</span>
<span class="sd">        - Delete the ID if it is not proper</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cds-&quot;</span><span class="p">,</span> <span class="s2">&quot;cds:&quot;</span><span class="p">]</span>
        <span class="n">cds_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">cds_id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

        <span class="c1"># Special case: if the ID doesn&#39;t look like one, remove it</span>
        <span class="c1"># It needs to be regenerated</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">cds_id</span><span class="p">):</span>
            <span class="n">cds_id</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">return</span> <span class="n">cds_id</span>

    <span class="k">def</span> <span class="nf">normalize_pseudogene_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Ensure CDS from a pseudogene have a proper ID</span>
<span class="sd">        - different from the gene</span>
<span class="sd">        - derived from the gene if it is not proper</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                        <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_cds&quot;</span>
                        <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>

    <span class="k">def</span> <span class="nf">remove_cds_from_pseudogene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Remove CDS from a pseudogene</span>
<span class="sd">        This assumes the CDSs are sub features of the transcript or the gene</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">gene_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="n">new_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="n">new_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">new_subfeats</span>
            <span class="n">gene_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subfeats</span>

    <span class="k">def</span> <span class="nf">remove_prefixes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Remove prefixes from an identifier if they are found</span>
<span class="sd">        Return the unaltered identifier otherwise</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="n">prefixes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">identifier</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">prefix</span><span class="p">):</span>
                <span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="p">:]</span>
        <span class="k">return</span> <span class="n">identifier</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.build_transcript" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">build_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a transcript with same metadata as the gene provided.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">build_transcript</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a transcript with same metadata as the gene provided.&quot;&quot;&quot;</span>

    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.cds_gene" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">cds_gene</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a gene created for a lone CDS.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">cds_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a gene created for a lone CDS.&quot;&quot;&quot;</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Put the lone CDS in gene-mRNA parent features&quot;</span><span class="p">)</span>

    <span class="c1"># Create a transcript, add the CDS</span>
    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">cds</span><span class="p">]</span>

    <span class="c1"># Add an exon too</span>
    <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
    <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

    <span class="c1"># Create a gene, add the transcript</span>
    <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;pseudo&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;true&quot;</span><span class="p">):</span>
        <span class="n">gene_type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">gene_type</span><span class="p">)</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.clean_extra_exons" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">clean_extra_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Remove extra exons, already existing in the mRNA.</p>
<p>This is a special case where a gene contains proper mRNAs, etc. but also
extra exons for the same features. Those exons usually have an ID starting with
"id-", so that's what we use to detect them.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">clean_extra_exons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove extra exons, already existing in the mRNA.</span>

<span class="sd">    This is a special case where a gene contains proper mRNAs, etc. but also</span>
<span class="sd">    extra exons for the same features. Those exons usually have an ID starting with</span>
<span class="sd">    &quot;id-&quot;, so that&#39;s what we use to detect them.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">others</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
            <span class="n">exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
            <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">others</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">exons</span> <span class="ow">and</span> <span class="n">mrnas</span><span class="p">:</span>
        <span class="n">exon_has_id</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="c1"># Check if the exon ids start with &quot;id-&quot;, which is an indication that they do not belong here</span>
        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">exons</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">exon</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;id-&quot;</span><span class="p">):</span>
                <span class="n">exon_has_id</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">exon_has_id</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">exon_has_id</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons</span><span class="p">):</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove </span><span class="si">{</span><span class="n">exon_has_id</span><span class="si">}</span><span class="s2"> extra exons from </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">mrnas</span>
                <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">others</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t remove extra exons for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">, not all start with &#39;id-&#39;&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.format_gene_segments" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">format_gene_segments</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns the equivalent Ensembl biotype feature for gene segment transcript features.</p>
<p>Supported features: "C_gene_segment" and "V_gene_segment".</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>transcript</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Gene segment transcript feature.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">format_gene_segments</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the equivalent Ensembl biotype feature for gene segment transcript features.</span>

<span class="sd">    Supported features: &quot;C_gene_segment&quot; and &quot;V_gene_segment&quot;.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcript: Gene segment transcript feature.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Change V/C_gene_segment into a its corresponding transcript names</span>
    <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;C_gene_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;V_gene_segment&quot;</span><span class="p">):</span>
        <span class="n">standard_name</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;standard_name&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">biotype</span> <span class="o">=</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_segment&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\b(immunoglobulin|ig)\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
            <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;IG_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">elif</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\bt[- _]cell\b&quot;</span><span class="p">,</span> <span class="n">standard_name</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">):</span>
            <span class="n">biotype</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;TR_</span><span class="si">{</span><span class="n">biotype</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unexpected &#39;standard_name&#39; content for feature </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">standard_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">transcript</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="n">biotype</span>
    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.format_mobile_element" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">format_mobile_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Given a mobile_genetic_element feature, transform it into a transposable_element</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">format_mobile_element</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a mobile_genetic_element feature, transform it into a transposable_element&quot;&quot;&quot;</span>

    <span class="c1"># Change mobile_genetic_element into a transposable_element feature</span>
    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">:</span>
        <span class="n">mobile_element_type</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mobile_element_type&quot;</span><span class="p">,</span> <span class="p">[])</span>
        <span class="k">if</span> <span class="n">mobile_element_type</span><span class="p">:</span>
            <span class="c1"># Get the type (and name) from the attrib</span>
            <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">element_type</span><span class="p">,</span> <span class="n">element_name</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                <span class="n">description</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">element_name</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">element_type</span> <span class="o">=</span> <span class="n">mobile_element_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">description</span> <span class="o">=</span> <span class="n">element_type</span>

            <span class="c1"># Keep the metadata in the description if the type is known</span>
            <span class="k">if</span> <span class="n">element_type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;transposon&quot;</span><span class="p">,</span> <span class="s2">&quot;retrotransposon&quot;</span><span class="p">):</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;transposable_element&quot;</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;product&quot;</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;product&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">description</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mobile genetic element &#39;mobile_element_type&#39; is not transposon: </span><span class="si">{</span><span class="n">element_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">feat</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mobile genetic element does not have a &#39;mobile_element_type&#39; tag&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">feat</span>
    <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> is not a supported TE feature </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">feat</span>

    <span class="c1"># Generate ID if needed and add it to the functional annotation</span>
    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">)</span>
    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">}</span>

    <span class="k">return</span> <span class="n">feat</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_cds" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">gene_to_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a list of transcripts (with exons), from a gene with only CDS children.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">gene_to_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a list of transcripts (with exons), from a gene with only CDS children.&quot;&quot;&quot;</span>

    <span class="n">gene_cds_skip_others</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_cds_skip_others</span>
    <span class="n">transcripts_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">del_transcript</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">count</span><span class="p">,</span> <span class="n">cds</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">gene_cds_skip_others</span><span class="p">:</span>
                <span class="n">del_transcript</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
                <span class="s2">&quot;Can not create a chain &#39;transcript - exon - CDS&#39;&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; when the gene children are not all CDSs&quot;</span>
                <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> of type </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> is child of gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
            <span class="p">)</span>

        <span class="n">exon</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">cds</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>

        <span class="c1"># Add to transcript or create a new one</span>
        <span class="k">if</span> <span class="n">cds</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transcripts_dict</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Create new mRNA for </span><span class="si">{</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">build_transcript</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript</span>
        <span class="n">exon</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
        <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>
        <span class="n">transcripts_dict</span><span class="p">[</span><span class="n">cds</span><span class="o">.</span><span class="n">id</span><span class="p">]</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">del_transcript</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">elt</span><span class="p">)</span>

    <span class="n">transcripts</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">transcripts_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">transcripts</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.gene_to_exon" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">gene_to_exon</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns an intermediary transcript for a gene with direct exon children.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">gene_to_exon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns an intermediary transcript for a gene with direct exon children.&quot;&quot;&quot;</span>

    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">exon</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.generate_stable_id" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">generate_stable_id</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a new unique gene stable_id with a prefix.</p>
<p>The id is made up of a prefix and a number, which is auto incremented.
Define the prefix with the param "stable_id_prefix",
or use the genome organism_abbrev and prepend "TMP_" to it.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">generate_stable_id</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a new unique gene stable_id with a prefix.</span>

<span class="sd">    The id is made up of a prefix and a number, which is auto incremented.</span>
<span class="sd">    Define the prefix with the param &quot;stable_id_prefix&quot;,</span>
<span class="sd">    or use the genome organism_abbrev and prepend &quot;TMP_&quot; to it.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span><span class="p">:</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span>
            <span class="n">org</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;BRC4&quot;</span><span class="p">,</span> <span class="p">{})</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;organism_abbrev&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">org</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_PREFIX_&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;TMP_&quot;</span> <span class="o">+</span> <span class="n">org</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">stable_id_prefix</span> <span class="o">=</span> <span class="n">prefix</span>

    <span class="n">number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">new_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">prefix</span><span class="si">}{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">current_stable_id_number</span> <span class="o">=</span> <span class="n">number</span>

    <span class="k">return</span> <span class="n">new_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.move_cds_to_mrna" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">move_cds_to_mrna</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Move CDS child features of a gene, to the mRNA.</p>
<p>This is to fix the case where we have the following structure:
gene -&gt; [ mRNA, CDSs ]
and change it to
gene -&gt; [ mRNA -&gt; [ CDSs ] ]
The mRNA might have exons, in which case check that they match the CDS coordinates.</p>
<p>Raises an exception if the feature structure is not recognized.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The gene where the CDSs have been moved, if needed.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">move_cds_to_mrna</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Move CDS child features of a gene, to the mRNA.</span>

<span class="sd">    This is to fix the case where we have the following structure:</span>
<span class="sd">    gene -&gt; [ mRNA, CDSs ]</span>
<span class="sd">    and change it to</span>
<span class="sd">    gene -&gt; [ mRNA -&gt; [ CDSs ] ]</span>
<span class="sd">    The mRNA might have exons, in which case check that they match the CDS coordinates.</span>

<span class="sd">    Raises an exception if the feature structure is not recognized.</span>

<span class="sd">    Args:</span>
<span class="sd">        A gene with only one transcript, to check and fix.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The gene where the CDSs have been moved, if needed.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First, count the types</span>
    <span class="n">mrnas</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">cdss</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">gene_subf_clean</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
            <span class="n">mrnas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_subf_clean</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Nothing to fix here, no CDSs to move</span>
        <span class="k">return</span> <span class="n">gene</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mrnas</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Can&#39;t fix gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: contains several mRNAs and CDSs, all children of the gene&quot;</span>
        <span class="p">)</span>

    <span class="n">mrna</span> <span class="o">=</span> <span class="n">mrnas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Check if there are exons (or CDSs) under the mRNA</span>
    <span class="n">sub_exons</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">sub_cdss</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subf</span> <span class="ow">in</span> <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">sub_cdss</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">subf</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
            <span class="n">sub_exons</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subf</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_cdss</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">sub_cdss</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_check_sub_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">cdss</span><span class="p">,</span> <span class="n">sub_exons</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">: move </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">cdss</span><span class="p">)</span><span class="si">}</span><span class="s2"> CDSs to the mRNA&quot;</span><span class="p">)</span>
    <span class="c1"># No more issues? move the CDSs</span>
    <span class="n">mrna</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">+=</span> <span class="n">cdss</span>
    <span class="c1"># And remove them from the gene</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subf_clean</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mrna</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_cds_id" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">cds_id</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Check the CDS ID is proper:
- Remove any unnecessary prefixes around the CDS ID
- Delete the ID if it is not proper</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_cds_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cds_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check the CDS ID is proper:</span>
<span class="sd">    - Remove any unnecessary prefixes around the CDS ID</span>
<span class="sd">    - Delete the ID if it is not proper</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cds-&quot;</span><span class="p">,</span> <span class="s2">&quot;cds:&quot;</span><span class="p">]</span>
    <span class="n">cds_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">cds_id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

    <span class="c1"># Special case: if the ID doesn&#39;t look like one, remove it</span>
    <span class="c1"># It needs to be regenerated</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">cds_id</span><span class="p">):</span>
        <span class="n">cds_id</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">return</span> <span class="n">cds_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">normalize_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a normalized gene structure, separate from the functional elements.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Gene object to normalize.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>functional_annotation</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>List of feature annotations (appended by this method).</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>fail_types</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, int]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>List of feature types that are not supported (appended by this method).</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a normalized gene structure, separate from the functional elements.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene: Gene object to normalize.</span>
<span class="sd">        functional_annotation: List of feature annotations (appended by this method).</span>
<span class="sd">        fail_types: List of feature types that are not supported (appended by this method).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># New gene ID</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Gene with no subfeatures: need to create a transcript at least</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for lone gene </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>

    <span class="c1"># Count features</span>
    <span class="n">fcounter</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">([</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">])</span>

    <span class="c1"># Transform gene - CDS to gene-transcript-exon-CDS</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fcounter</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
            <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript-exon feats for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> CDSs)&quot;</span><span class="p">)</span>
            <span class="n">transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">transcripts</span>

        <span class="c1"># Transform gene - exon to gene-transcript-exon</span>
        <span class="k">elif</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
            <span class="n">num_subs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Insert transcript for </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">num_subs</span><span class="si">}</span><span class="s2"> exons)&quot;</span><span class="p">)</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_exon</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">transcript</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Check that we don&#39;t mix</span>
        <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">):</span>
            <span class="c1"># Move CDS(s) from parent gene to parent mRNA if needed</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">move_cds_to_mrna</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">fcounter</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">):</span>
            <span class="c1"># Special case with extra exons</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">clean_extra_exons</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Remove CDS from pseudogenes</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># TRANSCRIPTS</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalize_transcripts</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">)</span>

    <span class="c1"># PSEUDOGENE CDS IDs</span>
    <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">allow_pseudogene_with_CDS</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">normalize_pseudogene_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

    <span class="c1"># Finally, store gene functional annotation</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">)</span>

    <span class="c1"># replace qualifiers</span>
    <span class="n">old_gene_qualifiers</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span> <span class="n">old_gene_qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]}</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_gene_id" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">normalize_gene_id</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Remove any unnecessary prefixes around the gene ID.</p>
<p>Generate a new stable id if it is not recognized as valid.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gene</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Gene feature to normalize.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A normalized gene id.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_gene_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove any unnecessary prefixes around the gene ID.</span>

<span class="sd">    Generate a new stable id if it is not recognized as valid.</span>

<span class="sd">    Args:</span>
<span class="sd">        gene: Gene feature to normalize.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A normalized gene id.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">prefixes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene-&quot;</span><span class="p">,</span> <span class="s2">&quot;gene:&quot;</span><span class="p">]</span>
    <span class="n">new_gene_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span>

    <span class="c1"># In case the gene id is not valid, use the GeneID</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">valid_id</span><span class="p">(</span><span class="n">new_gene_id</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene id is not valid: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">qual</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="k">if</span> <span class="s2">&quot;Dbxref&quot;</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">xref</span> <span class="ow">in</span> <span class="n">qual</span><span class="p">[</span><span class="s2">&quot;Dbxref&quot;</span><span class="p">]:</span>
                <span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="o">=</span> <span class="n">xref</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s2">&quot;GeneID&quot;</span><span class="p">:</span>
                    <span class="n">new_gene_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using GeneID </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> for stable_id instead of </span><span class="si">{</span><span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">return</span> <span class="n">new_gene_id</span>

        <span class="c1"># Make a new stable_id</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">:</span>
            <span class="n">new_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_stable_id</span><span class="p">()</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;New id: </span><span class="si">{</span><span class="n">new_gene_id</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">new_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">new_id</span>
        <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Can&#39;t use invalid gene id for </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">new_gene_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_pseudogene_cds" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">normalize_pseudogene_cds</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Ensure CDS from a pseudogene have a proper ID
- different from the gene
- derived from the gene if it is not proper</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_pseudogene_cds</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Ensure CDS from a pseudogene have a proper ID</span>
<span class="sd">    - different from the gene</span>
<span class="sd">    - derived from the gene if it is not proper</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_cds_id</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">_cds&quot;</span>
                    <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.normalize_transcript_id" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">normalize_transcript_id</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">number</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Use a gene ID and a number to make a formatted transcript ID.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize_transcript_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">number</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Use a gene ID and a number to make a formatted transcript ID.&quot;&quot;&quot;</span>

    <span class="n">transcript_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="s2">_t</span><span class="si">{</span><span class="n">number</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">return</span> <span class="n">transcript_id</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.remove_cds_from_pseudogene" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">remove_cds_from_pseudogene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Remove CDS from a pseudogene
This assumes the CDSs are sub features of the transcript or the gene</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">remove_cds_from_pseudogene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove CDS from a pseudogene</span>
<span class="sd">    This assumes the CDSs are sub features of the transcript or the gene</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gene_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">transcript</span> <span class="ow">in</span> <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">transcript</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">transcript</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="n">new_subfeats</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Remove pseudo CDS </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="n">new_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
        <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">new_subfeats</span>
        <span class="n">gene_subfeats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="n">gene_subfeats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.remove_prefixes" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">remove_prefixes</span><span class="p">(</span><span class="n">identifier</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Remove prefixes from an identifier if they are found
Return the unaltered identifier otherwise</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">remove_prefixes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">identifier</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">prefixes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remove prefixes from an identifier if they are found</span>
<span class="sd">    Return the unaltered identifier otherwise</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">prefix</span> <span class="ow">in</span> <span class="n">prefixes</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">identifier</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">prefix</span><span class="p">):</span>
            <span class="n">identifier</span> <span class="o">=</span> <span class="n">identifier</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="p">:]</span>
    <span class="k">return</span> <span class="n">identifier</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.simpler_gff3" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">simpler_gff3</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Load a GFF3 from INSDC and rewrite it in a simpler version,
and also write a functional_annotation file</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">simpler_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">in_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load a GFF3 from INSDC and rewrite it in a simpler version,</span>
<span class="sd">    and also write a functional_annotation file</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">allowed_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_types</span>
    <span class="n">ignored_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ignored_gene_types</span>
    <span class="n">transcript_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_types</span>
    <span class="n">allowed_non_gene_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_gene_types</span>
    <span class="n">skip_unrecognized</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">skip_unrecognized</span>
    <span class="n">to_exclude</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">exclude_seq_regions</span>

    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_gff_fh</span><span class="p">:</span>
        <span class="n">fail_types</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">in_gff_fh</span><span class="p">):</span>
            <span class="n">new_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">to_exclude</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Skip seq_region </span><span class="si">{</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1"># Root features (usually genes)</span>
            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="c1"># Skip or format depending on the feature type</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">ignored_gene_types</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">transcript_types</span><span class="p">:</span>
                    <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcript_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                    <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">cds_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mobile_genetic_element&quot;</span><span class="p">,</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">):</span>
                    <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">format_mobile_element</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>

                <span class="c1"># Normalize the gene structure</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">allowed_gene_types</span><span class="p">:</span>
                    <span class="n">feat</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">normalize_gene</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="n">fail_types</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="n">allowed_non_gene_types</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">fail_types</span><span class="p">[</span><span class="s2">&quot;gene=&quot;</span> <span class="o">+</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="n">message</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Unsupported feature type: </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> (for </span><span class="si">{</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">)&quot;</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">skip_unrecognized</span><span class="p">:</span>
                        <span class="k">del</span> <span class="n">feat</span>
                        <span class="k">continue</span>

                <span class="n">new_record</span><span class="o">.</span><span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">feat</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_record</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fail_types</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">skip_unrecognized</span><span class="p">:</span>
            <span class="n">fail_errors</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fail_types</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="k">raise</span> <span class="n">GFFParserError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unrecognized types found:</span><span class="se">\n</span><span class="s2">   </span><span class="si">{</span><span class="n">fail_errors</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_for_gene" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">transcript_for_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a transcript, from a gene without one.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">transcript_for_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a transcript, from a gene without one.&quot;&quot;&quot;</span>

    <span class="n">transcript</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;transcript&quot;</span><span class="p">)</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">transcript</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">return</span> <span class="n">transcript</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.transcript_gene" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">transcript_gene</span><span class="p">(</span><span class="n">ncrna</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Create a gene for lone transcripts: 'gene' for tRNA/rRNA, and 'ncRNA' for all others</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>ncrna</code></td>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the transcript for which we want to create a gene.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="Bio.SeqFeature.SeqFeature">SeqFeature</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The gene that contains the transcript.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">transcript_gene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ncrna</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqFeature</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a gene for lone transcripts: &#39;gene&#39; for tRNA/rRNA, and &#39;ncRNA&#39; for all others</span>

<span class="sd">    Args:</span>
<span class="sd">        ncrna: the transcript for which we want to create a gene.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The gene that contains the transcript.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;ncRNA_gene&quot;</span>
    <span class="k">if</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">):</span>
        <span class="n">new_type</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Put the transcript </span><span class="si">{</span><span class="n">ncrna</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2"> in a </span><span class="si">{</span><span class="n">new_type</span><span class="si">}</span><span class="s2"> parent feature&quot;</span><span class="p">)</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">ncrna</span><span class="o">.</span><span class="n">location</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">new_type</span><span class="p">)</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">sub_features</span> <span class="o">=</span> <span class="p">[</span><span class="n">ncrna</span><span class="p">]</span>
    <span class="n">gene</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">ncrna</span><span class="o">.</span><span class="n">id</span>

    <span class="k">return</span> <span class="n">gene</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.GFFSimplifier.valid_id" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">valid_id</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Check that the format of a stable id is valid.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">valid_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check that the format of a stable id is valid.&quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">validate_gene_id</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="n">min_length</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">min_id_length</span>

    <span class="c1"># Trna (from tRNAscan)</span>
    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^Trna&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is a Trna from tRNA-scan: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Coordinates</span>
    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;^.+:\d+..\d+&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is a coordinate: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Special characters</span>
    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;[ |]&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id contains special characters: </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Min length</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_length</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stable id is too short (&lt;</span><span class="si">{</span><span class="n">min_length</span><span class="si">}</span><span class="s2">) </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="k">return</span> <span class="kc">True</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.gff3.process.Records" class="doc doc-heading">
          <code>Records</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>list</code></p>

  
      <p>List of GFF3 SeqRecords.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Records</span><span class="p">(</span><span class="nb">list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;List of GFF3 SeqRecords.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">to_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Print out the current list of records in a GFF3 file.</span>

<span class="sd">        Args:</span>
<span class="sd">            out_gff_path: Path to GFF3 file where to write the records.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
            <span class="n">GFF</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_fh</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.gff3.process.Records.to_gff" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">to_gff</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Print out the current list of records in a GFF3 file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>out_gff_path</code></td>
          <td>
                <code><span title="os.PathLike">PathLike</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to GFF3 file where to write the records.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">to_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Print out the current list of records in a GFF3 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        out_gff_path: Path to GFF3 file where to write the records.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_gff_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_gff_fh</span><span class="p">:</span>
        <span class="n">GFF</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_gff_fh</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>



<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.gff3.process.main" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Main script entry-point.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/gff3/process.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main script entry-point.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Standardize the gene model representation of a GFF3 file, and extract the functional &quot;</span>
            <span class="s2">&quot;annotation in a separate file.&quot;</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span><span class="s2">&quot;--in_gff_path&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Input GFF3 file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_src_path</span><span class="p">(</span><span class="s2">&quot;--genome_data&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Genome JSON file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--make_missing_stable_ids&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Generate stable IDs when missing or invalid&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_dst_path</span><span class="p">(</span><span class="s2">&quot;--out_gff_path&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;gene_models.gff3&quot;</span><span class="p">),</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output GFF3 file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_dst_path</span><span class="p">(</span>
        <span class="s2">&quot;--out_func_path&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;functional_annotation.json&quot;</span><span class="p">),</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Output functional annotation JSON file&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="c1"># Merge multiline gene features in a separate file</span>
    <span class="n">interim_gff_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">args</span><span class="o">.</span><span class="n">in_gff_path</span><span class="si">}</span><span class="s2">_INTERIM_MERGE&quot;</span><span class="p">)</span>
    <span class="n">merger</span> <span class="o">=</span> <span class="n">GFFGeneMerger</span><span class="p">()</span>
    <span class="n">merged_genes</span> <span class="o">=</span> <span class="n">merger</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">in_gff_path</span><span class="p">,</span> <span class="n">interim_gff_path</span><span class="p">)</span>
    <span class="n">num_merged_genes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">merged_genes</span><span class="p">)</span>
    <span class="n">in_gff_path</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">in_gff_path</span>
    <span class="c1"># If there are split genes, decide to merge, or just die</span>
    <span class="k">if</span> <span class="n">num_merged_genes</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Report the list of merged genes in case something does not look right</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">num_merged_genes</span><span class="si">}</span><span class="s2"> genes merged:</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merged_genes</span><span class="p">))</span>
        <span class="c1"># Use the GFF with the merged genes for the next part</span>
        <span class="n">in_gff_path</span> <span class="o">=</span> <span class="n">interim_gff_path</span>

    <span class="c1"># Load GFF3 data and write a simpler version that follows our specifications as well as a</span>
    <span class="c1"># functional annotation JSON file</span>
    <span class="n">gff_data</span> <span class="o">=</span> <span class="n">GFFSimplifier</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">genome_data</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">make_missing_stable_ids</span><span class="p">)</span>
    <span class="n">gff_data</span><span class="o">.</span><span class="n">simpler_gff3</span><span class="p">(</span><span class="n">in_gff_path</span><span class="p">)</span>
    <span class="n">gff_data</span><span class="o">.</span><span class="n">records</span><span class="o">.</span><span class="n">to_gff</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_gff_path</span><span class="p">)</span>
    <span class="n">gff_data</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">to_json</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">out_func_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../../../../..", "features": [], "search": "../../../../../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../../../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
    
  </body>
</html>