
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../gff3/process_gff3/">
      
      
        <link rel="next" href="../manifest_maker/">
      
      
      <link rel="icon" href="../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>integrity - Ensembl-Genomio</title>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/stylesheets/main.4b4a2bd9.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#src.ensembl.io.genomio.integrity" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../.." title="Ensembl-Genomio" class="md-header__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-Genomio
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              integrity
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../.." title="Ensembl-Genomio" class="md-nav__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../images/logo.png" alt="logo">

    </a>
    Ensembl-Genomio
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    src
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            src
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [brc4]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [brc4]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1_1" id="__nav_4_1_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [runnable]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [runnable]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2" id="__nav_4_1_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [io]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2">
            <span class="md-nav__icon md-icon"></span>
            [io]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2_1" id="__nav_4_1_1_2_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [genomio]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            [genomio]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../assembly/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    assembly
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_1" id="__nav_4_1_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/get_assembly_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_assembly_data
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../db_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    db_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../events/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    events
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_3" id="__nav_4_1_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../events/format_events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_events
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../events_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../events_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_loader
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../fastaprep/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    fastaprep
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_6" id="__nav_4_1_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            fastaprep
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../fastaprep/process_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_fasta
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../genbank/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genbank
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_7" id="__nav_4_1_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/extract_from_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/get_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_genbank
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../genome_metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_8" id="__nav_4_1_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/compare_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/dump_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/dump_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../gff3/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    gff3
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_9" id="__nav_4_1_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/functional_annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    functional_annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/process_gff3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_gff3
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    integrity
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity" class="md-nav__link">
    integrity
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool" class="md-nav__link">
    IntegrityTool
  </a>
  
    <nav class="md-nav" aria-label="IntegrityTool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.add_errors" class="md-nav__link">
    add_errors()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_ids" class="md-nav__link">
    check_ids()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_integrity" class="md-nav__link">
    check_integrity()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_lengths" class="md-nav__link">
    check_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_seq_region_lengths" class="md-nav__link">
    check_seq_region_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.set_brc_mode" class="md-nav__link">
    set_brc_mode()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.set_ignore_final_stops" class="md-nav__link">
    set_ignore_final_stops()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.InvalidIntegrityError" class="md-nav__link">
    InvalidIntegrityError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest" class="md-nav__link">
    Manifest
  </a>
  
    <nav class="md-nav" aria-label="Manifest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_agp_seq_regions" class="md-nav__link">
    get_agp_seq_regions()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_circular" class="md-nav__link">
    get_circular()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_fasta_lengths" class="md-nav__link">
    get_fasta_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_functional_annotation" class="md-nav__link">
    get_functional_annotation()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_gff3" class="md-nav__link">
    get_gff3()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_lengths" class="md-nav__link">
    get_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_manifest" class="md-nav__link">
    get_manifest()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.has_lengths" class="md-nav__link">
    has_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.prepare_integrity_data" class="md-nav__link">
    prepare_integrity_data()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../manifest_maker/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_maker
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_13" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_13" id="__nav_4_1_1_2_1_13_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_13_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_13">
            <span class="md-nav__icon md-icon"></span>
            metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../metadata/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../metadata/prepare_seq_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_seq_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../metadata/update_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    update_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_14" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_14" id="__nav_4_1_1_2_1_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_14_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_14">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json_schema_validator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_validator
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seq_region_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_16" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../utils/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    utils
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_16" id="__nav_4_1_1_2_1_16_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_16_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_16">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../utils/archive_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    archive_utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../utils/json_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [example]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            [example]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../example/calculate/calculations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    calculations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../example/google_docs/google_docs_eg/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    google_docs
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../license/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    License
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity" class="md-nav__link">
    integrity
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool" class="md-nav__link">
    IntegrityTool
  </a>
  
    <nav class="md-nav" aria-label="IntegrityTool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.add_errors" class="md-nav__link">
    add_errors()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_ids" class="md-nav__link">
    check_ids()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_integrity" class="md-nav__link">
    check_integrity()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_lengths" class="md-nav__link">
    check_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.check_seq_region_lengths" class="md-nav__link">
    check_seq_region_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.set_brc_mode" class="md-nav__link">
    set_brc_mode()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.IntegrityTool.set_ignore_final_stops" class="md-nav__link">
    set_ignore_final_stops()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.InvalidIntegrityError" class="md-nav__link">
    InvalidIntegrityError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest" class="md-nav__link">
    Manifest
  </a>
  
    <nav class="md-nav" aria-label="Manifest">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_agp_seq_regions" class="md-nav__link">
    get_agp_seq_regions()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_circular" class="md-nav__link">
    get_circular()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_fasta_lengths" class="md-nav__link">
    get_fasta_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_functional_annotation" class="md-nav__link">
    get_functional_annotation()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_gff3" class="md-nav__link">
    get_gff3()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_lengths" class="md-nav__link">
    get_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.get_manifest" class="md-nav__link">
    get_manifest()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.has_lengths" class="md-nav__link">
    has_lengths()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.Manifest.prepare_integrity_data" class="md-nav__link">
    prepare_integrity_data()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.integrity.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>integrity</h1>

<div class="doc doc-object doc-module">



<a id="src.ensembl.io.genomio.integrity"></a>
  <div class="doc doc-contents first">
  
      <p>Compare the genomic data in a DNA fasta file, seq_region json, gene models GFF3 and peptide fasta
to ensure their contents are in sync.</p>

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.integrity.InputSchema" class="doc doc-heading">
          <code>InputSchema</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="argschema.ArgSchema">ArgSchema</span></code></p>

  
      <p>Input arguments expected by this script.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InputSchema</span><span class="p">(</span><span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchema</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Input arguments expected by this script.&quot;&quot;&quot;</span>

    <span class="c1"># Server parameters</span>
    <span class="n">manifest_file</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">InputFile</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Manifest file for the data to check&quot;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">brc_mode</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">Boolean</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;BRC mode&quot;</span><span class="p">})</span>
    <span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">Boolean</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Ignore final stop when calculating peptide length&quot;</span><span class="p">}</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.integrity.IntegrityTool" class="doc doc-heading">
          <code>IntegrityTool</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Check the integrity of sequence and annotation files in the genome</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">IntegrityTool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check the integrity of sequence and annotation files in the genome&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span> <span class="o">=</span> <span class="n">Manifest</span><span class="p">(</span><span class="n">manifest_file</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_brc_mode</span><span class="p">(</span><span class="n">brc_mode</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_ignore_final_stops</span><span class="p">(</span><span class="n">ignore_final_stops</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">add_errors</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Store the given errors in the list.&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">errors</span> <span class="o">+=</span> <span class="n">error</span>

    <span class="k">def</span> <span class="nf">check_integrity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Load files listed in the manifest.json and check the integrity.</span>
<span class="sd">        Check if the files are correct by verifying the MD5 hash.</span>
<span class="sd">        Check if translation, functional annotation and sequence region ids</span>
<span class="sd">        and lengths are consistent with the information in gff.</span>
<span class="sd">        Compare sequence length from fasta_dna file to seq_region.json metadata.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># Load the manifest integrity counts</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span>
        <span class="n">manifest</span><span class="o">.</span><span class="n">prepare_integrity_data</span><span class="p">()</span>

        <span class="n">genome</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">genome</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">seq_regions</span>

        <span class="n">dna</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">)</span>
        <span class="n">pep</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">)</span>
        <span class="n">seq_lengths</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>
        <span class="n">seq_circular</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_circular</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>

        <span class="n">agp_seqr</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;agp&quot;</span><span class="p">)</span>

        <span class="c1"># Then, run the checks</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_check_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

        <span class="c1"># Check gff3</span>
        <span class="k">if</span> <span class="n">manifest</span><span class="o">.</span><span class="n">has_lengths</span><span class="p">(</span><span class="s2">&quot;gff_genes&quot;</span><span class="p">):</span>
            <span class="n">gff_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_genes&quot;</span><span class="p">)</span>
            <span class="n">gff_seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_seq_regions&quot;</span><span class="p">)</span>
            <span class="n">gff_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_translations&quot;</span><span class="p">)</span>
            <span class="n">gff_all_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_all_translations&quot;</span><span class="p">)</span>
            <span class="n">gff_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_transposable_elements&quot;</span><span class="p">)</span>

            <span class="n">ann_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_genes&quot;</span><span class="p">)</span>
            <span class="n">ann_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_translations&quot;</span><span class="p">)</span>
            <span class="n">ann_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">)</span>

            <span class="c1"># Check fasta_pep.fa integrity</span>
            <span class="c1"># The sequence length and id retrieved from the fasta_pep file</span>
            <span class="c1"># and compared to the translated CDS id and length in the gff</span>
            <span class="c1"># We don&#39;t compare the peptide lengths because of seqedits</span>
            <span class="k">if</span> <span class="n">pep</span><span class="p">:</span>
                <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                    <span class="n">pep</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Fasta translations vs gff&quot;</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tr_errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># The pseudo CDSs are included in this check</span>
                    <span class="c1"># Pseudo CDSs are not translated, if the pseudo translation ids are not ignored</span>
                    <span class="c1"># in the gff it will give an error</span>
                    <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                        <span class="n">pep</span><span class="p">,</span>
                        <span class="n">gff_all_translations</span><span class="p">,</span>
                        <span class="s2">&quot;Fasta translations vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                        <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">tr_errors</span><span class="p">)</span>

            <span class="c1"># Check functional_annotation.json integrity</span>
            <span class="c1"># Gene ids, translated CDS ids and translated CDSs</span>
            <span class="c1"># including pseudogenes are compared to the gff</span>
            <span class="k">if</span> <span class="n">ann_genes</span><span class="p">:</span>
                <span class="n">errors</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span><span class="n">ann_genes</span><span class="p">,</span> <span class="n">gff_genes</span><span class="p">,</span> <span class="s2">&quot;Gene ids metadata vs gff&quot;</span><span class="p">)</span>
                <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                    <span class="n">ann_translations</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Translation ids metadata vs gff&quot;</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tr_errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                        <span class="n">ann_translations</span><span class="p">,</span>
                        <span class="n">gff_all_translations</span><span class="p">,</span>
                        <span class="s2">&quot;Translation ids metadata vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">tr_errors</span><span class="p">)</span>
                <span class="n">errors</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                    <span class="n">ann_transposable_elements</span><span class="p">,</span>
                    <span class="n">gff_transposable_elements</span><span class="p">,</span>
                    <span class="s2">&quot;TE ids metadata vs gff&quot;</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="c1"># Check the seq.json intregrity</span>
            <span class="c1"># Compare the length and id retrieved from seq.json to the gff</span>
            <span class="k">if</span> <span class="n">seq_regions</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span>
                    <span class="n">seq_lengths</span><span class="p">,</span> <span class="n">gff_seq_regions</span><span class="p">,</span> <span class="s2">&quot;Seq_regions metadata vs gff&quot;</span><span class="p">,</span> <span class="n">seq_circular</span>
                <span class="p">)</span>

        <span class="c1"># Check fasta dna and seq_region integrity</span>
        <span class="k">if</span> <span class="n">dna</span> <span class="ow">and</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">dna</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs dna&quot;</span><span class="p">)</span>

        <span class="c1"># Check agp and seq_region integrity</span>
        <span class="k">if</span> <span class="n">agp_seqr</span> <span class="ow">and</span> <span class="n">seq_lengths</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">agp_seqr</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs agps&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">errors</span><span class="p">:</span>
            <span class="n">errors_str</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">InvalidIntegrityError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Integrity test failed:</span><span class="se">\n</span><span class="si">{</span><span class="n">errors_str</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">set_brc_mode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set brc mode for this tool and the manifest.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>

    <span class="k">def</span> <span class="nf">set_ignore_final_stops</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Set ignore_final_stops (when calculating peptide length) for this tool and the manifest.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>

    <span class="k">def</span> <span class="nf">_check_genome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if the accession is correct in genome.json.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genome</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;assembly&quot;</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
                <span class="n">genome_ass</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;assembly&quot;</span><span class="p">]</span>
                <span class="k">if</span> <span class="s2">&quot;accession&quot;</span> <span class="ow">in</span> <span class="n">genome_ass</span><span class="p">:</span>
                    <span class="n">genome_acc</span> <span class="o">=</span> <span class="n">genome_ass</span><span class="p">[</span><span class="s2">&quot;accession&quot;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;GC[AF]_\d</span><span class="si">{9}</span><span class="s2">(\.\d+)?&quot;</span><span class="p">,</span> <span class="n">genome_acc</span><span class="p">):</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genome assembly accession is wrong: &#39;</span><span class="si">{</span><span class="n">genome_acc</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">check_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare the ids in list1 and list2.</span>

<span class="sd">        Args:</span>
<span class="sd">            list1: dict containing sequence ids retrieved from functional.json.</span>
<span class="sd">            list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">            name:  string</span>

<span class="sd">        Return:</span>
<span class="sd">            Error if the ids in functional.json and gff do not match.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">only1</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">only2</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">common</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list1</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
                <span class="n">common</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">only1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">common</span><span class="p">:</span>
                <span class="n">only2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>

        <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only1</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in first list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only2</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only2</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">errors</span>

    <span class="k">def</span> <span class="nf">check_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">allowed_len_diff</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check the difference in ids and length between list1 and list2.</span>
<span class="sd">            There are a few special cases here where we allow a certain asymmetry</span>
<span class="sd">            by changing the values of the arguments.</span>

<span class="sd">        Args:</span>
<span class="sd">            list1: dict containing length and id of the sequence from fasta files.</span>
<span class="sd">            list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">            name:  string</span>

<span class="sd">        allowed_len_diff : None to to not accept differences in length between list1 and list2.</span>
<span class="sd">            The value can be changed based on how much difference in sequence length we are wanting to accept.</span>

<span class="sd">        special_diff: set as False when no special length difference is expected between the lists.</span>
<span class="sd">                    This can be changed if we want to report common sequences with 1 BP difference.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Error if there is a difference in length or ids between the lists.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># check list diffferences, checks if abs(values diff) &lt; allowed_len_diff</span>

        <span class="n">set1</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list1</span><span class="p">)</span>
        <span class="n">set2</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list2</span><span class="p">)</span>
        <span class="n">list1_2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set1</span> <span class="o">-</span> <span class="n">set2</span><span class="p">)</span>
        <span class="n">list2_1</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set2</span> <span class="o">-</span> <span class="n">set1</span><span class="p">)</span>

        <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the first list only (i.e. </span><span class="si">{</span><span class="n">list1_2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the second list only (i.e. </span><span class="si">{</span><span class="n">list2_1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

        <span class="n">common_len</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">allowed_len_diff</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">common_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># check for the sequence length difference</span>
            <span class="n">diff_len_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">diff_len_special_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">:</span>
                <span class="n">dl12</span> <span class="o">=</span> <span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span> <span class="o">-</span> <span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">dl12</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">allowed_len_diff</span><span class="p">:</span>
                    <span class="n">common_len</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_list</span>
                    <span class="c1"># Special case: 1 AA /BP shorter,</span>
                    <span class="c1">#   so assuming the stop codon is not included in the CDS (when it should be)</span>
                    <span class="k">if</span> <span class="n">dl12</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">special_diff</span><span class="p">:</span>
                        <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_special_list</span>
                    <span class="n">_dlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">diff_len_special_list</span><span class="p">:</span>
                <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_special_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with one BP/AA length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_special_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="n">diff_len_list</span><span class="p">:</span>
                <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>
        <span class="k">if</span> <span class="n">common_len</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">common_len</span><span class="si">}</span><span class="s2"> common elements between lists for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">errors</span>

    <span class="k">def</span> <span class="nf">check_seq_region_lengths</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">seqrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="n">feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
        <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">circular</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check the integrity of seq_region.json file by comparing the length of the sequence</span>
<span class="sd">            to fasta files and the gff.</span>

<span class="sd">            Seq_region file is in json format containing the metadata of the sequence.</span>
<span class="sd">            It contains sequence id, length, location and the synonyms for the sequence name</span>
<span class="sd">            from different sources.</span>

<span class="sd">        Args:</span>
<span class="sd">            seqs: Sequence name and length retrieved from seq_region.json file.</span>
<span class="sd">            feats: Sequence name and length retrieved from the fasta and gff file.</span>
<span class="sd">            name: Name of the check to show in the logs.</span>
<span class="sd">            circular: Whether any sequence is circular.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Error if there are common sequences with difference in ids</span>
<span class="sd">            and if the sequences are not consistent in the files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compare_seqs</span><span class="p">(</span><span class="n">seqrs</span><span class="p">,</span> <span class="n">feats</span><span class="p">,</span> <span class="n">circular</span><span class="p">)</span>

        <span class="n">common</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span>
        <span class="n">diff_circular</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span>
        <span class="n">only_seqr</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_seqr&quot;</span><span class="p">]</span>
        <span class="n">only_feat</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_feat&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_circular</span><span class="p">:</span>
            <span class="n">example</span> <span class="o">=</span> <span class="n">diff_circular</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_circular</span><span class="p">)</span><span class="si">}</span><span class="s2"> differences for circular elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">example</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">diff</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with higher length in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">diff</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only_seqr</span><span class="p">:</span>
            <span class="c1"># Not an error!</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_seqr</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in seq_region list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_seqr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">only_feat</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_feat</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_feat</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_compare_seqs</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">seqrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="n">circular</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Give the intersection and other comparison between two groups of sequences.</span>

<span class="sd">        Args:</span>
<span class="sd">            seqs: Sequence name and length retrieved from seq_region.json file.</span>
<span class="sd">            feats: Sequence name and length retrieved from the fasta and gff file.</span>
<span class="sd">            circular: Whether any sequence is circular.</span>

<span class="sd">        Returns: Dict with 5 stats:</span>
<span class="sd">            common: Common elements.</span>
<span class="sd">            only_seqr: Elements only in the first one.</span>
<span class="sd">            only_feat: Elements only in the second one.</span>
<span class="sd">            diff: Elements that differ.</span>
<span class="sd">            diff_circular: Elements that differ in a circular sequence.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">comp</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;common&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;only_seqr&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;only_feat&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;diff&quot;</span><span class="p">:</span> <span class="p">[],</span>
            <span class="s2">&quot;diff_circular&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">seqrs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">feats</span><span class="p">:</span>
                <span class="c1"># Check that feature is within the seq_region length</span>
                <span class="k">if</span> <span class="n">feats</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">seqrs</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]:</span>
                    <span class="n">diff_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seq_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">seqrs</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">feats</span><span class="p">[</span><span class="n">seq_id</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="k">if</span> <span class="n">circular</span> <span class="ow">and</span> <span class="n">circular</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">seq_id</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
                        <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">diff_str</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">diff_str</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_id</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_seqr&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_id</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">seq_id</span> <span class="ow">in</span> <span class="n">feats</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">seq_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span>
                <span class="ow">and</span> <span class="n">seq_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span>
                <span class="ow">and</span> <span class="n">seq_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span>
            <span class="p">):</span>
                <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_feat&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_id</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">comp</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.IntegrityTool.add_errors" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Store the given errors in the list.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_errors</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Store the given errors in the list.&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">error</span> <span class="ow">in</span> <span class="n">args</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span> <span class="o">+=</span> <span class="n">error</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.IntegrityTool.check_ids" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">check_ids</span><span class="p">(</span><span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Compare the ids in list1 and list2.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>list1</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing sequence ids retrieved from functional.json.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>list2</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing length and id in the retrieved from the gff.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>name</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>string</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="return" open>
  <summary>Return</summary>
  <p>Error if the ids in functional.json and gff do not match.</p>
</details>
          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare the ids in list1 and list2.</span>

<span class="sd">    Args:</span>
<span class="sd">        list1: dict containing sequence ids retrieved from functional.json.</span>
<span class="sd">        list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">        name:  string</span>

<span class="sd">    Return:</span>
<span class="sd">        Error if the ids in functional.json and gff do not match.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">only1</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">only2</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">common</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list1</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
            <span class="n">common</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">only1</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">item_id</span> <span class="ow">in</span> <span class="n">list2</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">item_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">common</span><span class="p">:</span>
            <span class="n">only2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item_id</span><span class="p">)</span>

    <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only1</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only1</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in first list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only2</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only2</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">errors</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.IntegrityTool.check_integrity" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">check_integrity</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Load files listed in the manifest.json and check the integrity.
Check if the files are correct by verifying the MD5 hash.
Check if translation, functional annotation and sequence region ids
and lengths are consistent with the information in gff.
Compare sequence length from fasta_dna file to seq_region.json metadata.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_integrity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load files listed in the manifest.json and check the integrity.</span>
<span class="sd">    Check if the files are correct by verifying the MD5 hash.</span>
<span class="sd">    Check if translation, functional annotation and sequence region ids</span>
<span class="sd">    and lengths are consistent with the information in gff.</span>
<span class="sd">    Compare sequence length from fasta_dna file to seq_region.json metadata.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Load the manifest integrity counts</span>
    <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span>
    <span class="n">manifest</span><span class="o">.</span><span class="n">prepare_integrity_data</span><span class="p">()</span>

    <span class="n">genome</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">genome</span>
    <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">seq_regions</span>

    <span class="n">dna</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">)</span>
    <span class="n">pep</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">)</span>
    <span class="n">seq_lengths</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>
    <span class="n">seq_circular</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_circular</span><span class="p">(</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">)</span>

    <span class="n">agp_seqr</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;agp&quot;</span><span class="p">)</span>

    <span class="c1"># Then, run the checks</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_check_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>

    <span class="c1"># Check gff3</span>
    <span class="k">if</span> <span class="n">manifest</span><span class="o">.</span><span class="n">has_lengths</span><span class="p">(</span><span class="s2">&quot;gff_genes&quot;</span><span class="p">):</span>
        <span class="n">gff_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_genes&quot;</span><span class="p">)</span>
        <span class="n">gff_seq_regions</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_seq_regions&quot;</span><span class="p">)</span>
        <span class="n">gff_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_translations&quot;</span><span class="p">)</span>
        <span class="n">gff_all_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_all_translations&quot;</span><span class="p">)</span>
        <span class="n">gff_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;gff_transposable_elements&quot;</span><span class="p">)</span>

        <span class="n">ann_genes</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_genes&quot;</span><span class="p">)</span>
        <span class="n">ann_translations</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_translations&quot;</span><span class="p">)</span>
        <span class="n">ann_transposable_elements</span> <span class="o">=</span> <span class="n">manifest</span><span class="o">.</span><span class="n">get_lengths</span><span class="p">(</span><span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">)</span>

        <span class="c1"># Check fasta_pep.fa integrity</span>
        <span class="c1"># The sequence length and id retrieved from the fasta_pep file</span>
        <span class="c1"># and compared to the translated CDS id and length in the gff</span>
        <span class="c1"># We don&#39;t compare the peptide lengths because of seqedits</span>
        <span class="k">if</span> <span class="n">pep</span><span class="p">:</span>
            <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                <span class="n">pep</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Fasta translations vs gff&quot;</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tr_errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># The pseudo CDSs are included in this check</span>
                <span class="c1"># Pseudo CDSs are not translated, if the pseudo translation ids are not ignored</span>
                <span class="c1"># in the gff it will give an error</span>
                <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_lengths</span><span class="p">(</span>
                    <span class="n">pep</span><span class="p">,</span>
                    <span class="n">gff_all_translations</span><span class="p">,</span>
                    <span class="s2">&quot;Fasta translations vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                    <span class="n">special_diff</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">tr_errors</span><span class="p">)</span>

        <span class="c1"># Check functional_annotation.json integrity</span>
        <span class="c1"># Gene ids, translated CDS ids and translated CDSs</span>
        <span class="c1"># including pseudogenes are compared to the gff</span>
        <span class="k">if</span> <span class="n">ann_genes</span><span class="p">:</span>
            <span class="n">errors</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span><span class="n">ann_genes</span><span class="p">,</span> <span class="n">gff_genes</span><span class="p">,</span> <span class="s2">&quot;Gene ids metadata vs gff&quot;</span><span class="p">)</span>
            <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                <span class="n">ann_translations</span><span class="p">,</span> <span class="n">gff_translations</span><span class="p">,</span> <span class="s2">&quot;Translation ids metadata vs gff&quot;</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tr_errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">tr_errors</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                    <span class="n">ann_translations</span><span class="p">,</span>
                    <span class="n">gff_all_translations</span><span class="p">,</span>
                    <span class="s2">&quot;Translation ids metadata vs gff (include pseudo CDS)&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="o">*</span><span class="n">tr_errors</span><span class="p">)</span>
            <span class="n">errors</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ids</span><span class="p">(</span>
                <span class="n">ann_transposable_elements</span><span class="p">,</span>
                <span class="n">gff_transposable_elements</span><span class="p">,</span>
                <span class="s2">&quot;TE ids metadata vs gff&quot;</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="c1"># Check the seq.json intregrity</span>
        <span class="c1"># Compare the length and id retrieved from seq.json to the gff</span>
        <span class="k">if</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span>
                <span class="n">seq_lengths</span><span class="p">,</span> <span class="n">gff_seq_regions</span><span class="p">,</span> <span class="s2">&quot;Seq_regions metadata vs gff&quot;</span><span class="p">,</span> <span class="n">seq_circular</span>
            <span class="p">)</span>

    <span class="c1"># Check fasta dna and seq_region integrity</span>
    <span class="k">if</span> <span class="n">dna</span> <span class="ow">and</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">dna</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs dna&quot;</span><span class="p">)</span>

    <span class="c1"># Check agp and seq_region integrity</span>
    <span class="k">if</span> <span class="n">agp_seqr</span> <span class="ow">and</span> <span class="n">seq_lengths</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seq_lengths</span><span class="p">,</span> <span class="n">agp_seqr</span><span class="p">,</span> <span class="s2">&quot;seq_regions json vs agps&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">errors</span><span class="p">:</span>
        <span class="n">errors_str</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">InvalidIntegrityError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Integrity test failed:</span><span class="se">\n</span><span class="si">{</span><span class="n">errors_str</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.IntegrityTool.check_lengths" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">check_lengths</span><span class="p">(</span><span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">allowed_len_diff</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Check the difference in ids and length between list1 and list2.
    There are a few special cases here where we allow a certain asymmetry
    by changing the values of the arguments.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>list1</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing length and id of the sequence from fasta files.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>list2</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing length and id in the retrieved from the gff.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>name</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>string</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="allowed_len_diff-" open>
  <summary>None to to not accept differences in length between list1 and list2.</summary>
  <p>The value can be changed based on how much difference in sequence length we are wanting to accept.</p>
</details>
<details class="special_diff" open>
  <summary>set as False when no special length difference is expected between the lists.</summary>
  <p>This can be changed if we want to report common sequences with 1 BP difference.</p>
</details>


  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Error if there is a difference in length or ids between the lists.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">list1</span><span class="p">,</span> <span class="n">list2</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">allowed_len_diff</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">special_diff</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check the difference in ids and length between list1 and list2.</span>
<span class="sd">        There are a few special cases here where we allow a certain asymmetry</span>
<span class="sd">        by changing the values of the arguments.</span>

<span class="sd">    Args:</span>
<span class="sd">        list1: dict containing length and id of the sequence from fasta files.</span>
<span class="sd">        list2: dict containing length and id in the retrieved from the gff.</span>
<span class="sd">        name:  string</span>

<span class="sd">    allowed_len_diff : None to to not accept differences in length between list1 and list2.</span>
<span class="sd">        The value can be changed based on how much difference in sequence length we are wanting to accept.</span>

<span class="sd">    special_diff: set as False when no special length difference is expected between the lists.</span>
<span class="sd">                This can be changed if we want to report common sequences with 1 BP difference.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Error if there is a difference in length or ids between the lists.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># check list diffferences, checks if abs(values diff) &lt; allowed_len_diff</span>

    <span class="n">set1</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list1</span><span class="p">)</span>
    <span class="n">set2</span> <span class="o">=</span> <span class="nb">frozenset</span><span class="p">(</span><span class="n">list2</span><span class="p">)</span>
    <span class="n">list1_2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set1</span> <span class="o">-</span> <span class="n">set2</span><span class="p">)</span>
    <span class="n">list2_1</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">set2</span> <span class="o">-</span> <span class="n">set1</span><span class="p">)</span>

    <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list1_2</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the first list only (i.e. </span><span class="si">{</span><span class="n">list1_2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">list2_1</span><span class="p">)</span><span class="si">}</span><span class="s2"> from the second list only (i.e. </span><span class="si">{</span><span class="n">list2_1</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>

    <span class="n">common_len</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">allowed_len_diff</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">common_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># check for the sequence length difference</span>
        <span class="n">diff_len_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">diff_len_special_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">set1</span> <span class="o">&amp;</span> <span class="n">set2</span><span class="p">:</span>
            <span class="n">dl12</span> <span class="o">=</span> <span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span> <span class="o">-</span> <span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="n">dl12</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">allowed_len_diff</span><span class="p">:</span>
                <span class="n">common_len</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_list</span>
                <span class="c1"># Special case: 1 AA /BP shorter,</span>
                <span class="c1">#   so assuming the stop codon is not included in the CDS (when it should be)</span>
                <span class="k">if</span> <span class="n">dl12</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">special_diff</span><span class="p">:</span>
                    <span class="n">_dlist</span> <span class="o">=</span> <span class="n">diff_len_special_list</span>
                <span class="n">_dlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">list1</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">list2</span><span class="p">[</span><span class="n">e</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_len_special_list</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_special_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with one BP/AA length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_special_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">diff_len_list</span><span class="p">:</span>
            <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_len_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with length diff for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;(e.g. </span><span class="si">{</span><span class="n">diff_len_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span>
                <span class="p">)</span>
            <span class="p">)</span>
    <span class="k">if</span> <span class="n">common_len</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">common_len</span><span class="si">}</span><span class="s2"> common elements between lists for </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">errors</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.IntegrityTool.check_seq_region_lengths" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">check_seq_region_lengths</span><span class="p">(</span><span class="n">seqrs</span><span class="p">,</span> <span class="n">feats</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Check the integrity of seq_region.json file by comparing the length of the sequence
    to fasta files and the gff.</p>
<pre><code>Seq_region file is in json format containing the metadata of the sequence.
It contains sequence id, length, location and the synonyms for the sequence name
from different sources.
</code></pre>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seqs</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence name and length retrieved from seq_region.json file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feats</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Sequence name and length retrieved from the fasta and gff file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>name</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Name of the check to show in the logs.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>circular</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Whether any sequence is circular.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>None</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Error if there are common sequences with difference in ids</p>
            </div>
          </td>
        </tr>
        <tr>
          <td>
                <code>None</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>and if the sequences are not consistent in the files.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_seq_region_lengths</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">seqrs</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">circular</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check the integrity of seq_region.json file by comparing the length of the sequence</span>
<span class="sd">        to fasta files and the gff.</span>

<span class="sd">        Seq_region file is in json format containing the metadata of the sequence.</span>
<span class="sd">        It contains sequence id, length, location and the synonyms for the sequence name</span>
<span class="sd">        from different sources.</span>

<span class="sd">    Args:</span>
<span class="sd">        seqs: Sequence name and length retrieved from seq_region.json file.</span>
<span class="sd">        feats: Sequence name and length retrieved from the fasta and gff file.</span>
<span class="sd">        name: Name of the check to show in the logs.</span>
<span class="sd">        circular: Whether any sequence is circular.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Error if there are common sequences with difference in ids</span>
<span class="sd">        and if the sequences are not consistent in the files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_compare_seqs</span><span class="p">(</span><span class="n">seqrs</span><span class="p">,</span> <span class="n">feats</span><span class="p">,</span> <span class="n">circular</span><span class="p">)</span>

    <span class="n">common</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;common&quot;</span><span class="p">]</span>
    <span class="n">diff</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff&quot;</span><span class="p">]</span>
    <span class="n">diff_circular</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;diff_circular&quot;</span><span class="p">]</span>
    <span class="n">only_seqr</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_seqr&quot;</span><span class="p">]</span>
    <span class="n">only_feat</span> <span class="o">=</span> <span class="n">comp</span><span class="p">[</span><span class="s2">&quot;only_feat&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">common</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">common</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">diff_circular</span><span class="p">:</span>
        <span class="n">example</span> <span class="o">=</span> <span class="n">diff_circular</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff_circular</span><span class="p">)</span><span class="si">}</span><span class="s2"> differences for circular elements in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">example</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">diff</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span><span class="si">}</span><span class="s2"> common elements with higher length in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (e.g. </span><span class="si">{</span><span class="n">diff</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only_seqr</span><span class="p">:</span>
        <span class="c1"># Not an error!</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_seqr</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in seq_region list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_seqr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">only_feat</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_errors</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">only_feat</span><span class="p">)</span><span class="si">}</span><span class="s2"> only in second list in </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> (first: </span><span class="si">{</span><span class="n">only_feat</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.IntegrityTool.set_brc_mode" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">set_brc_mode</span><span class="p">(</span><span class="n">brc_mode</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Set brc mode for this tool and the manifest.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_brc_mode</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">brc_mode</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set brc mode for this tool and the manifest.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="n">brc_mode</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.IntegrityTool.set_ignore_final_stops" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">set_ignore_final_stops</span><span class="p">(</span><span class="n">ignore_final_stops</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Set ignore_final_stops (when calculating peptide length) for this tool and the manifest.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_ignore_final_stops</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set ignore_final_stops (when calculating peptide length) for this tool and the manifest.&quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="n">ignore_final_stops</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.integrity.InvalidIntegrityError" class="doc doc-heading">
          <code>InvalidIntegrityError</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>When a file integrity check fails</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">41</span>
<span class="normal">42</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InvalidIntegrityError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;When a file integrity check fails&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.integrity.Manifest" class="doc doc-heading">
          <code>Manifest</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Representation of the manifest and its files.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Manifest</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Representation of the manifest and its files.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_manifest</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_regions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;dna_sequences&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;peptide_sequences&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;seq_region_levels&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;annotations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;agp&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_genes&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_translations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;ann_genes&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;ann_translations&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">:</span> <span class="p">{},</span>
            <span class="s2">&quot;seq_regions&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;seq_regions&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">has_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if a given name has lengths records.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">get_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their length from a given file name.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No length available for key </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_circular</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their is_circular flag from a given file name.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No length available for key </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_add_error</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">error</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">error</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Load the content of a manifest file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict: Content of the manifest file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">manifest_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">manifest_fh</span><span class="p">:</span>
            <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">manifest_fh</span><span class="p">)</span>

            <span class="c1"># Use dir name from the manifest</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
                <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                    <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                    <span class="c1"># check if the md5sum is correct</span>
                    <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                    <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                            <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                            <span class="c1"># check if the md5sum is correct</span>
                            <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                            <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                            <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
            <span class="k">return</span> <span class="n">manifest</span>

    <span class="k">def</span> <span class="nf">_check_md5sum</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Verify the integrity of the files in manifest.json.</span>

<span class="sd">        An MD5 hash is generated using the path provided which is then compared to the hash in manifest.json.</span>
<span class="sd">        Errors are stored in ``self.errors``.</span>

<span class="sd">        Args:</span>
<span class="sd">            file_path: Path to a genome file.</span>
<span class="sd">            md5sum: MD5 hash for the files.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">with</span> <span class="n">file_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">bytes_obj</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
            <span class="n">readable_hash</span> <span class="o">=</span> <span class="n">hashlib</span><span class="o">.</span><span class="n">md5</span><span class="p">(</span><span class="n">bytes_obj</span><span class="p">)</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">readable_hash</span> <span class="o">!=</span> <span class="n">md5sum</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">InvalidIntegrityError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid md5 checksum for </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">prepare_integrity_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Read all the files and keep a record (IDs and their lengths)</span>
<span class="sd">        for each cases to be compared later.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># First, get the Data</span>
        <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a gff&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;fasta_dna&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a fasta dna&quot;</span><span class="p">)</span>
            <span class="c1"># Verify if the length and id for the sequence is unique</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_dna&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;fasta_pep&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a fasta pep&quot;</span><span class="p">)</span>
            <span class="c1"># Verify if the length and id for the sequence is unique</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_pep&quot;</span><span class="p">],</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a seq_regions&quot;</span><span class="p">)</span>
            <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>
            <span class="n">seqr_seqlevel</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">seq_lengths</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">seq_circular</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="c1"># Store the length as int</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
                <span class="n">seq_lengths</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
                <span class="n">seq_circular</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;circular&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;contig&quot;</span><span class="p">:</span>
                    <span class="n">seqr_seqlevel</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_lengths</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_circular</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">seq_regions</span> <span class="o">=</span> <span class="n">seq_regions</span>
        <span class="k">if</span> <span class="s2">&quot;functional_annotation&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a func_anns&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;functional_annotation&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;agp&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got agp files&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="s2">&quot;genome&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a genome&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]))</span>

    <span class="k">def</span> <span class="nf">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if the fasta files have the correct ids and no stop codon.</span>

<span class="sd">        Args:</span>
<span class="sd">            fasta_path: Path to fasta_dna and fasta_pep files.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Error if any empty ids, non-unique ids or stop codons are found in the fasta files.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">non_unique</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">non_unique_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">empty_id_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">contains_stop_codon</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
            <span class="c1"># Flag empty ids</span>
            <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">empty_id_count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Flag redundant ids</span>
                <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                    <span class="n">non_unique</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="n">non_unique_count</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1"># Store sequence id and length</span>
                <span class="n">data</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
                <span class="n">stops</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">stops</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">stops</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">ignore_final_stops</span><span class="p">:</span>
                        <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">empty_id_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">empty_id_count</span><span class="si">}</span><span class="s2"> sequences with empty ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">non_unique_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">non_unique_count</span><span class="si">}</span><span class="s2"> non unique sequence ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">contains_stop_codon</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">contains_stop_codon</span><span class="si">}</span><span class="s2"> sequences with stop codons in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">data</span>

    <span class="k">def</span> <span class="nf">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">json_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Load the functional annotation file to retrieve the gene_id and translation id.</span>
<span class="sd">            A functional annotation file contains information about a gene.</span>
<span class="sd">            The functional annotation file is stored in a json format containing</span>
<span class="sd">            the description, id and object type (eg: &quot;gene&quot;, &quot;transcript&quot;, &quot;translation&quot;).</span>

<span class="sd">        Args:</span>
<span class="sd">            json_path: Path to functional_annotation.json.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict with gene and translation ids.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Load the json file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

        <span class="c1"># Get gene ids and translation ids</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">translations</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">transposons</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span>
                <span class="n">genes</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;translation&quot;</span><span class="p">:</span>
                <span class="n">translations</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
                <span class="n">transposons</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;ann_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
            <span class="s2">&quot;ann_translations&quot;</span><span class="p">:</span> <span class="n">translations</span><span class="p">,</span>
            <span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">:</span> <span class="n">transposons</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;A GFF parser is used to retrieve information in the GFF file such as</span>
<span class="sd">           gene and CDS ids and their corresponding lengths.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path: Path to gff3 file.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">seqs</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">genes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">all_peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">tes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
            <span class="n">gff</span> <span class="o">=</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">gff</span><span class="p">:</span>
                <span class="n">seqs</span><span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

                <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="n">feat_length</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                    <span class="c1"># Store gene id and length</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;ncRNA_gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">]:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_retrieve_gff_gene_lengths</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peps</span><span class="p">,</span> <span class="n">all_peps</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
                        <span class="n">tes</span><span class="p">[</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_length</span>

        <span class="n">stats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">:</span> <span class="n">seqs</span><span class="p">,</span>
            <span class="s2">&quot;gff3_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
            <span class="s2">&quot;gff3_translations&quot;</span><span class="p">:</span> <span class="n">peps</span><span class="p">,</span>
            <span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">:</span> <span class="n">all_peps</span><span class="p">,</span>
            <span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">:</span> <span class="n">tes</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>

    <span class="k">def</span> <span class="nf">_retrieve_gff_gene_lengths</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">genes</span><span class="p">:</span> <span class="n">Lengths</span><span class="p">,</span> <span class="n">peps</span><span class="p">:</span> <span class="n">Lengths</span><span class="p">,</span> <span class="n">all_peps</span><span class="p">:</span> <span class="n">Lengths</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Record genes and peptides lengths from a feature.</span>

<span class="sd">        Args:</span>
<span class="sd">            feat : Gene feature to check.</span>
<span class="sd">            genes: Record of genes lengths to update.</span>
<span class="sd">            peps: Record of peptides lengths to update.</span>
<span class="sd">            all_peps: Record of all peptides lengths to update (include pseudogenes).</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gene_id</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">id</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span><span class="p">:</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">gene_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;gene:&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">genes</span><span class="p">[</span><span class="n">gene_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="c1"># Get CDS id and length</span>
        <span class="k">for</span> <span class="n">feat2</span> <span class="ow">in</span> <span class="n">feat</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat2</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogenic_transcript&quot;</span><span class="p">):</span>
                <span class="n">length</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">for</span> <span class="n">feat3</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">feat3</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                        <span class="n">pep_id</span> <span class="o">=</span> <span class="n">feat3</span><span class="o">.</span><span class="n">id</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">brc_mode</span><span class="p">:</span>
                            <span class="n">pep_id</span> <span class="o">=</span> <span class="n">pep_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;CDS:&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">pep_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">length</span><span class="p">:</span>
                            <span class="n">length</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
                        <span class="n">length</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">+=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat3</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat3</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">pep_id</span><span class="p">,</span> <span class="n">pep_length</span> <span class="ow">in</span> <span class="n">length</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="c1"># Store length for translations, add pseudo translations separately</span>
                    <span class="n">pep_length</span> <span class="o">=</span> <span class="n">floor</span><span class="p">(</span><span class="n">pep_length</span> <span class="o">/</span> <span class="mi">3</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
                        <span class="n">peps</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pep_length</span>
                    <span class="n">all_peps</span><span class="p">[</span><span class="n">pep_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pep_length</span>

    <span class="k">def</span> <span class="nf">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">agp_dict</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;AGP files describe the assembly of larger sequence objects using smaller objects.</span>
<span class="sd">            Eg: describes the assembly of scaffolds from contigs.</span>

<span class="sd">        Args:</span>
<span class="sd">            agp_dict: dict containing the information about the sequence.</span>

<span class="sd">        Note:</span>
<span class="sd">            AGP file is only used in the older builds, not used for current processing.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">seqr</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">agp</span> <span class="ow">in</span> <span class="n">agp_dict</span><span class="p">:</span>
            <span class="n">agp_path</span> <span class="o">=</span> <span class="n">agp_dict</span><span class="p">[</span><span class="n">agp</span><span class="p">]</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">agp_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">agph</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">agph</span><span class="p">:</span>
                    <span class="p">(</span>
                        <span class="n">asm_id</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_start</span>
                        <span class="n">asm_end</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_part</span>
                        <span class="n">typ</span><span class="p">,</span>
                        <span class="n">cmp_id</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_start</span>
                        <span class="n">cmp_end</span><span class="p">,</span>
                        <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_strand</span>
                    <span class="p">)</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="c1"># Ignore WGS contig</span>
                    <span class="k">if</span> <span class="n">typ</span> <span class="o">!=</span> <span class="s2">&quot;W&quot;</span><span class="p">:</span>
                        <span class="k">continue</span>

                    <span class="c1"># Assembled seq length</span>
                    <span class="k">if</span> <span class="n">asm_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">):</span>
                        <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">)</span>

                    <span class="c1"># Composite seq length</span>
                    <span class="k">if</span> <span class="n">cmp_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">):</span>
                        <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">seqr</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.get_agp_seq_regions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_agp_seq_regions</span><span class="p">(</span><span class="n">agp_dict</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>AGP files describe the assembly of larger sequence objects using smaller objects.
    Eg: describes the assembly of scaffolds from contigs.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>agp_dict</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict containing the information about the sequence.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

<details class="note" open>
  <summary>Note</summary>
  <p>AGP file is only used in the older builds, not used for current processing.</p>
</details>
          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">agp_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;AGP files describe the assembly of larger sequence objects using smaller objects.</span>
<span class="sd">        Eg: describes the assembly of scaffolds from contigs.</span>

<span class="sd">    Args:</span>
<span class="sd">        agp_dict: dict containing the information about the sequence.</span>

<span class="sd">    Note:</span>
<span class="sd">        AGP file is only used in the older builds, not used for current processing.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">seqr</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">agp</span> <span class="ow">in</span> <span class="n">agp_dict</span><span class="p">:</span>
        <span class="n">agp_path</span> <span class="o">=</span> <span class="n">agp_dict</span><span class="p">[</span><span class="n">agp</span><span class="p">]</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">agp_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">agph</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">agph</span><span class="p">:</span>
                <span class="p">(</span>
                    <span class="n">asm_id</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_start</span>
                    <span class="n">asm_end</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># asm_part</span>
                    <span class="n">typ</span><span class="p">,</span>
                    <span class="n">cmp_id</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_start</span>
                    <span class="n">cmp_end</span><span class="p">,</span>
                    <span class="n">_</span><span class="p">,</span>  <span class="c1"># cmp_strand</span>
                <span class="p">)</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="c1"># Ignore WGS contig</span>
                <span class="k">if</span> <span class="n">typ</span> <span class="o">!=</span> <span class="s2">&quot;W&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c1"># Assembled seq length</span>
                <span class="k">if</span> <span class="n">asm_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">):</span>
                    <span class="n">seqr</span><span class="p">[</span><span class="n">asm_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">asm_end</span><span class="p">)</span>

                <span class="c1"># Composite seq length</span>
                <span class="k">if</span> <span class="n">cmp_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">):</span>
                    <span class="n">seqr</span><span class="p">[</span><span class="n">cmp_id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cmp_end</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">seqr</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.get_circular" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_circular</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a dict associating IDs with their is_circular flag from a given file name.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_circular</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their is_circular flag from a given file name.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No length available for key </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.get_fasta_lengths" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_fasta_lengths</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Check if the fasta files have the correct ids and no stop codon.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>fasta_path</code></td>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to fasta_dna and fasta_pep files.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Error if any empty ids, non-unique ids or stop codons are found in the fasta files.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_path</span><span class="p">,</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if the fasta files have the correct ids and no stop codon.</span>

<span class="sd">    Args:</span>
<span class="sd">        fasta_path: Path to fasta_dna and fasta_pep files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Error if any empty ids, non-unique ids or stop codons are found in the fasta files.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">non_unique</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">non_unique_count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">empty_id_count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">contains_stop_codon</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
        <span class="c1"># Flag empty ids</span>
        <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">empty_id_count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Flag redundant ids</span>
            <span class="k">if</span> <span class="n">rec</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                <span class="n">non_unique</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="n">non_unique_count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c1"># Store sequence id and length</span>
            <span class="n">data</span><span class="p">[</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">stops</span> <span class="o">=</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">stops</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">stops</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">ignore_final_stops</span><span class="p">:</span>
                    <span class="n">contains_stop_codon</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">if</span> <span class="n">empty_id_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">empty_id_count</span><span class="si">}</span><span class="s2"> sequences with empty ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">non_unique_count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">non_unique_count</span><span class="si">}</span><span class="s2"> non unique sequence ids in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">contains_stop_codon</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_add_error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">contains_stop_codon</span><span class="si">}</span><span class="s2"> sequences with stop codons in </span><span class="si">{</span><span class="n">fasta_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.get_functional_annotation" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_functional_annotation</span><span class="p">(</span><span class="n">json_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Load the functional annotation file to retrieve the gene_id and translation id.
    A functional annotation file contains information about a gene.
    The functional annotation file is stored in a json format containing
    the description, id and object type (eg: "gene", "transcript", "translation").</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>json_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to functional_annotation.json.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>None</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>dict with gene and translation ids.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">json_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load the functional annotation file to retrieve the gene_id and translation id.</span>
<span class="sd">        A functional annotation file contains information about a gene.</span>
<span class="sd">        The functional annotation file is stored in a json format containing</span>
<span class="sd">        the description, id and object type (eg: &quot;gene&quot;, &quot;transcript&quot;, &quot;translation&quot;).</span>

<span class="sd">    Args:</span>
<span class="sd">        json_path: Path to functional_annotation.json.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict with gene and translation ids.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Load the json file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">json_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

    <span class="c1"># Get gene ids and translation ids</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">translations</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">transposons</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span>
            <span class="n">genes</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;translation&quot;</span><span class="p">:</span>
            <span class="n">translations</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
            <span class="n">transposons</span><span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;ann_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
        <span class="s2">&quot;ann_translations&quot;</span><span class="p">:</span> <span class="n">translations</span><span class="p">,</span>
        <span class="s2">&quot;ann_transposable_elements&quot;</span><span class="p">:</span> <span class="n">transposons</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.get_gff3" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_gff3</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>A GFF parser is used to retrieve information in the GFF file such as
   gene and CDS ids and their corresponding lengths.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gff3_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to gff3 file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A GFF parser is used to retrieve information in the GFF file such as</span>
<span class="sd">       gene and CDS ids and their corresponding lengths.</span>

<span class="sd">    Args:</span>
<span class="sd">        gff3_path: Path to gff3 file.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">seqs</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">genes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">all_peps</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">tes</span><span class="p">:</span> <span class="n">Lengths</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
        <span class="n">gff</span> <span class="o">=</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">gff</span><span class="p">:</span>
            <span class="n">seqs</span><span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="n">feat_length</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span> <span class="o">-</span> <span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="c1"># Store gene id and length</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;ncRNA_gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">]:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_retrieve_gff_gene_lengths</span><span class="p">(</span><span class="n">feat</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">peps</span><span class="p">,</span> <span class="n">all_peps</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transposable_element&quot;</span><span class="p">:</span>
                    <span class="n">tes</span><span class="p">[</span><span class="n">feat</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_length</span>

    <span class="n">stats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Lengths</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;gff3_seq_regions&quot;</span><span class="p">:</span> <span class="n">seqs</span><span class="p">,</span>
        <span class="s2">&quot;gff3_genes&quot;</span><span class="p">:</span> <span class="n">genes</span><span class="p">,</span>
        <span class="s2">&quot;gff3_translations&quot;</span><span class="p">:</span> <span class="n">peps</span><span class="p">,</span>
        <span class="s2">&quot;gff3_all_translations&quot;</span><span class="p">:</span> <span class="n">all_peps</span><span class="p">,</span>
        <span class="s2">&quot;gff3_transposable_elements&quot;</span><span class="p">:</span> <span class="n">tes</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">,</span> <span class="o">**</span><span class="n">stats</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.get_lengths" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_lengths</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Returns a dict associating IDs with their length from a given file name.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns a dict associating IDs with their length from a given file name.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No length available for key </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.get_manifest" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_manifest</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Load the content of a manifest file.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>Dict</code></td>          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Content of the manifest file.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load the content of a manifest file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict: Content of the manifest file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">manifest_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">manifest_path</span><span class="p">)</span>
    <span class="k">with</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">manifest_fh</span><span class="p">:</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">manifest_fh</span><span class="p">)</span>

        <span class="c1"># Use dir name from the manifest</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                <span class="c1"># check if the md5sum is correct</span>
                <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                        <span class="n">file_path</span> <span class="o">=</span> <span class="n">manifest_path</span><span class="o">.</span><span class="n">parent</span> <span class="o">/</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                        <span class="c1"># check if the md5sum is correct</span>
                        <span class="n">md5sum</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;md5sum&quot;</span><span class="p">]</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_check_md5sum</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">md5sum</span><span class="p">)</span>

                        <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_path</span>
        <span class="k">return</span> <span class="n">manifest</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.has_lengths" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">has_lengths</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Check if a given name has lengths records.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">has_lengths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if a given name has lengths records.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.integrity.Manifest.prepare_integrity_data" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">prepare_integrity_data</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Read all the files and keep a record (IDs and their lengths)
for each cases to be compared later.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">prepare_integrity_data</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read all the files and keep a record (IDs and their lengths)</span>
<span class="sd">    for each cases to be compared later.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First, get the Data</span>
    <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a gff&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;fasta_dna&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a fasta dna&quot;</span><span class="p">)</span>
        <span class="c1"># Verify if the length and id for the sequence is unique</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;dna_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_dna&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;fasta_pep&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a fasta pep&quot;</span><span class="p">)</span>
        <span class="c1"># Verify if the length and id for the sequence is unique</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;peptide_sequences&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_fasta_lengths</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;fasta_pep&quot;</span><span class="p">],</span> <span class="n">ignore_final_stops</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ignore_final_stops</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a seq_regions&quot;</span><span class="p">)</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>
        <span class="n">seqr_seqlevel</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">seq_lengths</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">seq_circular</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="c1"># Store the length as int</span>
        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="n">seq_lengths</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
            <span class="n">seq_circular</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;circular&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;contig&quot;</span><span class="p">:</span>
                <span class="n">seqr_seqlevel</span><span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_lengths</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">circular</span><span class="p">[</span><span class="s2">&quot;seq_regions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_circular</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_regions</span> <span class="o">=</span> <span class="n">seq_regions</span>
    <span class="k">if</span> <span class="s2">&quot;functional_annotation&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a func_anns&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">get_functional_annotation</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;functional_annotation&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;agp&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got agp files&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_agp_seq_regions</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;agp&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="s2">&quot;genome&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Got a genome&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">get_json</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest_files</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]))</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>



<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.integrity.main" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Main entrypoint.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/integrity.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main entrypoint.&quot;&quot;&quot;</span>
    <span class="n">mod</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchemaParser</span><span class="p">(</span><span class="n">schema_type</span><span class="o">=</span><span class="n">InputSchema</span><span class="p">)</span>

    <span class="c1"># Start</span>
    <span class="n">inspector</span> <span class="o">=</span> <span class="n">IntegrityTool</span><span class="p">(</span><span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;manifest_file&quot;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;brc_mode&quot;</span><span class="p">):</span>
        <span class="n">inspector</span><span class="o">.</span><span class="n">set_brc_mode</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;ignore_final_stops&quot;</span><span class="p">):</span>
        <span class="n">inspector</span><span class="o">.</span><span class="n">set_ignore_final_stops</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">inspector</span><span class="o">.</span><span class="n">check_integrity</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../../../..", "features": [], "search": "../../../../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
    
  </body>
</html>