
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../manifest_maker/">
      
      
        <link rel="next" href="../metadata/">
      
      
      <link rel="icon" href="../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>manifest_stats - Ensembl-Genomio</title>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/stylesheets/main.4b4a2bd9.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#src.ensembl.io.genomio.manifest_stats" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../.." title="Ensembl-Genomio" class="md-header__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-Genomio
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              manifest_stats
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../.." title="Ensembl-Genomio" class="md-nav__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../images/logo.png" alt="logo">

    </a>
    Ensembl-Genomio
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    src
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            src
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [brc4]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [brc4]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1_1" id="__nav_4_1_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [runnable]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [runnable]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2" id="__nav_4_1_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [io]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2">
            <span class="md-nav__icon md-icon"></span>
            [io]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2_1" id="__nav_4_1_1_2_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [genomio]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            [genomio]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../assembly/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    assembly
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_1" id="__nav_4_1_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/get_assembly_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_assembly_data
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../db_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    db_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../events/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    events
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_3" id="__nav_4_1_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../events/format_events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_events
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../events_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../events_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_loader
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../fastaprep/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    fastaprep
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_6" id="__nav_4_1_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            fastaprep
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../fastaprep/process_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_fasta
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../genbank/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genbank
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_7" id="__nav_4_1_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/extract_from_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genbank/get_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_genbank
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../genome_metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_8" id="__nav_4_1_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/compare_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/dump_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../genome_metadata/dump_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../gff3/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    gff3
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_9" id="__nav_4_1_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/functional_annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    functional_annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../gff3/process_gff3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_gff3
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../manifest_maker/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_maker
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats" class="md-nav__link">
    manifest_stats
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter" class="md-nav__link">
    BiotypeCounter
  </a>
  
    <nav class="md-nav" aria-label="BiotypeCounter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter.add_id" class="md-nav__link">
    add_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter.unique_count" class="md-nav__link">
    unique_count()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.StatsError" class="md-nav__link">
    StatsError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats" class="md-nav__link">
    manifest_stats
  </a>
  
    <nav class="md-nav" aria-label="manifest_stats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.biotypes_stats" class="md-nav__link">
    biotypes_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.check_ncbi_stats" class="md-nav__link">
    check_ncbi_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.compare_ncbi_counts" class="md-nav__link">
    compare_ncbi_counts()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.coord_systems_stats" class="md-nav__link">
    coord_systems_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.count_biotypes" class="md-nav__link">
    count_biotypes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.get_gff3_stats" class="md-nav__link">
    get_gff3_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.get_manifest" class="md-nav__link">
    get_manifest()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.get_seq_region_stats" class="md-nav__link">
    get_seq_region_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.increment_biotype" class="md-nav__link">
    increment_biotype()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.run" class="md-nav__link">
    run()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.seq_region_special_stats" class="md-nav__link">
    seq_region_special_stats()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_13" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_13" id="__nav_4_1_1_2_1_13_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_13_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_13">
            <span class="md-nav__icon md-icon"></span>
            metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../metadata/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../metadata/prepare_seq_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_seq_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../metadata/update_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    update_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_14" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_14" id="__nav_4_1_1_2_1_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_14_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_14">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../schemas/json_schema_validator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_validator
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../seq_region_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seq_region_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_16" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../utils/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    utils
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_16" id="__nav_4_1_1_2_1_16_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_16_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_16">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../utils/archive_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    archive_utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../utils/json_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [example]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            [example]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../example/calculate/calculations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    calculations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../example/google_docs/google_docs_eg/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    google_docs
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../license/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    License
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats" class="md-nav__link">
    manifest_stats
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter" class="md-nav__link">
    BiotypeCounter
  </a>
  
    <nav class="md-nav" aria-label="BiotypeCounter">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter.add_id" class="md-nav__link">
    add_id()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter.unique_count" class="md-nav__link">
    unique_count()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.StatsError" class="md-nav__link">
    StatsError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats" class="md-nav__link">
    manifest_stats
  </a>
  
    <nav class="md-nav" aria-label="manifest_stats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.biotypes_stats" class="md-nav__link">
    biotypes_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.check_ncbi_stats" class="md-nav__link">
    check_ncbi_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.compare_ncbi_counts" class="md-nav__link">
    compare_ncbi_counts()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.coord_systems_stats" class="md-nav__link">
    coord_systems_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.count_biotypes" class="md-nav__link">
    count_biotypes()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.get_gff3_stats" class="md-nav__link">
    get_gff3_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.get_manifest" class="md-nav__link">
    get_manifest()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.get_seq_region_stats" class="md-nav__link">
    get_seq_region_stats()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.increment_biotype" class="md-nav__link">
    increment_biotype()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.run" class="md-nav__link">
    run()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.manifest_stats.seq_region_special_stats" class="md-nav__link">
    seq_region_special_stats()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.manifest_stats.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>manifest_stats</h1>

<div class="doc doc-object doc-module">



<a id="src.ensembl.io.genomio.manifest_stats"></a>
  <div class="doc doc-contents first">
  
      <p>Compute stats from the current genome files associated with the manifest.</p>

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.manifest_stats.BiotypeCounter" class="doc doc-heading">
          <code>BiotypeCounter</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>A counter for a given biotype, given a list of features.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">BiotypeCounter</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A counter for a given biotype, given a list of features.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">ids</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">example</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">count</span>
        <span class="k">if</span> <span class="n">ids</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">ids</span>
        <span class="k">if</span> <span class="n">example</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">example</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">example</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">example</span>

    <span class="k">def</span> <span class="nf">add_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add a feature to the counter.</span>

<span class="sd">        Args:</span>
<span class="sd">            feature_id (str): Feature id to add.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">unique_count</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Total number feature ids added to the counter so far.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: number of features in the counter.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.BiotypeCounter.add_id" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">add_id</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Add a feature to the counter.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>feature_id</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Feature id to add.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add a feature to the counter.</span>

<span class="sd">    Args:</span>
<span class="sd">        feature_id (str): Feature id to add.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.BiotypeCounter.unique_count" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">unique_count</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Total number feature ids added to the counter so far.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>int</code></td>          <td>
                <code>int</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>number of features in the counter.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">unique_count</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Total number feature ids added to the counter so far.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: number of features in the counter.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ids</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.manifest_stats.InputSchema" class="doc doc-heading">
          <code>InputSchema</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="argschema.ArgSchema">ArgSchema</span></code></p>

  
      <p>Input arguments expected by this script.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InputSchema</span><span class="p">(</span><span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchema</span><span class="p">):</span>  <span class="c1"># need more metadata/&#39;True&#39; requirements</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Input arguments expected by this script.&quot;&quot;&quot;</span>

    <span class="n">manifest_dir</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">String</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Manifest file path&quot;</span><span class="p">})</span>
    <span class="n">accession</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">String</span><span class="p">(</span>
        <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence accession ID to compare stats with NCBI&quot;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">datasets_bin</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">String</span><span class="p">(</span><span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Datasets bin status&quot;</span><span class="p">})</span>
    <span class="n">stats_file</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">files</span><span class="o">.</span><span class="n">OutputFile</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Output file with the stats&quot;</span><span class="p">},</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.manifest_stats.StatsError" class="doc doc-heading">
          <code>StatsError</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Raised when stats could not be computed.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">62</span>
<span class="normal">63</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">StatsError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Raised when stats could not be computed.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.manifest_stats.manifest_stats" class="doc doc-heading">
          <code>manifest_stats</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Representation of the statistics of the set of files listed in the manifest file provided.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">manifest_stats</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Representation of the statistics of the set of files listed in the manifest file provided.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">manifest_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">accession</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">datasets_bin</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_dir</span><span class="si">}</span><span class="s2">/manifest.json&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">accession</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">error</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">datasets_bin</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">datasets_bin</span> <span class="o">=</span> <span class="s2">&quot;datasets&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span> <span class="o">=</span> <span class="n">datasets_bin</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span> <span class="o">=</span> <span class="n">manifest_dir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stats_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute stats in the files and output a stats.txt file in the same folder.</span>

<span class="sd">        Raises:</span>
<span class="sd">            StatsError: Could not compute some stats.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_manifest</span><span class="p">()</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">)</span>

        <span class="c1"># Compute the stats from the GFF3 file</span>
        <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">]))</span>

        <span class="c1"># Compute the stats from the seq_region file</span>
        <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq_region_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>

        <span class="c1"># Print out the stats in a separate file</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">stats_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stats_out</span><span class="p">:</span>
            <span class="n">stats_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">stats</span><span class="p">))</span>

        <span class="c1"># Die if there were errors in stats comparison</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">StatsError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stats count errors, check the file </span><span class="si">{</span><span class="n">stats_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get the files metadata from the manifest json file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dict: A representation of the manifest json data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_json</span><span class="p">:</span>
            <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f_json</span><span class="p">)</span>
            <span class="n">manifest_root</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span>

        <span class="c1"># Use dir name from the manifest</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_root</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                        <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                        <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest_root</span><span class="p">,</span> <span class="n">file_name</span>
                        <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>

        <span class="k">return</span> <span class="n">manifest</span>

    <span class="k">def</span> <span class="nf">get_seq_region_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_region_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compute stats from the seq_region json file.</span>

<span class="sd">        Args:</span>
<span class="sd">            seq_region_path (Path): the seq_region json file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List[str]: Stats from the seq_regions.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="n">seq_region_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
            <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

        <span class="c1"># Get basic data</span>
        <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">circular</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">locations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">codon_tables</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
            <span class="c1"># Get readable seq_region name:</span>
            <span class="c1"># either use a Genbank synonym, or just the provided seq_region name</span>
            <span class="n">genbank</span> <span class="o">=</span> <span class="s2">&quot;synonyms&quot;</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">and</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GenBank&quot;</span><span class="p">]</span>
            <span class="n">seqr_name</span> <span class="o">=</span> <span class="n">genbank</span> <span class="ow">and</span> <span class="n">genbank</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>

            <span class="c1"># Record the lengths of the elements of each coord_system</span>
            <span class="n">coord_level</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">coord_level</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="p">:</span>
                <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>

            <span class="c1"># Additional metadata records to count</span>
            <span class="k">if</span> <span class="s2">&quot;circular&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">circular</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="s2">&quot;codon_table&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">codon_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;codon_table&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
                <span class="n">locations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;location&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Stats</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_region_path</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">coord_systems_stats</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_region_special_stats</span><span class="p">(</span><span class="n">circular</span><span class="p">,</span> <span class="n">locations</span><span class="p">,</span> <span class="n">codon_tables</span><span class="p">)</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">coord_systems_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;For each coord_system compute various stats:</span>
<span class="sd">            - number of sequences</span>
<span class="sd">            - sequence length sum, minimum, maximum, mean</span>

<span class="sd">        Args:</span>
<span class="sd">            coord_systems: Coordinate system dictionary of lengths.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total coord_systems </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">coord_name</span><span class="p">,</span> <span class="n">lengths</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Coord_system: </span><span class="si">{</span><span class="n">coord_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">stat_counts</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;Number of sequences&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length sum&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length minimum&quot;</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length maximum&quot;</span><span class="p">:</span> <span class="nb">max</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
                <span class="s2">&quot;Sequence length mean&quot;</span><span class="p">:</span> <span class="n">mean</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="p">}</span>

            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">stat_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">seq_region_special_stats</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">circular</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">locations</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">codon_tables</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Prepare stats in case there are circular regions, specific locations and codon_tables.</span>
<span class="sd">                stats.append(f&quot;{count: 9f}\t{name}&quot;)</span>

<span class="sd">        Args:</span>
<span class="sd">            circular: Number of circular regions. Defaults to 0.</span>
<span class="sd">            locations: The regions and their location. Defaults to None.</span>
<span class="sd">            codon_tables: The regions and their codon_table. Defaults to None.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">circular</span> <span class="ow">or</span> <span class="n">locations</span> <span class="ow">or</span> <span class="n">codon_tables</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Special&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">circular</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="s2">circular sequences&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">locations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">locations</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with location&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">locations</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">loc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">codon_tables</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">codon_tables</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with codon_table&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">codon_tables</span><span class="p">:</span>
                    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">table</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">get_gff3_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the gene models from the GFF3 file and compute stats.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path (Path): the GFF3 file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            List: Stats from the gene model.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">biotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_biotypes</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span>
        <span class="c1"># Compile final stats</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biotypes_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">count_biotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Count the biotypes in a GFF3 file.</span>

<span class="sd">        Args:</span>
<span class="sd">            gff3_path: Path to the GFF3 file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Dictionary of biotype counters.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">feat1</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                    <span class="c1"># Check if the gene contains proteins (CDSs),</span>
                    <span class="c1"># and keep a count of all hierarchies (e.g. gene-mRNA-CDS)</span>
                    <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">for</span> <span class="n">feat2</span> <span class="ow">in</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">feat2</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                            <span class="n">types2</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">}</span>
                            <span class="k">if</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">in</span> <span class="n">types2</span><span class="p">:</span>
                                <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat2</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">for</span> <span class="n">feat3</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">feat3</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                                <span class="k">continue</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span>
                                <span class="n">biotypes</span><span class="p">,</span> <span class="n">feat3</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat3</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span>
                            <span class="p">)</span>

                    <span class="c1"># Main categories counts</span>
                    <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="n">is_protein</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Special case, undefined gene-transcript</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span>
                            <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span>
                            <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transcript&quot;</span>
                        <span class="p">):</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;NONPROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                    <span class="c1"># Total</span>
                    <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">):</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">biotypes</span>

    <span class="k">def</span> <span class="nf">biotypes_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Prepare biotype stats in order of their name.</span>

<span class="sd">        Args:</span>
<span class="sd">            biotypes: Biotypes counters.</span>

<span class="sd">        Returns:</span>
<span class="sd">            A list with the computed statistics in a printable format.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sorted_biotypes</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">biotypes</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">data</span><span class="p">:</span> <span class="n">BiotypeCounter</span> <span class="o">=</span> <span class="n">biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="n">sorted_biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

        <span class="n">stats</span> <span class="o">=</span> <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">unique_count</span><span class="p">()</span><span class="si">:</span><span class="s2">&gt;9</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">biotype</span><span class="si">:</span><span class="s2">&lt;20</span><span class="si">}</span><span class="se">\t</span><span class="s2">ID = </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">example</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">biotype</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sorted_biotypes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">]</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">check_ncbi_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Use the dataset tool from NCBI to get stats and compare with what we have&quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span>

        <span class="n">datasets_bin</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">which</span><span class="p">(</span><span class="n">datasets_bin</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">stats</span>

        <span class="c1"># Get the dataset summary from NCBI</span>
        <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">datasets_bin</span><span class="p">,</span> <span class="s2">&quot;summary&quot;</span><span class="p">,</span> <span class="s2">&quot;genome&quot;</span><span class="p">,</span> <span class="s2">&quot;accession&quot;</span><span class="p">,</span> <span class="n">accession</span><span class="p">]</span>
        <span class="n">result_out</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">result_out</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>

        <span class="c1"># Get stats</span>
        <span class="k">if</span> <span class="s2">&quot;reports&quot;</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
            <span class="n">genome</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;reports&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;annotation_info&quot;</span> <span class="ow">in</span> <span class="n">genome</span> <span class="ow">and</span> <span class="s2">&quot;stats&quot;</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">]:</span>
                <span class="n">ncbi_stats</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>

                <span class="k">if</span> <span class="s2">&quot;gene_counts&quot;</span> <span class="ow">in</span> <span class="n">ncbi_stats</span><span class="p">:</span>
                    <span class="n">counts</span> <span class="o">=</span> <span class="n">ncbi_stats</span><span class="p">[</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">]</span>
                    <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_ncbi_counts</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">def</span> <span class="nf">compare_ncbi_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">ncbi</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compare specific gene stats from NCBI&quot;&quot;&quot;</span>
        <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">maps</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span><span class="s2">&quot;total&quot;</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;PROT_gene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;non_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;NONPROT_gene&quot;</span><span class="p">],</span>
            <span class="p">[</span><span class="s2">&quot;other&quot;</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">],</span>
        <span class="p">]</span>

        <span class="k">for</span> <span class="n">count_map</span> <span class="ow">in</span> <span class="n">maps</span><span class="p">:</span>
            <span class="n">ncbi_name</span><span class="p">,</span> <span class="n">prep_name</span> <span class="o">=</span> <span class="n">count_map</span>
            <span class="n">ncbi_count</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ncbi_name</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
            <span class="n">preped</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="n">biotypes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">prep_name</span><span class="p">)</span>
            <span class="n">prep_count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">preped</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">prep_count</span> <span class="o">=</span> <span class="n">preped</span><span class="o">.</span><span class="n">count</span>

            <span class="k">if</span> <span class="n">prep_count</span> <span class="o">!=</span> <span class="n">ncbi_count</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">prep_count</span> <span class="o">-</span> <span class="n">ncbi_count</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DIFF gene count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">ncbi_count</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">diff</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">error</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Same count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">stats</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">feature_biotype</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Add the feature to their respective biotype counter.</span>

<span class="sd">        Args:</span>
<span class="sd">            biotypes (Dict[str, BiotypeCounter]): All current biotypes, with their counter.</span>
<span class="sd">            feature_id (str): Feature id to be counted.</span>
<span class="sd">            feature_biotype (str): The biotype of the feature.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">feature_biotype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">biotypes</span><span class="p">:</span>
            <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span> <span class="o">=</span> <span class="n">BiotypeCounter</span><span class="p">(</span><span class="n">example</span><span class="o">=</span><span class="n">feature_id</span><span class="p">)</span>
        <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span><span class="o">.</span><span class="n">add_id</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.biotypes_stats" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">biotypes_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Prepare biotype stats in order of their name.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>biotypes</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.manifest_stats.BiotypeCounter" href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter">BiotypeCounter</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Biotypes counters.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list with the computed statistics in a printable format.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">biotypes_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepare biotype stats in order of their name.</span>

<span class="sd">    Args:</span>
<span class="sd">        biotypes: Biotypes counters.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list with the computed statistics in a printable format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sorted_biotypes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">biotypes</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
        <span class="n">data</span><span class="p">:</span> <span class="n">BiotypeCounter</span> <span class="o">=</span> <span class="n">biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
        <span class="n">sorted_biotypes</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">unique_count</span><span class="p">()</span><span class="si">:</span><span class="s2">&gt;9</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">biotype</span><span class="si">:</span><span class="s2">&lt;20</span><span class="si">}</span><span class="se">\t</span><span class="s2">ID = </span><span class="si">{</span><span class="n">data</span><span class="o">.</span><span class="n">example</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">biotype</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span> <span class="ow">in</span> <span class="n">sorted_biotypes</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.check_ncbi_stats" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">check_ncbi_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Use the dataset tool from NCBI to get stats and compare with what we have</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_ncbi_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Use the dataset tool from NCBI to get stats and compare with what we have&quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="n">accession</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span>

    <span class="n">datasets_bin</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">datasets_bin</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">which</span><span class="p">(</span><span class="n">datasets_bin</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">stats</span>

    <span class="c1"># Get the dataset summary from NCBI</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="n">datasets_bin</span><span class="p">,</span> <span class="s2">&quot;summary&quot;</span><span class="p">,</span> <span class="s2">&quot;genome&quot;</span><span class="p">,</span> <span class="s2">&quot;accession&quot;</span><span class="p">,</span> <span class="n">accession</span><span class="p">]</span>
    <span class="n">result_out</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">result_out</span><span class="o">.</span><span class="n">stdout</span><span class="p">)</span>

    <span class="c1"># Get stats</span>
    <span class="k">if</span> <span class="s2">&quot;reports&quot;</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s2">&quot;reports&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="s2">&quot;annotation_info&quot;</span> <span class="ow">in</span> <span class="n">genome</span> <span class="ow">and</span> <span class="s2">&quot;stats&quot;</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">]:</span>
            <span class="n">ncbi_stats</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="s2">&quot;annotation_info&quot;</span><span class="p">][</span><span class="s2">&quot;stats&quot;</span><span class="p">]</span>

            <span class="k">if</span> <span class="s2">&quot;gene_counts&quot;</span> <span class="ow">in</span> <span class="n">ncbi_stats</span><span class="p">:</span>
                <span class="n">counts</span> <span class="o">=</span> <span class="n">ncbi_stats</span><span class="p">[</span><span class="s2">&quot;gene_counts&quot;</span><span class="p">]</span>
                <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_ncbi_counts</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">counts</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.compare_ncbi_counts" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">compare_ncbi_counts</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">ncbi</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Compare specific gene stats from NCBI</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compare_ncbi_counts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">ncbi</span><span class="p">:</span> <span class="n">Dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare specific gene stats from NCBI&quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">maps</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span><span class="s2">&quot;total&quot;</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;PROT_gene&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;non_coding&quot;</span><span class="p">,</span> <span class="s2">&quot;NONPROT_gene&quot;</span><span class="p">],</span>
        <span class="p">[</span><span class="s2">&quot;other&quot;</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">],</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">count_map</span> <span class="ow">in</span> <span class="n">maps</span><span class="p">:</span>
        <span class="n">ncbi_name</span><span class="p">,</span> <span class="n">prep_name</span> <span class="o">=</span> <span class="n">count_map</span>
        <span class="n">ncbi_count</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">ncbi_name</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">preped</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="n">biotypes</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">prep_name</span><span class="p">)</span>
        <span class="n">prep_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">preped</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">prep_count</span> <span class="o">=</span> <span class="n">preped</span><span class="o">.</span><span class="n">count</span>

        <span class="k">if</span> <span class="n">prep_count</span> <span class="o">!=</span> <span class="n">ncbi_count</span><span class="p">:</span>
            <span class="n">diff</span> <span class="o">=</span> <span class="n">prep_count</span> <span class="o">-</span> <span class="n">ncbi_count</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DIFF gene count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">ncbi_count</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">diff</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">error</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Same count for </span><span class="si">{</span><span class="n">count_map</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">prep_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.coord_systems_stats" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">coord_systems_stats</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  

<details class="for-each-coord_system-compute-various-stats" open>
  <summary>For each coord_system compute various stats</summary>
  <ul>
<li>number of sequences</li>
<li>sequence length sum, minimum, maximum, mean</li>
</ul>
</details>


  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>coord_systems</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[int]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Coordinate system dictionary of lengths.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list with the computed statistics in a printable format.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">coord_systems_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;For each coord_system compute various stats:</span>
<span class="sd">        - number of sequences</span>
<span class="sd">        - sequence length sum, minimum, maximum, mean</span>

<span class="sd">    Args:</span>
<span class="sd">        coord_systems: Coordinate system dictionary of lengths.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list with the computed statistics in a printable format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total coord_systems </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">coord_name</span><span class="p">,</span> <span class="n">lengths</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Coord_system: </span><span class="si">{</span><span class="n">coord_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">stat_counts</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;Number of sequences&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length sum&quot;</span><span class="p">:</span> <span class="nb">sum</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length minimum&quot;</span><span class="p">:</span> <span class="nb">min</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length maximum&quot;</span><span class="p">:</span> <span class="nb">max</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
            <span class="s2">&quot;Sequence length mean&quot;</span><span class="p">:</span> <span class="n">mean</span><span class="p">(</span><span class="n">lengths</span><span class="p">),</span>
        <span class="p">}</span>

        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">stat_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">count</span><span class="si">:</span><span class="s2"> 9f</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.count_biotypes" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">count_biotypes</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Count the biotypes in a GFF3 file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gff3_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Path to the GFF3 file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.manifest_stats.BiotypeCounter" href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter">BiotypeCounter</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Dictionary of biotype counters.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">count_biotypes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Count the biotypes in a GFF3 file.</span>

<span class="sd">    Args:</span>
<span class="sd">        gff3_path: Path to the GFF3 file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dictionary of biotype counters.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">with</span> <span class="n">open_gz_file</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff3_handle</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">GFF</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gff3_handle</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">feat1</span> <span class="ow">in</span> <span class="n">rec</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
                <span class="c1"># Check if the gene contains proteins (CDSs),</span>
                <span class="c1"># and keep a count of all hierarchies (e.g. gene-mRNA-CDS)</span>
                <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">feat2</span> <span class="ow">in</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">feat2</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">:</span>
                        <span class="n">types2</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">}</span>
                        <span class="k">if</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">in</span> <span class="n">types2</span><span class="p">:</span>
                            <span class="n">is_protein</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat2</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">feat3</span> <span class="ow">in</span> <span class="n">feat2</span><span class="o">.</span><span class="n">sub_features</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">feat3</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span>
                            <span class="k">continue</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span>
                            <span class="n">biotypes</span><span class="p">,</span> <span class="n">feat3</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat2</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">feat3</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span>
                        <span class="p">)</span>

                <span class="c1"># Main categories counts</span>
                <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">:</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">is_protein</span><span class="p">:</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;PROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Special case, undefined gene-transcript</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span>
                        <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span>
                        <span class="ow">and</span> <span class="n">feat1</span><span class="o">.</span><span class="n">sub_features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;transcript&quot;</span>
                    <span class="p">):</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;OTHER&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;NONPROT_</span><span class="si">{</span><span class="n">feat1</span><span class="o">.</span><span class="n">type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="c1"># Total</span>
                <span class="k">if</span> <span class="n">feat1</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;pseudogene&quot;</span><span class="p">):</span>
                    <span class="n">manifest_stats</span><span class="o">.</span><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feat1</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;ALL_GENES&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">biotypes</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.get_gff3_stats" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_gff3_stats</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Extract the gene models from the GFF3 file and compute stats.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>gff3_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the GFF3 file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>List</code></td>          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Stats from the gene model.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_gff3_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gff3_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract the gene models from the GFF3 file and compute stats.</span>

<span class="sd">    Args:</span>
<span class="sd">        gff3_path (Path): the GFF3 file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List: Stats from the gene model.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">biotypes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">count_biotypes</span><span class="p">(</span><span class="n">gff3_path</span><span class="p">)</span>
    <span class="c1"># Compile final stats</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">biotypes_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
    <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_ncbi_stats</span><span class="p">(</span><span class="n">biotypes</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.get_manifest" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_manifest</span><span class="p">()</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Get the files metadata from the manifest json file.</p>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>Dict</code></td>          <td>
                <code><span title="typing.Dict">Dict</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A representation of the manifest json data.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_manifest</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the files metadata from the manifest json file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict: A representation of the manifest json data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">manifest</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_json</span><span class="p">:</span>
        <span class="n">manifest</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f_json</span><span class="p">)</span>
        <span class="n">manifest_root</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">manifest_parent</span>

    <span class="c1"># Use dir name from the manifest</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
            <span class="n">file_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">manifest_root</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">]:</span>
                <span class="k">if</span> <span class="s2">&quot;file&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]:</span>
                    <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">][</span><span class="s2">&quot;file&quot;</span><span class="p">]</span>
                    <span class="n">file_name</span> <span class="o">=</span> <span class="n">manifest_root</span><span class="p">,</span> <span class="n">file_name</span>
                    <span class="n">manifest</span><span class="p">[</span><span class="n">name</span><span class="p">][</span><span class="n">f</span><span class="p">]</span> <span class="o">=</span> <span class="n">file_name</span>

    <span class="k">return</span> <span class="n">manifest</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.get_seq_region_stats" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_seq_region_stats</span><span class="p">(</span><span class="n">seq_region_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Compute stats from the seq_region json file.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>seq_region_path</code></td>
          <td>
                <code><span title="pathlib.Path">Path</span></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>the seq_region json file.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>List[str]: Stats from the seq_regions.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_seq_region_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq_region_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute stats from the seq_region json file.</span>

<span class="sd">    Args:</span>
<span class="sd">        seq_region_path (Path): the seq_region json file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: Stats from the seq_regions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">seq_region_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
        <span class="n">seq_regions</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>

    <span class="c1"># Get basic data</span>
    <span class="n">coord_systems</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">circular</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">locations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">codon_tables</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seqr</span> <span class="ow">in</span> <span class="n">seq_regions</span><span class="p">:</span>
        <span class="c1"># Get readable seq_region name:</span>
        <span class="c1"># either use a Genbank synonym, or just the provided seq_region name</span>
        <span class="n">genbank</span> <span class="o">=</span> <span class="s2">&quot;synonyms&quot;</span> <span class="ow">in</span> <span class="n">seqr</span> <span class="ow">and</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;synonyms&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;source&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;GenBank&quot;</span><span class="p">]</span>
        <span class="n">seqr_name</span> <span class="o">=</span> <span class="n">genbank</span> <span class="ow">and</span> <span class="n">genbank</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>

        <span class="c1"># Record the lengths of the elements of each coord_system</span>
        <span class="n">coord_level</span> <span class="o">=</span> <span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;coord_system_level&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">coord_level</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">coord_systems</span><span class="p">:</span>
            <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">coord_systems</span><span class="p">[</span><span class="n">coord_level</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqr</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">])</span>

        <span class="c1"># Additional metadata records to count</span>
        <span class="k">if</span> <span class="s2">&quot;circular&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
            <span class="n">circular</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="s2">&quot;codon_table&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
            <span class="n">codon_tables</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;codon_table&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;location&quot;</span> <span class="ow">in</span> <span class="n">seqr</span><span class="p">:</span>
            <span class="n">locations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seqr_name</span><span class="si">}</span><span class="s2"> = </span><span class="si">{</span><span class="n">seqr</span><span class="p">[</span><span class="s1">&#39;location&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Stats</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_region_path</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">coord_systems_stats</span><span class="p">(</span><span class="n">coord_systems</span><span class="p">)</span>
    <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_region_special_stats</span><span class="p">(</span><span class="n">circular</span><span class="p">,</span> <span class="n">locations</span><span class="p">,</span> <span class="n">codon_tables</span><span class="p">)</span>
    <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.increment_biotype" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">,</span> <span class="n">feature_id</span><span class="p">,</span> <span class="n">feature_biotype</span><span class="p">)</span></code>
  
  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h3>


  <div class="doc doc-contents ">
  
      <p>Add the feature to their respective biotype counter.</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>biotypes</code></td>
          <td>
                <code><span title="typing.Dict">Dict</span>[str, <a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.manifest_stats.BiotypeCounter" href="#src.ensembl.io.genomio.manifest_stats.BiotypeCounter">BiotypeCounter</a>]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>All current biotypes, with their counter.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feature_id</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Feature id to be counted.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>feature_biotype</code></td>
          <td>
                <code>str</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The biotype of the feature.</p>
            </div>
          </td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">increment_biotype</span><span class="p">(</span><span class="n">biotypes</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">BiotypeCounter</span><span class="p">],</span> <span class="n">feature_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">feature_biotype</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add the feature to their respective biotype counter.</span>

<span class="sd">    Args:</span>
<span class="sd">        biotypes (Dict[str, BiotypeCounter]): All current biotypes, with their counter.</span>
<span class="sd">        feature_id (str): Feature id to be counted.</span>
<span class="sd">        feature_biotype (str): The biotype of the feature.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">feature_biotype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">biotypes</span><span class="p">:</span>
        <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span> <span class="o">=</span> <span class="n">BiotypeCounter</span><span class="p">(</span><span class="n">example</span><span class="o">=</span><span class="n">feature_id</span><span class="p">)</span>
    <span class="n">biotypes</span><span class="p">[</span><span class="n">feature_biotype</span><span class="p">]</span><span class="o">.</span><span class="n">add_id</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.run" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">run</span><span class="p">(</span><span class="n">stats_path</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Compute stats in the files and output a stats.txt file in the same folder.</p>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.manifest_stats.StatsError" href="#src.ensembl.io.genomio.manifest_stats.StatsError">StatsError</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Could not compute some stats.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stats_path</span><span class="p">:</span> <span class="n">PathLike</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute stats in the files and output a stats.txt file in the same folder.</span>

<span class="sd">    Raises:</span>
<span class="sd">        StatsError: Could not compute some stats.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">manifest</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_manifest</span><span class="p">()</span>

    <span class="n">stats</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">accession</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">accession</span><span class="p">)</span>

    <span class="c1"># Compute the stats from the GFF3 file</span>
    <span class="k">if</span> <span class="s2">&quot;gff3&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gff3_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;gff3&quot;</span><span class="p">]))</span>

    <span class="c1"># Compute the stats from the seq_region file</span>
    <span class="k">if</span> <span class="s2">&quot;seq_region&quot;</span> <span class="ow">in</span> <span class="n">manifest</span><span class="p">:</span>
        <span class="n">stats</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq_region_stats</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">manifest</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]))</span>

    <span class="c1"># Print out the stats in a separate file</span>
    <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">stats_path</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">stats_out</span><span class="p">:</span>
        <span class="n">stats_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">stats</span><span class="p">))</span>

    <span class="c1"># Die if there were errors in stats comparison</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">StatsError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Stats count errors, check the file </span><span class="si">{</span><span class="n">stats_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.manifest_stats.manifest_stats.seq_region_special_stats" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">seq_region_special_stats</span><span class="p">(</span><span class="n">circular</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">locations</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">codon_tables</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Prepare stats in case there are circular regions, specific locations and codon_tables.
        stats.append(f"{count: 9f}      {name}")</p>



  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>circular</code></td>
          <td>
                <code>int</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>Number of circular regions. Defaults to 0.</p>
            </div>
          </td>
          <td>
                <code>0</code>
          </td>
        </tr>
        <tr>
          <td><code>locations</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The regions and their location. Defaults to None.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
        <tr>
          <td><code>codon_tables</code></td>
          <td>
                <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>The regions and their codon_table. Defaults to None.</p>
            </div>
          </td>
          <td>
                <code>None</code>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code><span title="typing.List">List</span>[str]</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>A list with the computed statistics in a printable format.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">seq_region_special_stats</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">circular</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">locations</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">codon_tables</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prepare stats in case there are circular regions, specific locations and codon_tables.</span>
<span class="sd">            stats.append(f&quot;{count: 9f}\t{name}&quot;)</span>

<span class="sd">    Args:</span>
<span class="sd">        circular: Number of circular regions. Defaults to 0.</span>
<span class="sd">        locations: The regions and their location. Defaults to None.</span>
<span class="sd">        codon_tables: The regions and their codon_table. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        A list with the computed statistics in a printable format.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">stats</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">circular</span> <span class="ow">or</span> <span class="n">locations</span> <span class="ow">or</span> <span class="n">codon_tables</span><span class="p">:</span>
        <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Special&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">circular</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="se">\t</span><span class="s2">circular sequences&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">locations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">locations</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with location&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">loc</span> <span class="ow">in</span> <span class="n">locations</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">loc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">codon_tables</span><span class="p">:</span>
            <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">codon_tables</span><span class="p">)</span><span class="si">:</span><span class="s2"> 9d</span><span class="si">}</span><span class="s2"> sequences with codon_table&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">table</span> <span class="ow">in</span> <span class="n">codon_tables</span><span class="p">:</span>
                <span class="n">stats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t\t\t</span><span class="si">{</span><span class="n">table</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">stats</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>



<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.manifest_stats.main" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Main entrypoint.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/manifest_stats.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main entrypoint.&quot;&quot;&quot;</span>
    <span class="n">mod</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchemaParser</span><span class="p">(</span><span class="n">schema_type</span><span class="o">=</span><span class="n">InputSchema</span><span class="p">)</span>
    <span class="n">mstats</span> <span class="o">=</span> <span class="n">manifest_stats</span><span class="p">(</span><span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;manifest_dir&quot;</span><span class="p">],</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;accession&quot;</span><span class="p">),</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;datasets_bin&quot;</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;accession&quot;</span><span class="p">):</span>
        <span class="n">mstats</span><span class="o">.</span><span class="n">check_ncbi</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="n">stats_file</span><span class="p">:</span> <span class="n">Path</span>
    <span class="k">if</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;stats_file&quot;</span><span class="p">):</span>
        <span class="n">stats_file</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;stats_file&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">stats_file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;manifest_dir&quot;</span><span class="p">))</span> <span class="o">/</span> <span class="s2">&quot;stats.txt&quot;</span>

    <span class="n">mstats</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">stats_file</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../../../..", "features": [], "search": "../../../../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
    
  </body>
</html>