
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../">
      
      
        <link rel="next" href="../get_genbank/">
      
      
      <link rel="icon" href="../../../../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.7">
    
    
      
        <title>extract_from_genbank - Ensembl-Genomio</title>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/stylesheets/main.4b4a2bd9.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../../../.." title="Ensembl-Genomio" class="md-header__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ensembl-Genomio
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              extract_from_genbank
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../../../.." title="Ensembl-Genomio" class="md-nav__button md-logo" aria-label="Ensembl-Genomio" data-md-component="logo">
      
  <img src="../../../../../../images/logo.png" alt="logo">

    </a>
    Ensembl-Genomio
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    src
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            src
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ensembl
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1" id="__nav_4_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [brc4]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [brc4]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1_1_1" id="__nav_4_1_1_1_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [runnable]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_1_1">
            <span class="md-nav__icon md-icon"></span>
            [runnable]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_fasta
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/compare_report/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_report
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/core_server/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    core_server
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/download_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    download_genbank
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/dump_stable_ids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_stable_ids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/extract_from_gb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    extract_from_gb
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/fill_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fill_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/gff3_specifier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gff3_specifier
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/load_sequence_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    load_sequence_data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/read_json/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    read_json
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/say_accession/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    say_accession
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../brc4/runnable/seqregion_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seqregion_parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2" id="__nav_4_1_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [io]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_4_1_1_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2">
            <span class="md-nav__icon md-icon"></span>
            [io]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1" checked>
        
          
          <label class="md-nav__link" for="__nav_4_1_1_2_1" id="__nav_4_1_1_2_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [genomio]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="5" aria-labelledby="__nav_4_1_1_2_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1">
            <span class="md-nav__icon md-icon"></span>
            [genomio]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../assembly/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    assembly
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_1" id="__nav_4_1_1_2_1_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_1">
            <span class="md-nav__icon md-icon"></span>
            assembly
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../assembly/get_assembly_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_assembly_data
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../db_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    db_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../events/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    events
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_3" id="__nav_4_1_1_2_1_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_3">
            <span class="md-nav__icon md-icon"></span>
            events
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events/format_events/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    format_events
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../events_loader/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    events_loader
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../fastaprep/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    fastaprep
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_6" id="__nav_4_1_1_2_1_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_6">
            <span class="md-nav__icon md-icon"></span>
            fastaprep
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../fastaprep/process_fasta/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_fasta
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_7" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genbank
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_7" id="__nav_4_1_1_2_1_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_7_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_7">
            <span class="md-nav__icon md-icon"></span>
            genbank
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    extract_from_genbank
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    extract_from_genbank
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank" class="md-nav__link">
    extract_from_genbank
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator" class="md-nav__link">
    FormattedFilesGenerator
  </a>
  
    <nav class="md-nav" aria-label="FormattedFilesGenerator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.extract_gb" class="md-nav__link">
    extract_gb()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.parse_genbank" class="md-nav__link">
    parse_genbank()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.set_prefix" class="md-nav__link">
    set_prefix()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.set_production_name" class="md-nav__link">
    set_production_name()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.GBParseError" class="md-nav__link">
    GBParseError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.GenomeFiles" class="md-nav__link">
    GenomeFiles
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.UnsupportedData" class="md-nav__link">
    UnsupportedData
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../get_genbank/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    get_genbank
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../genome_metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    genome_metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_8" id="__nav_4_1_1_2_1_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_8">
            <span class="md-nav__icon md-icon"></span>
            genome_metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/compare_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    compare_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/dump_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../genome_metadata/dump_genome_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dump_genome_stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../gff3/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    gff3
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_9" id="__nav_4_1_1_2_1_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_9">
            <span class="md-nav__icon md-icon"></span>
            gff3
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/functional_annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    functional_annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../gff3/process_gff3/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    process_gff3
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../integrity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    integrity
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest_maker/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_maker
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../manifest_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    manifest_stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_13" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../metadata/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    metadata
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_13" id="__nav_4_1_1_2_1_13_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_13_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_13">
            <span class="md-nav__icon md-icon"></span>
            metadata
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/prepare_genome/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_genome
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/prepare_seq_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    prepare_seq_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../metadata/update_genome_metadata/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    update_genome_metadata
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_14" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../schemas/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    schemas
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_14" id="__nav_4_1_1_2_1_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_14_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_14">
            <span class="md-nav__icon md-icon"></span>
            schemas
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json_schema_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../schemas/json_schema_validator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_schema_validator
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../seq_region_dumper/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    seq_region_dumper
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1_2_1_16" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../utils/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    utils
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4_1_1_2_1_16" id="__nav_4_1_1_2_1_16_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="6" aria-labelledby="__nav_4_1_1_2_1_16_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1_2_1_16">
            <span class="md-nav__icon md-icon"></span>
            utils
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/archive_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    archive_utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../utils/json_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    json_utils
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    [example]
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            [example]
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../example/calculate/calculations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    calculations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../example/google_docs/google_docs_eg/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    google_docs
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../nextflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Running Nextflow
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../pipelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Ensembl-Genomio Pipelines
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="../../../../../../license/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    License
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank" class="md-nav__link">
    extract_from_genbank
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator" class="md-nav__link">
    FormattedFilesGenerator
  </a>
  
    <nav class="md-nav" aria-label="FormattedFilesGenerator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.extract_gb" class="md-nav__link">
    extract_gb()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.parse_genbank" class="md-nav__link">
    parse_genbank()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.set_prefix" class="md-nav__link">
    set_prefix()
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.set_production_name" class="md-nav__link">
    set_production_name()
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.GBParseError" class="md-nav__link">
    GBParseError
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.GenomeFiles" class="md-nav__link">
    GenomeFiles
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.InputSchema" class="md-nav__link">
    InputSchema
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.UnsupportedData" class="md-nav__link">
    UnsupportedData
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.ensembl.io.genomio.genbank.extract_from_genbank.main" class="md-nav__link">
    main()
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>extract_from_genbank</h1>

<div class="doc doc-object doc-module">



<a id="src.ensembl.io.genomio.genbank.extract_from_genbank"></a>
  <div class="doc doc-contents first">
  
      <p>Parse a Genbank file and creates cleaned up files from it:
- DNA fasta
- Peptide fasta
- Gene models GFF3
- seq_regions json
- genome metadata json</p>



  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>GFFPArseError</code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If the structure of the gb file can't be parsed.</p>
            </div>
          </td>
        </tr>
        <tr>
          <td>
                <code><a class="autorefs autorefs-internal" title="src.ensembl.io.genomio.genbank.extract_from_genbank.UnsupportedData" href="#src.ensembl.io.genomio.genbank.extract_from_genbank.UnsupportedData">UnsupportedData</a></code>
          </td>
          <td>
            <div class="doc-md-description">
              <p>If some data is not as expected.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>



  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
<th>Name</th>        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
<td><code>json_output</code></td>          <td>
          </td>
          <td>
            <div class="doc-md-description">
              <p>json file with a dict that contains all genome files created.</p>
            </div>
          </td>
        </tr>
    </tbody>
  </table>

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator" class="doc doc-heading">
          <code>FormattedFilesGenerator</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Contains a parser to load data from a file, and output a set of files that follow our schema
for input into a core database</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">FormattedFilesGenerator</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Contains a parser to load data from a file, and output a set of files that follow our schema</span>
<span class="sd">    for input into a core database</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">locations</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;mitochondrion&quot;</span><span class="p">:</span> <span class="s2">&quot;mitochondrial_chromosome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;apicoplast&quot;</span><span class="p">:</span> <span class="s2">&quot;apicoplast_chromosome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chloroplast&quot;</span><span class="p">:</span> <span class="s2">&quot;chloroplast_chromosome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;chromoplast&quot;</span><span class="p">:</span> <span class="s2">&quot;chromoplast_chromosome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;cyanelle&quot;</span><span class="p">:</span> <span class="s2">&quot;cyanelle_chromosome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;leucoplast&quot;</span><span class="p">:</span> <span class="s2">&quot;leucoplast_chromosome&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">allowed_feat_types</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;transcript&quot;</span><span class="p">,</span>
        <span class="s2">&quot;tRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;rRNA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CDS&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prod_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">gb_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prod_name</span> <span class="o">=</span> <span class="n">prod_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gb_file</span> <span class="o">=</span> <span class="n">gb_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="n">GenomeFiles</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">set_prefix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Define a prefix to add to the feature IDs</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">prefix</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span>

    <span class="k">def</span> <span class="nf">set_production_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prod_name</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Define a production_name for the genome</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">prod_name</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">prod_name</span> <span class="o">=</span> <span class="n">prod_name</span>

    <span class="k">def</span> <span class="nf">extract_gb</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract data from a Genbank file and create files from it.&quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">out_dir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="n">GenomeFiles</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_prefix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">prefix</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_production_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">prod_name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parse_genbank</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gb_file</span><span class="p">))</span>

        <span class="c1"># Output the gff3 file</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span>

    <span class="k">def</span> <span class="nf">parse_genbank</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gb_file</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load a sequence from a Genbank file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">organella</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_organella</span><span class="p">(</span><span class="n">gb_file</span><span class="p">)</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gb_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbh</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
                <span class="c1"># We don&#39;t want the record description (especially for the fasta file)</span>
                <span class="n">record</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">record</span><span class="o">.</span><span class="n">organelle</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">organella</span><span class="p">:</span>
                    <span class="n">record</span><span class="o">.</span><span class="n">organelle</span> <span class="o">=</span> <span class="n">organella</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_write_genome_json</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_genes_gff</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_seq_region_json</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_fasta_dna</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_get_organella</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gb_file</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieve the organelle from the genbank file, using the specific GenBank object,</span>
<span class="sd">        because SeqIO does not support this field</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">organella</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gb_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">GenBank</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbh</span><span class="p">):</span>
                <span class="n">accession</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">version</span>
                <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">q</span><span class="o">.</span><span class="n">key</span> <span class="o">==</span> <span class="s2">&quot;/organelle=&quot;</span><span class="p">:</span>
                        <span class="n">organelle</span> <span class="o">=</span> <span class="n">q</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                        <span class="n">organella</span><span class="p">[</span><span class="n">accession</span><span class="p">]</span> <span class="o">=</span> <span class="n">organelle</span>
        <span class="k">return</span> <span class="n">organella</span>

    <span class="k">def</span> <span class="nf">_write_fasta_dna</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s2">&quot;fasta_dna&quot;</span><span class="p">],</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_fh</span><span class="p">:</span>
            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="p">,</span> <span class="n">fasta_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_write_genes_gff</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract gene models from the record, and write a GFF and peptide fasta file.</span>
<span class="sd">        Raise GBParseError If the IDs in all the records are not unique.&quot;&quot;&quot;</span>
        <span class="n">peptides</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_ids</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="p">:</span>
            <span class="n">new_record</span><span class="p">,</span> <span class="n">rec_ids</span><span class="p">,</span> <span class="n">rec_peptides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
            <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_record</span><span class="p">)</span>
            <span class="n">all_ids</span> <span class="o">+=</span> <span class="n">rec_ids</span>
            <span class="n">peptides</span> <span class="o">+=</span> <span class="n">rec_peptides</span>

        <span class="c1"># Write those records to a clean GFF</span>
        <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s2">&quot;gene_models&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff_fh</span><span class="p">:</span>
            <span class="n">GFF</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">gff_fh</span><span class="p">)</span>

        <span class="c1"># Write the peptide sequences to a fasta file</span>
        <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s2">&quot;fasta_pep&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_fh</span><span class="p">:</span>
            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">peptides</span><span class="p">,</span> <span class="n">fasta_fh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

        <span class="c1"># Warn if some IDs are not unique</span>
        <span class="n">count</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">all_ids</span><span class="p">))</span>
        <span class="n">num_duplicates</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">count</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">count</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">num_duplicates</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ID </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> is duplicated </span><span class="si">{</span><span class="n">count</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="si">}</span><span class="s2"> times&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_duplicates</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">GBParseError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Some </span><span class="si">{</span><span class="n">num_duplicates</span><span class="si">}</span><span class="s2"> IDs are duplicated&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_parse_record</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
        <span class="n">all_ids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">peptides</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="c1"># Silently skip any unsupported feature type</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">allowed_feat_types</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># Create a clean clone of the feature</span>
            <span class="n">gff_qualifiers</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
            <span class="n">gff_feat</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span>
                <span class="n">location</span><span class="o">=</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span>
                <span class="nb">type</span><span class="o">=</span><span class="n">feat</span><span class="o">.</span><span class="n">type</span><span class="p">,</span>
                <span class="n">strand</span><span class="o">=</span><span class="n">feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="n">qualifiers</span><span class="o">=</span><span class="n">gff_qualifiers</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="c1"># Only Genes should have a name: use either attribute gene or locus_tag</span>
            <span class="n">gene_name</span> <span class="o">=</span> <span class="n">gff_qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">gene_name</span> <span class="o">=</span> <span class="n">gff_qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">,</span> <span class="p">[</span><span class="kc">None</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>

            <span class="c1"># Parse this gene</span>
            <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">gene_feats</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">gene_peptides</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_gene_feat</span><span class="p">(</span><span class="n">gff_feat</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">)</span>
                <span class="n">peptides</span> <span class="o">+=</span> <span class="n">gene_peptides</span>
                <span class="n">feats</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">feats</span><span class="p">,</span> <span class="o">**</span><span class="n">gene_feats</span><span class="p">}</span>
                <span class="n">all_ids</span> <span class="o">+=</span> <span class="n">gene_ids</span>

            <span class="c1"># No gene ID: parse if it is a tRNA or rRNA</span>
            <span class="k">elif</span> <span class="n">gff_feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">):</span>
                <span class="n">rna_feats</span><span class="p">,</span> <span class="n">rna_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parse_rna_feat</span><span class="p">(</span><span class="n">gff_feat</span><span class="p">)</span>
                <span class="n">feats</span> <span class="o">=</span> <span class="p">{</span><span class="o">**</span><span class="n">feats</span><span class="p">,</span> <span class="o">**</span><span class="n">rna_feats</span><span class="p">}</span>
                <span class="n">all_ids</span> <span class="o">+=</span> <span class="n">rna_ids</span>

            <span class="c1"># Any other case? Fail here and check if we shoud support it, or add it to unsupported list</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">GBParseError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No ID for allowed feature: </span><span class="si">{</span><span class="n">feat</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">new_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
        <span class="n">new_record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">feats</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">new_record</span><span class="p">,</span> <span class="n">all_ids</span><span class="p">,</span> <span class="n">peptides</span>

    <span class="k">def</span> <span class="nf">_parse_gene_feat</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">gene_feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">:</span> <span class="nb">str</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqFeature</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]]:</span>
        <span class="n">gene_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="n">gene_name</span>
        <span class="n">gene_qualifiers</span> <span class="o">=</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="n">new_feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">peptides</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_ids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">gene_qualifiers</span><span class="p">:</span>
                <span class="n">gene_feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;pseudogene&quot;</span>
            <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_id</span>
            <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
            <span class="k">if</span> <span class="s2">&quot;gene&quot;</span> <span class="ow">in</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;locus_tag&quot;</span> <span class="ow">in</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">]</span>
            <span class="n">new_feats</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)]</span> <span class="o">=</span> <span class="n">gene_feat</span>
            <span class="n">all_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_id</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">type</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;tRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;rRNA&quot;</span><span class="p">):</span>
            <span class="n">tr_id</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="o">+</span> <span class="s2">&quot;_t1&quot;</span>
            <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tr_id</span>
            <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Parent&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_id</span>
            <span class="k">if</span> <span class="s2">&quot;gene&quot;</span> <span class="ow">in</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;locus_tag&quot;</span> <span class="ow">in</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">]</span>
            <span class="n">new_feats</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">tr_id</span><span class="p">)]</span> <span class="o">=</span> <span class="n">gene_feat</span>
            <span class="n">all_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">tr_id</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;pseudo&quot;</span> <span class="ow">in</span> <span class="n">gene_qualifiers</span><span class="p">:</span>
                <span class="n">gene_feat</span><span class="o">.</span><span class="n">type</span> <span class="o">=</span> <span class="s2">&quot;exon&quot;</span>
            <span class="n">cds_id</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="o">+</span> <span class="s2">&quot;_p1&quot;</span>
            <span class="n">tr_id</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="o">+</span> <span class="s2">&quot;_t1&quot;</span>
            <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds_id</span>
            <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Parent&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tr_id</span>
            <span class="k">if</span> <span class="s2">&quot;gene&quot;</span> <span class="ow">in</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="s2">&quot;locus_tag&quot;</span> <span class="ow">in</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">gene_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">]</span>

            <span class="c1"># Add fasta to pep fasta file</span>
            <span class="k">if</span> <span class="s2">&quot;translation&quot;</span> <span class="ow">in</span> <span class="n">gene_qualifiers</span><span class="p">:</span>
                <span class="n">new_pep_record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="n">gene_qualifiers</span><span class="p">[</span><span class="s2">&quot;translation&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">id</span><span class="o">=</span><span class="n">cds_id</span><span class="p">)</span>
                <span class="n">peptides</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_pep_record</span><span class="p">)</span>

            <span class="c1"># Also create a parent transcript for this translation</span>
            <span class="n">tr_qualifiers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">tr_id</span><span class="p">,</span> <span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="n">gene_name</span><span class="p">,</span> <span class="s2">&quot;Parent&quot;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">}</span>
            <span class="n">gff_tr</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span>
                <span class="n">location</span><span class="o">=</span><span class="n">gene_feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span>
                <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">,</span>
                <span class="n">strand</span><span class="o">=</span><span class="n">gene_feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                <span class="n">qualifiers</span><span class="o">=</span><span class="n">tr_qualifiers</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">new_feats</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">tr_id</span><span class="p">)]</span> <span class="o">=</span> <span class="n">gff_tr</span>
            <span class="n">new_feats</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">cds_id</span><span class="p">)]</span> <span class="o">=</span> <span class="n">gene_feat</span>
            <span class="n">all_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">tr_id</span><span class="p">))</span>
            <span class="n">all_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">cds_id</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">new_feats</span><span class="p">,</span> <span class="n">all_ids</span><span class="p">,</span> <span class="n">peptides</span>

    <span class="k">def</span> <span class="nf">_parse_rna_feat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">rna_feat</span><span class="p">:</span> <span class="n">SeqFeature</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">SeqFeature</span><span class="p">],</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqFeature</span><span class="p">]]:</span>
        <span class="n">new_feats</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">all_ids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">gff_qualifiers</span> <span class="o">=</span> <span class="n">rna_feat</span><span class="o">.</span><span class="n">qualifiers</span>
        <span class="n">feat_name</span> <span class="o">=</span> <span class="n">gff_qualifiers</span><span class="p">[</span><span class="s2">&quot;product&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">gene_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="n">feat_name</span>

        <span class="n">parts</span> <span class="o">=</span> <span class="n">gene_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Shortening gene_id to </span><span class="si">{</span><span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">gene_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_uniquify_id</span><span class="p">(</span><span class="n">gene_id</span><span class="p">,</span> <span class="n">all_ids</span><span class="p">)</span>

        <span class="n">feat_id</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="o">+</span> <span class="s2">&quot;_t1&quot;</span>
        <span class="n">rna_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_id</span>
        <span class="n">rna_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">feat_name</span>
        <span class="n">rna_feat</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;Parent&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_id</span>

        <span class="c1"># Also create a parent gene for this transcript</span>
        <span class="n">gene_qualifiers</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;ID&quot;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
            <span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="n">feat_name</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">gff_gene</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span>
            <span class="n">location</span><span class="o">=</span><span class="n">rna_feat</span><span class="o">.</span><span class="n">location</span><span class="p">,</span>
            <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span>
            <span class="n">strand</span><span class="o">=</span><span class="n">rna_feat</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
            <span class="n">qualifiers</span><span class="o">=</span><span class="n">gene_qualifiers</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">new_feats</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)]</span> <span class="o">=</span> <span class="n">gff_gene</span>
        <span class="n">new_feats</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">feat_id</span><span class="p">)]</span> <span class="o">=</span> <span class="n">rna_feat</span>
        <span class="n">all_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_id</span><span class="p">))</span>
        <span class="n">all_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">feat_id</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">new_feats</span><span class="p">,</span> <span class="n">all_ids</span>

    <span class="k">def</span> <span class="nf">_uniquify_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">all_ids</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Ensure the gene id used is unique,</span>
<span class="sd">        and append a number otherwise, starting at 2</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">new_id</span> <span class="o">=</span> <span class="n">gene_id</span>
        <span class="n">num</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">new_id</span> <span class="ow">in</span> <span class="n">all_ids</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">new_id</span><span class="si">}</span><span class="s2"> exists, update&quot;</span><span class="p">)</span>
            <span class="n">num</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">new_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">num</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using </span><span class="si">{</span><span class="n">new_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">new_id</span>

    <span class="k">def</span> <span class="nf">_write_seq_region_json</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">json_array</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="p">:</span>
            <span class="n">codon_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_codon_table</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">codon_table</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="s2">&quot;Warning: No codon table found. &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;Make sure to change the codon table number in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s1">&#39;seq_region&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> manually &quot;</span>
                        <span class="s2">&quot;if it is not the standard codon table&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>

                <span class="n">codon_table</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">codon_table</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">codon_table</span><span class="p">)</span>
            <span class="n">seq_obj</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;coord_system_level&quot;</span><span class="p">:</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="s2">&quot;circular&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;topology&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;circular&quot;</span><span class="p">),</span>
                <span class="s2">&quot;codon_table&quot;</span><span class="p">:</span> <span class="n">codon_table</span><span class="p">,</span>
                <span class="s2">&quot;length&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="n">seq</span><span class="o">.</span><span class="n">organelle</span><span class="p">:</span>
                <span class="n">seq_obj</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_prepare_location</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">organelle</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">codon_table</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span>
                        <span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Warning: &#39;</span><span class="si">{</span><span class="n">seq</span><span class="o">.</span><span class="n">organelle</span><span class="si">}</span><span class="s2">&#39; is an organelle: &quot;</span>
                            <span class="sa">f</span><span class="s2">&quot;make sure to change the codon table number in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s1">&#39;seq_region&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> &quot;</span>
                            <span class="s2">&quot;manually if it is not the standard codon table&quot;</span>
                        <span class="p">)</span>
                    <span class="p">)</span>

            <span class="c1"># Additional attributes for Ensembl</span>
            <span class="n">seq_obj</span><span class="p">[</span><span class="s2">&quot;added_sequence&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="s2">&quot;assembly_provider&quot;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;GenBank&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;url&quot;</span><span class="p">:</span> <span class="s2">&quot;https://www.ncbi.nlm.nih.gov/genbank&quot;</span><span class="p">,</span>
                <span class="p">},</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">seq_obj</span><span class="p">[</span><span class="s2">&quot;added_sequence&quot;</span><span class="p">][</span><span class="s2">&quot;assembly_provider&quot;</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="s2">&quot;Warning: please add the relevant provider name&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;for the assembly in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s1">&#39;seq_region&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">seq_obj</span><span class="p">[</span><span class="s2">&quot;added_sequence&quot;</span><span class="p">][</span><span class="s2">&quot;assembly_provider&quot;</span><span class="p">][</span><span class="s2">&quot;url&quot;</span><span class="p">]:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="s2">&quot;Warning: please add the relevant provider url&quot;</span>
                        <span class="sa">f</span><span class="s2">&quot; for the assembly in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s1">&#39;seq_region&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                <span class="p">)</span>

            <span class="n">json_array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_obj</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">],</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">seq_fh</span><span class="p">:</span>
            <span class="n">seq_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">json_array</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">_get_codon_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Look at the CDS features to see if they have a codon table</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">feat</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">feat</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">quals</span> <span class="o">=</span> <span class="n">feat</span><span class="o">.</span><span class="n">qualifiers</span>
                <span class="k">if</span> <span class="s2">&quot;transl_table&quot;</span> <span class="ow">in</span> <span class="n">quals</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">quals</span><span class="p">[</span><span class="s2">&quot;transl_table&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">_prepare_location</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">organelle</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given an organelle name, returns the SO term corresponding to its location</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">organelle</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">locations</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">locations</span><span class="p">[</span><span class="n">organelle</span><span class="p">]</span>
        <span class="k">raise</span> <span class="n">UnsupportedData</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unkown organelle: </span><span class="si">{</span><span class="n">organelle</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_write_genome_json</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write a draft for the genome json file</span>
<span class="sd">        Only the production_name is needed, but the rest of the fields need to be given</span>
<span class="sd">        for the validation of the json file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">prod_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">prod_name</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">prod_name</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

        <span class="n">genome_data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;production_name&quot;</span><span class="p">:</span> <span class="n">prod_name</span><span class="p">,</span>
                <span class="s2">&quot;taxonomy_id&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="s2">&quot;assembly&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;accession&quot;</span><span class="p">:</span> <span class="s2">&quot;GCA_000000000&quot;</span><span class="p">,</span> <span class="s2">&quot;version&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
            <span class="s2">&quot;added_seq&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;species&quot;</span><span class="p">][</span><span class="s2">&quot;production_name&quot;</span><span class="p">]:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Warning: please add the relevant production_name for this genome in </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s1">&#39;genome&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="p">]</span>
        <span class="n">genome_data</span><span class="p">[</span><span class="s2">&quot;added_seq&quot;</span><span class="p">][</span><span class="s2">&quot;region_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ids</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">files</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">],</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">genome_fh</span><span class="p">:</span>
            <span class="n">genome_fh</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">genome_data</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.extract_gb" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">extract_gb</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Extract data from a Genbank file and create files from it.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">extract_gb</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathLike</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract data from a Genbank file and create files from it.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">out_dir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">files</span> <span class="o">=</span> <span class="n">GenomeFiles</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">set_prefix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">prefix</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">set_production_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">prod_name</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">parse_genbank</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gb_file</span><span class="p">))</span>

    <span class="c1"># Output the gff3 file</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">files</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.parse_genbank" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">parse_genbank</span><span class="p">(</span><span class="n">gb_file</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Load a sequence from a Genbank file</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">parse_genbank</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gb_file</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load a sequence from a Genbank file</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">organella</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_organella</span><span class="p">(</span><span class="n">gb_file</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gb_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gbh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gbh</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">):</span>
            <span class="c1"># We don&#39;t want the record description (especially for the fasta file)</span>
            <span class="n">record</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">record</span><span class="o">.</span><span class="n">organelle</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">organella</span><span class="p">:</span>
                <span class="n">record</span><span class="o">.</span><span class="n">organelle</span> <span class="o">=</span> <span class="n">organella</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_write_genome_json</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_genes_gff</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_seq_region_json</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_write_fasta_dna</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.set_prefix" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">set_prefix</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Define a prefix to add to the feature IDs</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_prefix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Define a prefix to add to the feature IDs</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">prefix</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h3 id="src.ensembl.io.genomio.genbank.extract_from_genbank.FormattedFilesGenerator.set_production_name" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">set_production_name</span><span class="p">(</span><span class="n">prod_name</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Define a production_name for the genome</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_production_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prod_name</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Define a production_name for the genome</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">prod_name</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">prod_name</span> <span class="o">=</span> <span class="n">prod_name</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.genbank.extract_from_genbank.GBParseError" class="doc doc-heading">
          <code>GBParseError</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>Error when parsing the Genbank file.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">49</span>
<span class="normal">50</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GBParseError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Error when parsing the Genbank file.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.genbank.extract_from_genbank.GenomeFiles" class="doc doc-heading">
          <code>GenomeFiles</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>dict</code></p>

  
      <p>Store the representation of the genome files created.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">GenomeFiles</span><span class="p">(</span><span class="nb">dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Store the representation of the genome files created.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">:</span> <span class="n">PathLike</span> <span class="o">=</span> <span class="n">Path</span><span class="p">())</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="n">out_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">out_dir</span><span class="p">)</span>
        <span class="bp">self</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="s2">&quot;genome.json&quot;</span>
        <span class="bp">self</span><span class="p">[</span><span class="s2">&quot;seq_region&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="s2">&quot;seq_region.json&quot;</span>
        <span class="bp">self</span><span class="p">[</span><span class="s2">&quot;fasta_dna&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="s2">&quot;dna.fasta&quot;</span>
        <span class="bp">self</span><span class="p">[</span><span class="s2">&quot;fasta_pep&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="s2">&quot;pep.fasta&quot;</span>
        <span class="bp">self</span><span class="p">[</span><span class="s2">&quot;gene_models&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out_dir</span> <span class="o">/</span> <span class="s2">&quot;genes.gff&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.genbank.extract_from_genbank.InputSchema" class="doc doc-heading">
          <code>InputSchema</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code><span title="argschema.ArgSchema">ArgSchema</span></code></p>

  
      <p>Input arguments expected by this script.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">InputSchema</span><span class="p">(</span><span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchema</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Input arguments expected by this script.&quot;&quot;&quot;</span>

    <span class="n">prefix</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">String</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;prefix to add required&quot;</span><span class="p">})</span>
    <span class="n">prod_name</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">String</span><span class="p">(</span>
        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;production name for the species&quot;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">gb_file</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">InputFile</span><span class="p">(</span><span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Sequence accession file&quot;</span><span class="p">})</span>
    <span class="n">out_dir</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">OutputDir</span><span class="p">(</span>
        <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;description&quot;</span><span class="p">:</span> <span class="s2">&quot;Output dir where the generated files will be stored (default=current)&quot;</span><span class="p">}</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>

  

  <div class="doc doc-children">











  </div>

  </div>

</div>

<div class="doc doc-object doc-class">




<h2 id="src.ensembl.io.genomio.genbank.extract_from_genbank.UnsupportedData" class="doc doc-heading">
          <code>UnsupportedData</code>


</h2>


  <div class="doc doc-contents ">
          <p class="doc doc-class-bases">
            Bases: <code>Exception</code></p>

  
      <p>When an expected data is not supported by the current parser.</p>

            <details class="quote">
              <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">53</span>
<span class="normal">54</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">UnsupportedData</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;When an expected data is not supported by the current parser.&quot;&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>

  </div>

</div>



<div class="doc doc-object doc-function">




<h2 id="src.ensembl.io.genomio.genbank.extract_from_genbank.main" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">main</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Main script entry-point.</p>

          <details class="quote">
            <summary>Source code in <code>src/ensembl/io/genomio/genbank/extract_from_genbank.py</code></summary>
            <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main script entry-point.&quot;&quot;&quot;</span>
    <span class="n">mod</span> <span class="o">=</span> <span class="n">argschema</span><span class="o">.</span><span class="n">ArgSchemaParser</span><span class="p">(</span><span class="n">schema_type</span><span class="o">=</span><span class="n">InputSchema</span><span class="p">)</span>
    <span class="c1"># mod.args[&quot;metadata_types&quot;] will be a list-like string that needs to be parsed to List[str]</span>
    <span class="n">gb_extractor</span> <span class="o">=</span> <span class="n">FormattedFilesGenerator</span><span class="p">(</span>
        <span class="n">prefix</span><span class="o">=</span><span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;prefix&quot;</span><span class="p">],</span>
        <span class="n">prod_name</span><span class="o">=</span><span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;prod_name&quot;</span><span class="p">],</span>
        <span class="n">gb_file</span><span class="o">=</span><span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="p">[</span><span class="s2">&quot;gb_file&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">gb_output</span> <span class="o">=</span> <span class="n">gb_extractor</span><span class="o">.</span><span class="n">extract_gb</span><span class="p">(</span><span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;out_dir&quot;</span><span class="p">))</span>

    <span class="n">output_json_file</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;json_output&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">output_json_file</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_json_file</span><span class="p">)</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_fh</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">gb_output</span><span class="p">,</span> <span class="n">out_fh</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
          </details>
  </div>

</div>



  </div>

  </div>

</div>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../../../../..", "features": [], "search": "../../../../../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../../../../assets/javascripts/bundle.aecac24b.min.js"></script>
      
    
  </body>
</html>